{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
    "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5",
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2025-03-29T01:14:46.240921Z",
     "iopub.status.busy": "2025-03-29T01:14:46.240710Z",
     "iopub.status.idle": "2025-03-29T01:14:48.267401Z",
     "shell.execute_reply": "2025-03-29T01:14:48.266470Z",
     "shell.execute_reply.started": "2025-03-29T01:14:46.240900Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "trusted": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "/kaggle/input/stanford-rna-3d-folding/sample_submission.csv\n/kaggle/input/stanford-rna-3d-folding/validation_sequences.csv\n/kaggle/input/stanford-rna-3d-folding/test_sequences.csv\n/kaggle/input/stanford-rna-3d-folding/validation_labels.csv\n/kaggle/input/stanford-rna-3d-folding/train_labels.csv\n/kaggle/input/stanford-rna-3d-folding/train_sequences.csv\n/kaggle/input/stanford-rna-3d-folding/MSA/R1108.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8EVR_EC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZDI_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5FJ1_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NBY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DEG_J.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M21_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8U3M_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WW6_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FJE_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2K5Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T29_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7NK4_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AWD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M8K_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WW6_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3JD5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LP9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WB0_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7M57_ii.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6CYT_N.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L2J_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ZAP_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1NYB_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2O33_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5A18_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DU5_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2K66_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LKR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4N_A1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8BTZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6U79_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1BZ2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K4A_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RPT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7SHX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N4L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6X_A7.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KRL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5Y36_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6XRZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8K8B_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7DLZ_Y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1XWP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FQZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L6I_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3A2K_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5G_BB.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KUW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IHX_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UGA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MTJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2K4C_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6SY6_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5XBL_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MHI_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YSH_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_J.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8QO5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9EOW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MT4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6DU5_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4OQ9_3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2K95_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6BZ7_QW.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KUR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ORM_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GIO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9ENF_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6XWW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NOQ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1TXS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OJ8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2ESE_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YMO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7M2T_ss.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T3F_EC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1SZY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HS1_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MQT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1BVJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9CF3_W.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6NOA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1AQO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UYJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V9B_CD.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2FEY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R93_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BU.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/17RA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OPT_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5KH8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2O_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7A01_e2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8QO4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8PM4_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EP2_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3IZ4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZVK_e2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7S36_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NCI_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2ADT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MME_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7URM_v.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5FMZ_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZVK_h2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MFF_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2Y_tA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3JQ4_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YLG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8GAP_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AHU_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6IV6_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FFK_9.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1R2W_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6NM9_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3UZS_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MFG_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JO7_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_Z.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N8V_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7S4U_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QZC_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BQ5_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3AKZ_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7WB1_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4YVJ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8H6E_5A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8YGJ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1Q75_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LYS_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UYP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F9L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QFQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6C4H_x.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1A60_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8DC2_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8XCA_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ATO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2ZJQ_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4DR5_W.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ELE_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM3_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V93_a.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FOQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YSH_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2A_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1TLR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F6Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2FDT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1BAU_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ATW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4TZP_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6MXQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1S34_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6EVJ_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M12_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V7E_Ab.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1M5L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DET_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2J28_8.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ORJ_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M57_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ZRS_6.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AH3_T.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7B9V_5.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LBL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6U_A0.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L3E_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KKA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6FHI_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8JTJ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6OSI_RB.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2J28_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RVO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MC6_m.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2A9L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4ZT9_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6KUT_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1R7W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MI0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LUN_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1E4P_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7B9V_2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MC6_n.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8P0G_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1156.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EBQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LHD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1X18_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6OJ2_QY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NC0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7KUD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7JRS_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7MSF_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8KAI_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JPP_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8C9C_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3PHP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8KAM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1RNG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1IK1_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IXY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LMB_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NR0_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1R4H_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1P6V_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1P6V_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3A3A_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KE6_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WD9_2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8URY_d.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2QH3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NOQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LBS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1IBM_Y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4N_B1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7VTN_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZN1_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MNC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UYL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8E79_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2EVY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2XXA_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K5I_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V65_BB.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7SAM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8H69_5.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ZPQ_6.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7PTL_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L3C_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6W_A8.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7D8C_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N7X_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4ILM_M.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT7_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3HAY_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8IEW_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6S0X_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1XWU_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1WZ2_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RPK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8KI7_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1RHT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1PJY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YSH_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8Z8N_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DKN_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HS8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8QO3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2TPK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8EWG_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4WC3_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T3E_EC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KPV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6N7R_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6O3M_XV.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8CLR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3IZZ_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3CW1_v.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8BU8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MFC_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7S3H_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6D_AV.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MC6_BR.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BJ2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V6X_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ORI_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N3O_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3W1K_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9FO8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BL.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V65_A1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KUU_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1N8X_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1B36_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WFR_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7S4V_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6SDW_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HS2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8BWT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GVO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LYG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UMC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8FCS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8URY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1I4C_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3LWQ_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1AUD_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1I3X_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6UFM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT6_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1138.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7SLP_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1P5N_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7EXY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5KK5_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EXY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9CF1_W.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1P5P_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KIS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V17_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JSE_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IL9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6X_B2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7XW2_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LAC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8CTI_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5GAP_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DU6_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8JTR_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UME_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZVK_d2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4BY9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7QGG_v.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L1F_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6W_A7.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1D0U_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7RGU_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFO_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OPS_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6BZ7_XY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8CI5_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6QX3_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1VOP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ETF_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7YOJ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8J72_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3TUP_T.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F6X_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6OF6_XW.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4N_B2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3PGW_N.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V99_FX.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2AHT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9G8O_S2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IZN_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2Y95_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N6W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4X4S_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QWB_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6W3M_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1L1W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2W2H_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2AGN_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6FLQ_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BS0_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M23_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1X18_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8G1S_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4V_AA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LDL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5N5C_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V16_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5OA3_1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZIG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UBE_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HLX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5ZEY_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZIF_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JUR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5AMQ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1LS2_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8DVS_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2G1W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UMD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OKD_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7A01_K3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UPT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K6H_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1E7K_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QZB_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MEQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5Z3G_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1T4L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K4B_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M22_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J3V_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MS1_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8H6E_4A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GV4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L1F_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1G70_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9DTT_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ANR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JWV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8HUD_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1X18_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6QX9_1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6QWL_W.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3AMU_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1LC6_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LI4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5G_BA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8G1I_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1Y1Y_P.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J3W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6K0B_U.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JOX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KAJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6LTP_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UO1_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1RY1_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KRP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WUS_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6FT6_2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IZ8_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V93_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JR4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8Z90_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4W_B9.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZVK_K3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZBH_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RLU_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4YVK_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8JHP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YNE_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HVU_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OPP_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1TBK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V8J_AV.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5UF3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1UUU_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2AGN_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LVA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DEG_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1NC0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LHP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6S0X_a.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7KUB_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6HYK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V7E_Ae.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7XSN_N.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WFR_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2B6G_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1D6K_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HMH_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9ENE_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4N_A3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6W5C_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WW6_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IL9_M.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QZC_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6O0Y_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4C4Q_N.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MFE_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6XZR_IN1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MNX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MF0_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFJ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1SCL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L5Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SCF_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MSF_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KOC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6VAR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4QIL_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2ZJQ_Y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7RYG_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NR0_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6W_A5.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6W_BC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ULL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M4W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MS0_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1XSG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4WSA_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1126.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7PTQ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AEB_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6RFL_U.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MI0_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5MC6_BS.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2F87_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AWC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GRW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5V6X_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8TOC_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KD8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V7E_Aa.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2P_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JYM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9ENC_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N2P_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1BGZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4A4U_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8Z85_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT8_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8C8X_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EIY_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1T28_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7RQ5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7JRT_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2A64_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1YSV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LBJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F84_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT8_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT7_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EKZ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2M5U_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2D19_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HWQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5GAP_U.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7PTS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4TUA_XY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1A4T_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4YVI_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1IDV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6W64_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1LDZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ESH_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8YE6_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4ILL_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UO0_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2Y_tP.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2P_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RO2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1A1T_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BK.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7D1A_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8T2B_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KPD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2F4X_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BH.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1BIV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_AF.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7M57_cc.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WFQ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2PN9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WQZ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5ELS_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V7E_Ac.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6DTI_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZVK_E1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFH_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8EVQ_EC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6Y0C_IN1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6XZD_IN1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5WQ1_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KMJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8FZT_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BQ.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N0R_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1NA2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1OSW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6PK9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6POM_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1CQ5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DEG_K.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6MCE_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7S38_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/484D_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LU0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6V5B_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2J37_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8Z1G_T.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KZL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4WJ3_Q.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NC1_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1190.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JZC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JTW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8XPP_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GO5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DER_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2HNS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1Z31_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7OZS_3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LPA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N6S_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EP2_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7U2A_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3IYR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_J.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KPD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3IZD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1E95_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8QO2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6GBM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DU4_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WB1_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JTP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ESY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5ZAL_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IZ8_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1116.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LA5_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8CQ1_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6OPE_XY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4TZV_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5LSN_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6U_B1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5EPI_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LV0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8DK7_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LJJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NR0_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BQ5_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4X0A_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4X4P_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F5U_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EQ3_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4Z7L_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1Z30_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_AH.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4WZJ_VV.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LQZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6IFO_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7M5O_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2QH4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ATV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6K4S_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MFK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6RR7_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AGB_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6HYU_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WPI_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1117v2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IXZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GO5_9.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WC1_Q.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N3Q_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DKN_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1RNK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J8G_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2HUA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZC8_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OST_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6VZC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FHK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7ZRZ_ZN1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4G_CA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7QDU_Q.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LJY_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7KUC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LYT_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1L1C_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5O7H_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UO6_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8EXY_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8KAG_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7VW3_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3JB9_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J8G_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1R2P_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MFD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFN_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2U2A_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1128.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LDT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EQ4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ORD_QY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J6B_e.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BS0_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6UES_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MFY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8AW3_1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6W62_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WD2_1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3J3V_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3NDB_M.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1A9N_Q.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KIS_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1XSH_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AHU_T.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8TVZ_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LK3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1AFX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4KJI_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V7E_Ad.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1SLO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KP7_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2BJ2_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1MFJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4A4T_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8WT6_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5KMZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2B2E_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QC8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8Z1F_T.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SA6_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QWA_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5XJ2_G.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7X34_3.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MXJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7K4L_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8E29_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2DR2_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6XWJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5UZT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HS4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4A4S_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EUQ_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1A51_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DEG_I.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GIP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2B2E_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5ZAM_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NBX_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1EBR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6OJ2_QW.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2MXL_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UVT_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1TJZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4X0B_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1S9S_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EP2_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6MJ0_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4P8Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N7M_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N6T_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2EUY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1QZA_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFL_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8D4A_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1U63_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K2G_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UAU_R.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8D49_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6Z6B_UUU.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1RFR_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6K3Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SFI_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1FYO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3EQ3_Y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7V59_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1189.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1OQ0_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4KR6_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KPC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N2O_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8XCC_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1OW9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LC8_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6S0Z_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N6X_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2QH2_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2JXV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/28SP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1WKS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZIH_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BM.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NR0_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LUB_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5F_CA.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6NDK_XY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NBZ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KHY_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8IDF_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7LMA_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2N1Q_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1IKD_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8I0P_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6O0Z_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2FY1_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V4V_B0.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9FO9_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V5Z_BP.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DKN_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7UR5_y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2HGH_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLO_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6C66_J.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5IEM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1P5M_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K6G_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7DD4_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1N34_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JTJ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1I46_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1F85_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7NQ4_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6BZ8_XY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/9G7C_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OM7_H.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1KKS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6KUV_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IY3_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6X_A5.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2L3J_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6WLN_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1107.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6ZDU_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2OB7_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WC2_Q.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2LBK_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6K0A_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6QW6_5.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KUV_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3DEG_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2AKE_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1M82_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2RN1_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6W_B2.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V8J_CW.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7YR7_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8KDA_O.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6X_A8.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2X7N_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8VCI_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4D5N_X.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1Z2J_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1HS3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2PCW_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2NOQ_E.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7OBQ_1.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2GV3_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2R1G_F.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4V6X_BC.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ROQ_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8BD5_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1ZDK_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8UYS_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1XHP_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2D1B_B.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1N66_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8C98_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2HEM_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8OMR_C.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/7V06_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1JWC_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/4TUC_QY.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2KVN_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3WC2_P.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8P0B_V.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2IZM_S.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/5KQE_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1149.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/3PIP_Y.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1R7Z_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/R1136.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/1K9W_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/2F88_A.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/6AAX_D.MSA.fasta\n/kaggle/input/stanford-rna-3d-folding/MSA/8SCH_A.MSA.fasta\n/kaggle/input/rna-structure-predict/rna_predict-0.1.0-py3-none-any.whl\n/kaggle/input/rna-structure-predict/rna_predict-0.1.3-py3-none-any.whl\n/kaggle/input/rna-structure-predict/rna_predict-0.1.1-py3-none-any.whl\n/kaggle/input/rna-structure-predict/rna_predict-0.1.2-py3-none-any.whl\n/kaggle/input/block-sparse-wheels/block_sparse_attn-0.0.1cu118torch2.0cxx11abiTRUE-cp310-cp310-linux_x86_64.whl\n"
    }
   ],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load\n",
    "\n",
    "# Input data files are available in the read-only \"../input/\" directory\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n",
    "import os\n",
    "\n",
    "import numpy as np  # linear algebra\n",
    "import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "\n",
    "for dirname, _, filenames in os.walk(\"/kaggle/input\"):\n",
    "    for filename in filenames:\n",
    "        print(os.path.join(dirname, filename))\n",
    "\n",
    "# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\"\n",
    "# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:14:48.268761Z",
     "iopub.status.busy": "2025-03-29T01:14:48.268278Z",
     "iopub.status.idle": "2025-03-29T01:17:51.675373Z",
     "shell.execute_reply": "2025-03-29T01:17:51.674367Z",
     "shell.execute_reply.started": "2025-03-29T01:14:48.268718Z"
    },
    "trusted": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "\n=== [System Information] ===\n\n[Python Version]\n3.10.12 (main, Nov  6 2024, 20:22:13) [GCC 11.4.0]\n\n[Kernel and OS Information]\nLinux ef5a5af58c89 6.6.56+ #1 SMP PREEMPT_DYNAMIC Sun Nov 10 10:07:59 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux\n\n\n[CPU Information]\nArchitecture:                         x86_64\nCPU op-mode(s):                       32-bit, 64-bit\nAddress sizes:                        46 bits physical, 48 bits virtual\nByte Order:                           Little Endian\nCPU(s):                               4\nOn-line CPU(s) list:                  0-3\nVendor ID:                            GenuineIntel\nModel name:                           Intel(R) Xeon(R) CPU @ 2.00GHz\nCPU family:                           6\nModel:                                85\nThread(s) per core:                   2\nCore(s) per socket:                   2\nSocket(s):                            1\nStepping:                             3\nBogoMIPS:                             4000.33\nFlags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities\nHypervisor vendor:                    KVM\nVirtualization type:                  full\nL1d cache:                            64 KiB (2 instances)\nL1i cache:                            64 KiB (2 instances)\nL2 cache:                             2 MiB (2 instances)\nL3 cache:                             38.5 MiB (1 instance)\nNUMA node(s):                         1\nNUMA node0 CPU(s):                    0-3\nVulnerability Gather data sampling:   Not affected\nVulnerability Itlb multihit:          Not affected\nVulnerability L1tf:                   Mitigation; PTE Inversion\nVulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown\nVulnerability Meltdown:               Mitigation; PTI\nVulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown\nVulnerability Reg file data sampling: Not affected\nVulnerability Retbleed:               Mitigation; IBRS\nVulnerability Spec rstack overflow:   Not affected\nVulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl\nVulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization\nVulnerability Spectre v2:             Mitigation; IBRS; IBPB conditional; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop\nVulnerability Srbds:                  Not affected\nVulnerability Tsx async abort:        Mitigation; Clear CPU buffers; SMT Host state unknown\n\n\n[Memory Information]\n               total        used        free      shared  buff/cache   available\nMem:            31Gi       768Mi        23Gi       1.0Mi       6.8Gi        30Gi\nSwap:             0B          0B          0B\n\n\n[Disk Information]\nNAME    MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS\nloop0     7:0    0    96G  0 loop \nloop1     7:1    0    20G  0 loop /etc/secrets/kaggle\n                                  /kaggle/working\n                                  /kaggle/input\n                                  /kaggle/lib\nsda       8:0    0   128G  0 disk \nsda1    8:1    0 123.8G  0 part /usr/local/cuda-12.2/targets/x86_64-linux/lib/stubs\n                                 /usr/local/nvidia/lib64\n                                 /opt/bin\nsda2    8:2    0    16M  0 part \nsda3    8:3    0     2G  0 part \nsda4    8:4    0    16M  0 part \nsda5    8:5    0     2G  0 part \nsda6    8:6    0   512B  0 part \nsda7    8:7    0   512B  0 part \nsda8    8:8    0    16M  0 part \nsda9    8:9    0   512B  0 part \nsda10   8:10   0   512B  0 part \nsda11   8:11   0     8M  0 part \nsda12   8:12   0    32M  0 part \nsdb       8:16   0     8T  1 disk \n\n\n=== [End of System Information] ===\n\n[INFO] Uninstalling Torch, rna_predict, scikit-learn, SciPy, NumPy, Seaborn...\n\nFound existing installation: torch 2.5.1+cu121\nUninstalling torch-2.5.1+cu121:\n  Successfully uninstalled torch-2.5.1+cu121\nFound existing installation: torchvision 0.20.1+cu121\nUninstalling torchvision-0.20.1+cu121:\n  Successfully uninstalled torchvision-0.20.1+cu121\nFound existing installation: torchaudio 2.5.1+cu121\nUninstalling torchaudio-2.5.1+cu121:\n  Successfully uninstalled torchaudio-2.5.1+cu121\nFound existing installation: scikit-learn 1.2.2\nUninstalling scikit-learn-1.2.2:\n  Successfully uninstalled scikit-learn-1.2.2\nFound existing installation: scipy 1.13.1\nUninstalling scipy-1.13.1:\n  Successfully uninstalled scipy-1.13.1\nFound existing installation: numpy 1.26.4\nUninstalling numpy-1.26.4:\n  Successfully uninstalled numpy-1.26.4\nFound existing installation: seaborn 0.12.2\nUninstalling seaborn-0.12.2:\n  Successfully uninstalled seaborn-0.12.2\n\nWARNING: Skipping rna-predict as it is not installed.\n\n\n[INFO] Installing Torch: torch==2.0.1+cu118 from https://download.pytorch.org/whl/cu118\n\nLooking in indexes: https://download.pytorch.org/whl/cu118\nCollecting torch==2.0.1+cu118\n  Downloading https://download.pytorch.org/whl/cu118/torch-2.0.1%2Bcu118-cp310-cp310-linux_x86_64.whl (2267.3 MB)\n      2.3/2.3 GB 88.3 MB/s eta 0:00:00\nRequirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch==2.0.1+cu118) (3.17.0)\nRequirement already satisfied: typing-extensions in /usr/local/lib/python3.10/dist-packages (from torch==2.0.1+cu118) (4.12.2)\nRequirement already satisfied: sympy in /usr/local/lib/python3.10/dist-packages (from torch==2.0.1+cu118) (1.13.1)\nRequirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch==2.0.1+cu118) (3.4.2)\nRequirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch==2.0.1+cu118) (3.1.4)\nCollecting triton==2.0.0 (from torch==2.0.1+cu118)\n  Downloading https://download.pytorch.org/whl/triton-2.0.0-1-cp310-cp310-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (63.3 MB)\n      63.3/63.3 MB 94.5 MB/s eta 0:00:00\nRequirement already satisfied: cmake in /usr/local/lib/python3.10/dist-packages (from triton==2.0.0->torch==2.0.1+cu118) (3.31.2)\nCollecting lit (from triton==2.0.0->torch==2.0.1+cu118)\n  Downloading https://download.pytorch.org/whl/lit-15.0.7.tar.gz (132 kB)\n      132.3/132.3 kB 303.8 MB/s eta 0:00:00\n  Preparing metadata (setup.py): started\n  Preparing metadata (setup.py): finished with status 'done'\nRequirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch==2.0.1+cu118) (3.0.2)\nRequirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from sympy->torch==2.0.1+cu118) (1.3.0)\nBuilding wheels for collected packages: lit\n  Building wheel for lit (setup.py): started\n  Building wheel for lit (setup.py): finished with status 'done'\n  Created wheel for lit: filename=lit-15.0.7-py3-none-any.whl size=89991 sha256=7a8406cf6584936ef0f81e0d6f13f31f9f2b65af75c07055a405b6e37c35fe1c\n  Stored in directory: /tmp/pip-ephem-wheel-cache-mufbfl9s/wheels/27/2c/b6/3ed2983b1b44fe0dea1bb35234b09f2c22fb8ebb308679c922\nSuccessfully built lit\nInstalling collected packages: lit, triton, torch\nSuccessfully installed lit-15.0.7 torch-2.0.1+cu118 triton-2.0.0\n\nERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\naccelerate 1.2.1 requires numpy<3.0.0,>=1.17, which is not installed.\neasyocr 1.7.2 requires numpy, which is not installed.\neasyocr 1.7.2 requires scipy, which is not installed.\neasyocr 1.7.2 requires torchvision>=0.5, which is not installed.\nfastai 2.7.18 requires scikit-learn, which is not installed.\nfastai 2.7.18 requires scipy, which is not installed.\nfastai 2.7.18 requires torchvision>=0.11, which is not installed.\npeft 0.14.0 requires numpy>=1.17, which is not installed.\nsentence-transformers 3.3.1 requires scikit-learn, which is not installed.\nsentence-transformers 3.3.1 requires scipy, which is not installed.\nstable-baselines3 2.1.0 requires numpy>=1.20, which is not installed.\ntimm 1.0.12 requires torchvision, which is not installed.\ntorchmetrics 1.6.1 requires numpy>1.20.0, which is not installed.\npytorch-lightning 2.5.0.post0 requires torch>=2.1.0, but you have torch 2.0.1+cu118 which is incompatible.\n\n\n[INFO] Copied and renamed block-sparse-attn wheel for PEP 440 compliance.\n\n\n[INFO] Installing block-sparse-attn from: /kaggle/working/block_sparse_attn-0.0.1+cu118torch2.0-cp310-cp310-linux_x86_64.whl\nProcessing ./block_sparse_attn-0.0.1+cu118torch2.0-cp310-cp310-linux_x86_64.whl\nInstalling collected packages: block-sparse-attn\nSuccessfully installed block-sparse-attn-0.0.1\n\n\n[INFO] Installing rna_predict from: /kaggle/input/rna-structure-predict/rna_predict-0.1.3-py3-none-any.whl\nProcessing /kaggle/input/rna-structure-predict/rna_predict-0.1.3-py3-none-any.whl\nInstalling collected packages: rna-predict\nSuccessfully installed rna-predict-0.1.3\n\n\n[INFO] Reinstalling scikit-learn, SciPy, NumPy, Seaborn to fix version conflicts...\nCollecting numpy==1.24.3\n  Downloading numpy-1.24.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.6 kB)\nCollecting scipy==1.10.1\n  Downloading scipy-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (58 kB)\n      58.9/58.9 kB 8.4 MB/s eta 0:00:00\nCollecting scikit-learn==1.2.2\n  Downloading scikit_learn-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)\nCollecting seaborn==0.12.2\n  Downloading seaborn-0.12.2-py3-none-any.whl.metadata (5.4 kB)\nCollecting joblib>=1.1.1 (from scikit-learn==1.2.2)\n  Downloading joblib-1.4.2-py3-none-any.whl.metadata (5.4 kB)\nCollecting threadpoolctl>=2.0.0 (from scikit-learn==1.2.2)\n  Downloading threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)\nCollecting pandas>=0.25 (from seaborn==0.12.2)\n  Downloading pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (89 kB)\n      89.9/89.9 kB 88.7 MB/s eta 0:00:00\nCollecting matplotlib!=3.6.1,>=3.1 (from seaborn==0.12.2)\n  Downloading matplotlib-3.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)\nCollecting contourpy>=1.0.1 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading contourpy-1.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.4 kB)\nCollecting cycler>=0.10 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)\nCollecting fonttools>=4.22.0 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading fonttools-4.56.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (101 kB)\n      101.9/101.9 kB 313.4 MB/s eta 0:00:00\nCollecting kiwisolver>=1.3.1 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading kiwisolver-1.4.8-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl.metadata (6.2 kB)\nCollecting packaging>=20.0 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading packaging-24.2-py3-none-any.whl.metadata (3.2 kB)\nCollecting pillow>=8 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading pillow-11.1.0-cp310-cp310-manylinux_2_28_x86_64.whl.metadata (9.1 kB)\nCollecting pyparsing>=2.3.1 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading pyparsing-3.2.3-py3-none-any.whl.metadata (5.0 kB)\nCollecting python-dateutil>=2.7 (from matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)\nCollecting pytz>=2020.1 (from pandas>=0.25->seaborn==0.12.2)\n  Downloading pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)\nCollecting tzdata>=2022.7 (from pandas>=0.25->seaborn==0.12.2)\n  Downloading tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)\nCollecting six>=1.5 (from python-dateutil>=2.7->matplotlib!=3.6.1,>=3.1->seaborn==0.12.2)\n  Downloading six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)\nDownloading numpy-1.24.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)\n    17.3/17.3 MB 279.8 MB/s eta 0:00:00\nDownloading scipy-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (34.4 MB)\n    34.4/34.4 MB 290.1 MB/s eta 0:00:00\nDownloading scikit_learn-1.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (9.6 MB)\n    9.6/9.6 MB 267.0 MB/s eta 0:00:00\nDownloading seaborn-0.12.2-py3-none-any.whl (293 kB)\n    293.3/293.3 kB 346.8 MB/s eta 0:00:00\nDownloading joblib-1.4.2-py3-none-any.whl (301 kB)\n    301.8/301.8 kB 338.2 MB/s eta 0:00:00\nDownloading matplotlib-3.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.6 MB)\n    8.6/8.6 MB 261.2 MB/s eta 0:00:00\nDownloading pandas-2.2.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n    13.1/13.1 MB 291.8 MB/s eta 0:00:00\nDownloading threadpoolctl-3.6.0-py3-none-any.whl (18 kB)\nDownloading contourpy-1.3.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (324 kB)\n    325.0/325.0 kB 345.3 MB/s eta 0:00:00\nDownloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\nDownloading fonttools-4.56.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.6 MB)\n    4.6/4.6 MB 290.0 MB/s eta 0:00:00\nDownloading kiwisolver-1.4.8-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.6 MB)\n    1.6/1.6 MB 316.8 MB/s eta 0:00:00\nDownloading packaging-24.2-py3-none-any.whl (65 kB)\n    65.5/65.5 kB 225.3 MB/s eta 0:00:00\nDownloading pillow-11.1.0-cp310-cp310-manylinux_2_28_x86_64.whl (4.5 MB)\n    4.5/4.5 MB 247.9 MB/s eta 0:00:00\nDownloading pyparsing-3.2.3-py3-none-any.whl (111 kB)\n    111.1/111.1 kB 331.6 MB/s eta 0:00:00\nDownloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n    229.9/229.9 kB 355.9 MB/s eta 0:00:00\nDownloading pytz-2025.2-py2.py3-none-any.whl (509 kB)\n    509.2/509.2 kB 325.9 MB/s eta 0:00:00\nDownloading tzdata-2025.2-py2.py3-none-any.whl (347 kB)\n    347.8/347.8 kB 355.5 MB/s eta 0:00:00\nDownloading six-1.17.0-py2.py3-none-any.whl (11 kB)\nInstalling collected packages: pytz, tzdata, threadpoolctl, six, pyparsing, pillow, packaging, numpy, kiwisolver, joblib, fonttools, cycler, scipy, python-dateutil, contourpy, scikit-learn, pandas, matplotlib, seaborn\n  Attempting uninstall: pytz\n    Found existing installation: pytz 2025.1\n    Uninstalling pytz-2025.1:\n      Successfully uninstalled pytz-2025.1\n  Attempting uninstall: tzdata\n    Found existing installation: tzdata 2025.1\n    Uninstalling tzdata-2025.1:\n      Successfully uninstalled tzdata-2025.1\n  Attempting uninstall: threadpoolctl\n    Found existing installation: threadpoolctl 3.5.0\n    Uninstalling threadpoolctl-3.5.0:\n      Successfully uninstalled threadpoolctl-3.5.0\n  Attempting uninstall: six\n    Found existing installation: six 1.17.0\n    Uninstalling six-1.17.0:\n      Successfully uninstalled six-1.17.0\n  Attempting uninstall: pyparsing\n    Found existing installation: pyparsing 3.2.0\n    Uninstalling pyparsing-3.2.0:\n      Successfully uninstalled pyparsing-3.2.0\n  Attempting uninstall: pillow\n    Found existing installation: pillow 11.0.0\n    Uninstalling pillow-11.0.0:\n      Successfully uninstalled pillow-11.0.0\n  Attempting uninstall: packaging\n    Found existing installation: packaging 24.2\n    Uninstalling packaging-24.2:\n      Successfully uninstalled packaging-24.2\n  Attempting uninstall: kiwisolver\n    Found existing installation: kiwisolver 1.4.7\n    Uninstalling kiwisolver-1.4.7:\n      Successfully uninstalled kiwisolver-1.4.7\n  Attempting uninstall: joblib\n    Found existing installation: joblib 1.4.2\n    Uninstalling joblib-1.4.2:\n      Successfully uninstalled joblib-1.4.2\n  Attempting uninstall: fonttools\n    Found existing installation: fonttools 4.55.3\n    Uninstalling fonttools-4.55.3:\n      Successfully uninstalled fonttools-4.55.3\n  Attempting uninstall: cycler\n    Found existing installation: cycler 0.12.1\n    Uninstalling cycler-0.12.1:\n      Successfully uninstalled cycler-0.12.1\n  Attempting uninstall: python-dateutil\n    Found existing installation: python-dateutil 2.9.0.post0\n    Uninstalling python-dateutil-2.9.0.post0:\n      Successfully uninstalled python-dateutil-2.9.0.post0\n  Attempting uninstall: contourpy\n    Found existing installation: contourpy 1.3.1\n    Uninstalling contourpy-1.3.1:\n      Successfully uninstalled contourpy-1.3.1\n  Attempting uninstall: pandas\n    Found existing installation: pandas 2.2.3\n    Uninstalling pandas-2.2.3:\n      Successfully uninstalled pandas-2.2.3\n  Attempting uninstall: matplotlib\n    Found existing installation: matplotlib 3.7.5\n    Uninstalling matplotlib-3.7.5:\n      Successfully uninstalled matplotlib-3.7.5\nSuccessfully installed contourpy-1.3.1 cycler-0.12.1 fonttools-4.56.0 joblib-1.4.2 kiwisolver-1.4.8 matplotlib-3.10.1 numpy-1.24.3 packaging-24.2 pandas-2.2.3 pillow-11.1.0 pyparsing-3.2.3 python-dateutil-2.9.0.post0 pytz-2025.2 scikit-learn-1.2.2 scipy-1.10.1 seaborn-0.12.2 six-1.17.0 threadpoolctl-3.6.0 tzdata-2025.2\n\nERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\neasyocr 1.7.2 requires torchvision>=0.5, which is not installed.\nfastai 2.7.18 requires torchvision>=0.11, which is not installed.\nrna-predict 0.1.3 requires biopython>=1.85, which is not installed.\nrna-predict 0.1.3 requires mdanalysis>=2.9.0, which is not installed.\ntimm 1.0.12 requires torchvision, which is not installed.\nalbucore 0.0.19 requires numpy>=1.24.4, but you have numpy 1.24.3 which is incompatible.\nalbumentations 1.4.20 requires numpy>=1.24.4, but you have numpy 1.24.3 which is incompatible.\nbayesian-optimization 2.0.3 requires numpy>=1.25, but you have numpy 1.24.3 which is incompatible.\nfeaturetools 1.31.0 requires numpy>=1.25.0, but you have numpy 1.24.3 which is incompatible.\ngoogle-colab 1.0.0 requires notebook==6.5.5, but you have notebook 6.5.4 which is incompatible.\ngoogle-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.\nkaggle-environments 1.16.11 requires scipy>=1.11.2, but you have scipy 1.10.1 which is incompatible.\nlangchain 0.3.12 requires async-timeout<5.0.0,>=4.0.0; python_version < \"3.11\", but you have async-timeout 5.0.1 which is incompatible.\nmkl-fft 1.3.8 requires numpy<1.27.0,>=1.26.4, but you have numpy 1.24.3 which is incompatible.\nmkl-random 1.2.4 requires numpy<1.27.0,>=1.26.4, but you have numpy 1.24.3 which is incompatible.\nmkl-umath 0.1.1 requires numpy<1.27.0,>=1.26.4, but you have numpy 1.24.3 which is incompatible.\nmlxtend 0.23.3 requires scikit-learn>=1.3.1, but you have scikit-learn 1.2.2 which is incompatible.\npandas-gbq 0.25.0 requires google-api-core<3.0.0dev,>=2.10.2, but you have google-api-core 1.34.1 which is incompatible.\npymc 5.19.1 requires numpy>=1.25.0, but you have numpy 1.24.3 which is incompatible.\npytorch-lightning 2.5.0.post0 requires torch>=2.1.0, but you have torch 2.0.1+cu118 which is incompatible.\nrna-predict 0.1.3 requires datasets>=3.3.2, but you have datasets 3.3.1 which is incompatible.\nrna-predict 0.1.3 requires numpy>=2.2.3, but you have numpy 1.24.3 which is incompatible.\nscikit-image 0.25.0 requires scipy>=1.11.2, but you have scipy 1.10.1 which is incompatible.\ntensorflow-decision-forests 1.10.0 requires tensorflow==2.17.0, but you have tensorflow 2.17.1 which is incompatible.\nwoodwork 0.31.0 requires numpy>=1.25.0, but you have numpy 1.24.3 which is incompatible.\n\n\n=== [Final Package Versions] ===\nName: torch\nVersion: 2.0.1+cu118\nSummary: Tensors and Dynamic neural networks in Python with strong GPU acceleration\nHome-page: https://pytorch.org/\nAuthor: PyTorch Team\nAuthor-email: packages@pytorch.org\nLicense: BSD-3\nLocation: /usr/local/lib/python3.10/dist-packages\nRequires: filelock, jinja2, networkx, sympy, triton, typing-extensions\nRequired-by: accelerate, block-sparse-attn, easyocr, fastai, kornia, peft, pytorch-ignite, pytorch-lightning, rna-predict, sentence-transformers, stable-baselines3, timm, torchmetrics, triton\n\nName: scikit-learn\nVersion: 1.2.2\nSummary: A set of python modules for machine learning and data mining\nHome-page: http://scikit-learn.org\nAuthor: \nAuthor-email: \nLicense: new BSD\nLocation: /usr/local/lib/python3.10/dist-packages\nRequires: joblib, numpy, scipy, threadpoolctl\nRequired-by: bayesian-optimization, bigframes, Boruta, category_encoders, cesium, eli5, fastai, hep_ml, imbalanced-learn, librosa, lime, mlxtend, nilearn, pyLDAvis, rgf-python, scikit-learn-intelex, scikit-optimize, scikit-plot, sentence-transformers, shap, sklearn-pandas, TPOT, tsfresh, woodwork, yellowbrick\n\nName: scipy\nVersion: 1.10.1\nSummary: Fundamental algorithms for scientific computing in Python\nHome-page: https://scipy.org/\nAuthor: \nAuthor-email: \nLicense: Copyright (c) 2001-2002 Enthought, Inc. 2003-2022, SciPy Developers.\n        All rights reserved.\n        \n        Redistribution and use in source and binary forms, with or without\n        modification, are permitted provided that the following conditions\n        are met:\n        \n        1. Redistributions of source code must retain the above copyright\n           notice, this list of conditions and the following disclaimer.\n        \n        2. Redistributions in binary form must reproduce the above\n           copyright notice, this list of conditions and the following\n           disclaimer in the documentation and/or other materials provided\n           with the distribution.\n        \n        3. Neither the name of the copyright holder nor the names of its\n           contributors may be used to endorse or promote products derived\n           from this software without specific prior written permission.\n        \n        THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n        \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n        LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR\n        A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT\n        OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,\n        SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT\n        LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,\n        DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY\n        THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT\n        (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\n        OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n        \n        ----\n        \n        This binary distribution of SciPy also bundles the following software:\n        \n        \n        Name: OpenBLAS\n        Files: .libs/libopenb*.so\n        Description: bundled as a dynamically linked library\n        Availability: https://github.com/xianyi/OpenBLAS/\n        License: 3-clause BSD\n          Copyright (c) 2011-2014, The OpenBLAS Project\n          All rights reserved.\n          \n          Redistribution and use in source and binary forms, with or without\n          modification, are permitted provided that the following conditions are\n          met:\n          \n             1. Redistributions of source code must retain the above copyright\n                notice, this list of conditions and the following disclaimer.\n          \n             2. Redistributions in binary form must reproduce the above copyright\n                notice, this list of conditions and the following disclaimer in\n                the documentation and/or other materials provided with the\n                distribution.\n             3. Neither the name of the OpenBLAS project nor the names of \n                its contributors may be used to endorse or promote products \n                derived from this software without specific prior written \n                permission.\n          \n          THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\"\n          AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE\n          IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE\n          ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE\n          LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL\n          DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR\n          SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER\n          CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,\n          OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE\n          USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n        \n        \n        Name: LAPACK\n        Files: .libs/libopenb*.so\n        Description: bundled in OpenBLAS\n        Availability: https://github.com/xianyi/OpenBLAS/\n        License 3-clause BSD\n          Copyright (c) 1992-2013 The University of Tennessee and The University\n                                  of Tennessee Research Foundation.  All rights\n                                  reserved.\n          Copyright (c) 2000-2013 The University of California Berkeley. All\n                                  rights reserved.\n          Copyright (c) 2006-2013 The University of Colorado Denver.  All rights\n                                  reserved.\n          \n          $COPYRIGHT$\n          \n          Additional copyrights may follow\n          \n          $HEADER$\n          \n          Redistribution and use in source and binary forms, with or without\n          modification, are permitted provided that the following conditions are\n          met:\n          \n          - Redistributions of source code must retain the above copyright\n            notice, this list of conditions and the following disclaimer.\n          \n          - Redistributions in binary form must reproduce the above copyright\n            notice, this list of conditions and the following disclaimer listed\n            in this license in the documentation and/or other materials\n            provided with the distribution.\n          \n          - Neither the name of the copyright holders nor the names of its\n            contributors may be used to endorse or promote products derived from\n            this software without specific prior written permission.\n          \n          The copyright holders provide no reassurances that the source code\n          provided does not infringe any patent, copyright, or any other\n          intellectual property rights of third parties.  The copyright holders\n          disclaim any liability to any recipient for claims brought against\n          recipient by any third party for infringement of that parties\n          intellectual property rights.\n          \n          THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS\n          \"AS IS\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT\n          LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR\n          A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT\n          OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,\n          SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT\n          LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,\n          DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY\n          THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT\n          (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE\n          OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n        \n        \n        Name: GCC runtime library\n        Files: .libs/libgfortran*.so\n        Description: dynamically linked to files compiled with gcc\n        Availability: https://gcc.gnu.org/viewcvs/gcc/\n        License: GPLv3 + runtime exception\n          Copyright (C) 2002-2017 Free Software Foundation, Inc.\n          \n          Libgfortran is free software; you can redistribute it and/or modify\n          it under the terms of the GNU General Public License as published by\n          the Free Software Foundation; either version 3, or (at your option)\n          any later version.\n          \n          Libgfortran is distributed in the hope that it will be useful,\n          but WITHOUT ANY WARRANTY; without even the implied warranty of\n          MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n          GNU General Public License for more details.\n          \n          Under Section 7 of GPL version 3, you are granted additional\n          permissions described in the GCC Runtime Library Exception, version\n          3.1, as published by the Free Software Foundation.\n          \n          You should have received a copy of the GNU General Public License and\n          a copy of the GCC Runtime Library Exception along with this program;\n          see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see\n          <http://www.gnu.org/licenses/>.\n        \n        ----\n        \n        Full text of license texts referred to above follows (that they are\n        listed below does not necessarily imply the conditions apply to the\n        present binary release):\n        \n        ----\n        \n        GCC RUNTIME LIBRARY EXCEPTION\n        \n        Version 3.1, 31 March 2009\n        \n        Copyright (C) 2009 Free Software Foundation, Inc. <http://fsf.org/>\n        \n        Everyone is permitted to copy and distribute verbatim copies of this\n        license document, but changing it is not allowed.\n        \n        This GCC Runtime Library Exception (\"Exception\") is an additional\n        permission under section 7 of the GNU General Public License, version\n        3 (\"GPLv3\"). It applies to a given file (the \"Runtime Library\") that\n        bears a notice placed by the copyright holder of the file stating that\n        the file is governed by GPLv3 along with this Exception.\n        \n        When you use GCC to compile a program, GCC may combine portions of\n        certain GCC header files and runtime libraries with the compiled\n        program. The purpose of this Exception is to allow compilation of\n        non-GPL (including proprietary) programs to use, in this way, the\n        header files and runtime libraries covered by this Exception.\n        \n        0. Definitions.\n        \n        A file is an \"Independent Module\" if it either requires the Runtime\n        Library for execution after a Compilation Process, or makes use of an\n        interface provided by the Runtime Library, but is not otherwise based\n        on the Runtime Library.\n        \n        \"GCC\" means a version of the GNU Compiler Collection, with or without\n        modifications, governed by version 3 (or a specified later version) of\n        the GNU General Public License (GPL) with the option of using any\n        subsequent versions published by the FSF.\n        \n        \"GPL-compatible Software\" is software whose conditions of propagation,\n        modification and use would permit combination with GCC in accord with\n        the license of GCC.\n        \n        \"Target Code\" refers to output from any compiler for a real or virtual\n        target processor architecture, in executable form or suitable for\n        input to an assembler, loader, linker and/or execution\n        phase. Notwithstanding that, Target Code does not include data in any\n        format that is used as a compiler intermediate representation, or used\n        for producing a compiler intermediate representation.\n        \n        The \"Compilation Process\" transforms code entirely represented in\n        non-intermediate languages designed for human-written code, and/or in\n        Java Virtual Machine byte code, into Target Code. Thus, for example,\n        use of source code generators and preprocessors need not be considered\n        part of the Compilation Process, since the Compilation Process can be\n        understood as starting with the output of the generators or\n        preprocessors.\n        \n        A Compilation Process is \"Eligible\" if it is done using GCC, alone or\n        with other GPL-compatible software, or if it is done without using any\n        work based on GCC. For example, using non-GPL-compatible Software to\n        optimize any GCC intermediate representations would not qualify as an\n        Eligible Compilation Process.\n        \n        1. Grant of Additional Permission.\n        \n        You have permission to propagate a work of Target Code formed by\n        combining the Runtime Library with Independent Modules, even if such\n        propagation would otherwise violate the terms of GPLv3, provided that\n        all Target Code was generated by Eligible Compilation Processes. You\n        may then convey such a combination under terms of your choice,\n        consistent with the licensing of the Independent Modules.\n        \n        2. No Weakening of GCC Copyleft.\n        \n        The availability of this Exception does not imply any general\n        presumption that third-party software is unaffected by the copyleft\n        requirements of the license of GCC.\n        \n        ----\n        \n                            GNU GENERAL PUBLIC LICENSE\n                               Version 3, 29 June 2007\n        \n         Copyright (C) 2007 Free Software Foundation, Inc. <http://fsf.org/>\n         Everyone is permitted to copy and distribute verbatim copies\n         of this license document, but changing it is not allowed.\n        \n                                    Preamble\n        \n          The GNU General Public License is a free, copyleft license for\n        software and other kinds of works.\n        \n          The licenses for most software and other practical works are designed\n        to take away your freedom to share and change the works.  By contrast,\n        the GNU General Public License is intended to guarantee your freedom to\n        share and change all versions of a program--to make sure it remains free\n        software for all its users.  We, the Free Software Foundation, use the\n        GNU General Public License for most of our software; it applies also to\n        any other work released this way by its authors.  You can apply it to\n        your programs, too.\n        \n          When we speak of free software, we are referring to freedom, not\n        price.  Our General Public Licenses are designed to make sure that you\n        have the freedom to distribute copies of free software (and charge for\n        them if you wish), that you receive source code or can get it if you\n        want it, that you can change the software or use pieces of it in new\n        free programs, and that you know you can do these things.\n        \n          To protect your rights, we need to prevent others from denying you\n        these rights or asking you to surrender the rights.  Therefore, you have\n        certain responsibilities if you distribute copies of the software, or if\n        you modify it: responsibilities to respect the freedom of others.\n        \n          For example, if you distribute copies of such a program, whether\n        gratis or for a fee, you must pass on to the recipients the same\n        freedoms that you received.  You must make sure that they, too, receive\n        or can get the source code.  And you must show them these terms so they\n        know their rights.\n        \n          Developers that use the GNU GPL protect your rights with two steps:\n        (1) assert copyright on the software, and (2) offer you this License\n        giving you legal permission to copy, distribute and/or modify it.\n        \n          For the developers' and authors' protection, the GPL clearly explains\n        that there is no warranty for this free software.  For both users' and\n        authors' sake, the GPL requires that modified versions be marked as\n        changed, so that their problems will not be attributed erroneously to\n        authors of previous versions.\n        \n          Some devices are designed to deny users access to install or run\n        modified versions of the software inside them, although the manufacturer\n        can do so.  This is fundamentally incompatible with the aim of\n        protecting users' freedom to change the software.  The systematic\n        pattern of such abuse occurs in the area of products for individuals to\n        use, which is precisely where it is most unacceptable.  Therefore, we\n        have designed this version of the GPL to prohibit the practice for those\n        products.  If such problems arise substantially in other domains, we\n        stand ready to extend this provision to those domains in future versions\n        of the GPL, as needed to protect the freedom of users.\n        \n          Finally, every program is threatened constantly by software patents.\n        States should not allow patents to restrict development and use of\n        software on general-purpose computers, but in those that do, we wish to\n        avoid the special danger that patents applied to a free program could\n        make it effectively proprietary.  To prevent this, the GPL assures that\n        patents cannot be used to render the program non-free.\n        \n          The precise terms and conditions for copying, distribution and\n        modification follow.\n        \n                               TERMS AND CONDITIONS\n        \n          0. Definitions.\n        \n          \"This License\" refers to version 3 of the GNU General Public License.\n        \n          \"Copyright\" also means copyright-like laws that apply to other kinds of\n        works, such as semiconductor masks.\n        \n          \"The Program\" refers to any copyrightable work licensed under this\n        License.  Each licensee is addressed as \"you\".  \"Licensees\" and\n        \"recipients\" may be individuals or organizations.\n        \n          To \"modify\" a work means to copy from or adapt all or part of the work\n        in a fashion requiring copyright permission, other than the making of an\n        exact copy.  The resulting work is called a \"modified version\" of the\n        earlier work or a work \"based on\" the earlier work.\n        \n          A \"covered work\" means either the unmodified Program or a work based\n        on the Program.\n        \n          To \"propagate\" a work means to do anything with it that, without\n        permission, would make you directly or secondarily liable for\n        infringement under applicable copyright law, except executing it on a\n        computer or modifying a private copy.  Propagation includes copying,\n        distribution (with or without modification), making available to the\n        public, and in some countries other activities as well.\n        \n          To \"convey\" a work means any kind of propagation that enables other\n        parties to make or receive copies.  Mere interaction with a user through\n        a computer network, with no transfer of a copy, is not conveying.\n        \n          An interactive user interface displays \"Appropriate Legal Notices\"\n        to the extent that it includes a convenient and prominently visible\n        feature that (1) displays an appropriate copyright notice, and (2)\n        tells the user that there is no warranty for the work (except to the\n        extent that warranties are provided), that licensees may convey the\n        work under this License, and how to view a copy of this License.  If\n        the interface presents a list of user commands or options, such as a\n        menu, a prominent item in the list meets this criterion.\n        \n          1. Source Code.\n        \n          The \"source code\" for a work means the preferred form of the work\n        for making modifications to it.  \"Object code\" means any non-source\n        form of a work.\n        \n          A \"Standard Interface\" means an interface that either is an official\n        standard defined by a recognized standards body, or, in the case of\n        interfaces specified for a particular programming language, one that\n        is widely used among developers working in that language.\n        \n          The \"System Libraries\" of an executable work include anything, other\n        than the work as a whole, that (a) is included in the normal form of\n        packaging a Major Component, but which is not part of that Major\n        Component, and (b) serves only to enable use of the work with that\n        Major Component, or to implement a Standard Interface for which an\n        implementation is available to the public in source code form.  A\n        \"Major Component\", in this context, means a major essential component\n        (kernel, window system, and so on) of the specific operating system\n        (if any) on which the executable work runs, or a compiler used to\n        produce the work, or an object code interpreter used to run it.\n        \n          The \"Corresponding Source\" for a work in object code form means all\n        the source code needed to generate, install, and (for an executable\n        work) run the object code and to modify the work, including scripts to\n        control those activities.  However, it does not include the work's\n        System Libraries, or general-purpose tools or generally available free\n        programs which are used unmodified in performing those activities but\n        which are not part of the work.  For example, Corresponding Source\n        includes interface definition files associated with source files for\n        the work, and the source code for shared libraries and dynamically\n        linked subprograms that the work is specifically designed to require,\n        such as by intimate data communication or control flow between those\n        subprograms and other parts of the work.\n        \n          The Corresponding Source need not include anything that users\n        can regenerate automatically from other parts of the Corresponding\n        Source.\n        \n          The Corresponding Source for a work in source code form is that\n        same work.\n        \n          2. Basic Permissions.\n        \n          All rights granted under this License are granted for the term of\n        copyright on the Program, and are irrevocable provided the stated\n        conditions are met.  This License explicitly affirms your unlimited\n        permission to run the unmodified Program.  The output from running a\n        covered work is covered by this License only if the output, given its\n        content, constitutes a covered work.  This License acknowledges your\n        rights of fair use or other equivalent, as provided by copyright law.\n        \n          You may make, run and propagate covered works that you do not\n        convey, without conditions so long as your license otherwise remains\n        in force.  You may convey covered works to others for the sole purpose\n        of having them make modifications exclusively for you, or provide you\n        with facilities for running those works, provided that you comply with\n        the terms of this License in conveying all material for which you do\n        not control copyright.  Those thus making or running the covered works\n        for you must do so exclusively on your behalf, under your direction\n        and control, on terms that prohibit them from making any copies of\n        your copyrighted material outside their relationship with you.\n        \n          Conveying under any other circumstances is permitted solely under\n        the conditions stated below.  Sublicensing is not allowed; section 10\n        makes it unnecessary.\n        \n          3. Protecting Users' Legal Rights From Anti-Circumvention Law.\n        \n          No covered work shall be deemed part of an effective technological\n        measure under any applicable law fulfilling obligations under article\n        11 of the WIPO copyright treaty adopted on 20 December 1996, or\n        similar laws prohibiting or restricting circumvention of such\n        measures.\n        \n          When you convey a covered work, you waive any legal power to forbid\n        circumvention of technological measures to the extent such circumvention\n        is effected by exercising rights under this License with respect to\n        the covered work, and you disclaim any intention to limit operation or\n        modification of the work as a means of enforcing, against the work's\n        users, your or third parties' legal rights to forbid circumvention of\n        technological measures.\n        \n          4. Conveying Verbatim Copies.\n        \n          You may convey verbatim copies of the Program's source code as you\n        receive it, in any medium, provided that you conspicuously and\n        appropriately publish on each copy an appropriate copyright notice;\n        keep intact all notices stating that this License and any\n        non-permissive terms added in accord with section 7 apply to the code;\n        keep intact all notices of the absence of any warranty; and give all\n        recipients a copy of this License along with the Program.\n        \n          You may charge any price or no price for each copy that you convey,\n        and you may offer support or warranty protection for a fee.\n        \n          5. Conveying Modified Source Versions.\n        \n          You may convey a work based on the Program, or the modifications to\n        produce it from the Program, in the form of source code under the\n        terms of section 4, provided that you also meet all of these conditions:\n        \n            a) The work must carry prominent notices stating that you modified\n            it, and giving a relevant date.\n        \n            b) The work must carry prominent notices stating that it is\n            released under this License and any conditions added under section\n            7.  This requirement modifies the requirement in section 4 to\n            \"keep intact all notices\".\n        \n            c) You must license the entire work, as a whole, under this\n            License to anyone who comes into possession of a copy.  This\n            License will therefore apply, along with any applicable section 7\n            additional terms, to the whole of the work, and all its parts,\n            regardless of how they are packaged.  This License gives no\n            permission to license the work in any other way, but it does not\n            invalidate such permission if you have separately received it.\n        \n            d) If the work has interactive user interfaces, each must display\n            Appropriate Legal Notices; however, if the Program has interactive\n            interfaces that do not display Appropriate Legal Notices, your\n            work need not make them do so.\n        \n          A compilation of a covered work with other separate and independent\n        works, which are not by their nature extensions of the covered work,\n        and which are not combined with it such as to form a larger program,\n        in or on a volume of a storage or distribution medium, is called an\n        \"aggregate\" if the compilation and its resulting copyright are not\n        used to limit the access or legal rights of the compilation's users\n        beyond what the individual works permit.  Inclusion of a covered work\n        in an aggregate does not cause this License to apply to the other\n        parts of the aggregate.\n        \n          6. Conveying Non-Source Forms.\n        \n          You may convey a covered work in object code form under the terms\n        of sections 4 and 5, provided that you also convey the\n        machine-readable Corresponding Source under the terms of this License,\n        in one of these ways:\n        \n            a) Convey the object code in, or embodied in, a physical product\n            (including a physical distribution medium), accompanied by the\n            Corresponding Source fixed on a durable physical medium\n            customarily used for software interchange.\n        \n            b) Convey the object code in, or embodied in, a physical product\n            (including a physical distribution medium), accompanied by a\n            written offer, valid for at least three years and valid for as\n            long as you offer spare parts or customer support for that product\n            model, to give anyone who possesses the object code either (1) a\n            copy of the Corresponding Source for all the software in the\n            product that is covered by this License, on a durable physical\n            medium customarily used for software interchange, for a price no\n            more than your reasonable cost of physically performing this\n            conveying of source, or (2) access to copy the\n            Corresponding Source from a network server at no charge.\n        \n            c) Convey individual copies of the object code with a copy of the\n            written offer to provide the Corresponding Source.  This\n            alternative is allowed only occasionally and noncommercially, and\n            only if you received the object code with such an offer, in accord\n            with subsection 6b.\n        \n            d) Convey the object code by offering access from a designated\n            place (gratis or for a charge), and offer equivalent access to the\n            Corresponding Source in the same way through the same place at no\n            further charge.  You need not require recipients to copy the\n            Corresponding Source along with the object code.  If the place to\n            copy the object code is a network server, the Corresponding Source\n            may be on a different server (operated by you or a third party)\n            that supports equivalent copying facilities, provided you maintain\n            clear directions next to the object code saying where to find the\n            Corresponding Source.  Regardless of what server hosts the\n            Corresponding Source, you remain obligated to ensure that it is\n            available for as long as needed to satisfy these requirements.\n        \n            e) Convey the object code using peer-to-peer transmission, provided\n            you inform other peers where the object code and Corresponding\n            Source of the work are being offered to the general public at no\n            charge under subsection 6d.\n        \n          A separable portion of the object code, whose source code is excluded\n        from the Corresponding Source as a System Library, need not be\n        included in conveying the object code work.\n        \n          A \"User Product\" is either (1) a \"consumer product\", which means any\n        tangible personal property which is normally used for personal, family,\n        or household purposes, or (2) anything designed or sold for incorporation\n        into a dwelling.  In determining whether a product is a consumer product,\n        doubtful cases shall be resolved in favor of coverage.  For a particular\n        product received by a particular user, \"normally used\" refers to a\n        typical or common use of that class of product, regardless of the status\n        of the particular user or of the way in which the particular user\n        actually uses, or expects or is expected to use, the product.  A product\n        is a consumer product regardless of whether the product has substantial\n        commercial, industrial or non-consumer uses, unless such uses represent\n        the only significant mode of use of the product.\n        \n          \"Installation Information\" for a User Product means any methods,\n        procedures, authorization keys, or other information required to install\n        and execute modified versions of a covered work in that User Product from\n        a modified version of its Corresponding Source.  The information must\n        suffice to ensure that the continued functioning of the modified object\n        code is in no case prevented or interfered with solely because\n        modification has been made.\n        \n          If you convey an object code work under this section in, or with, or\n        specifically for use in, a User Product, and the conveying occurs as\n        part of a transaction in which the right of possession and use of the\n        User Product is transferred to the recipient in perpetuity or for a\n        fixed term (regardless of how the transaction is characterized), the\n        Corresponding Source conveyed under this section must be accompanied\n        by the Installation Information.  But this requirement does not apply\n        if neither you nor any third party retains the ability to install\n        modified object code on the User Product (for example, the work has\n        been installed in ROM).\n        \n          The requirement to provide Installation Information does not include a\n        requirement to continue to provide support service, warranty, or updates\n        for a work that has been modified or installed by the recipient, or for\n        the User Product in which it has been modified or installed.  Access to a\n        network may be denied when the modification itself materially and\n        adversely affects the operation of the network or violates the rules and\n        protocols for communication across the network.\n        \n          Corresponding Source conveyed, and Installation Information provided,\n        in accord with this section must be in a format that is publicly\n        documented (and with an implementation available to the public in\n        source code form), and must require no special password or key for\n        unpacking, reading or copying.\n        \n          7. Additional Terms.\n        \n          \"Additional permissions\" are terms that supplement the terms of this\n        License by making exceptions from one or more of its conditions.\n        Additional permissions that are applicable to the entire Program shall\n        be treated as though they were included in this License, to the extent\n        that they are valid under applicable law.  If additional permissions\n        apply only to part of the Program, that part may be used separately\n        under those permissions, but the entire Program remains governed by\n        this License without regard to the additional permissions.\n        \n          When you convey a copy of a covered work, you may at your option\n        remove any additional permissions from that copy, or from any part of\n        it.  (Additional permissions may be written to require their own\n        removal in certain cases when you modify the work.)  You may place\n        additional permissions on material, added by you to a covered work,\n        for which you have or can give appropriate copyright permission.\n        \n          Notwithstanding any other provision of this License, for material you\n        add to a covered work, you may (if authorized by the copyright holders of\n        that material) supplement the terms of this License with terms:\n        \n            a) Disclaiming warranty or limiting liability differently from the\n            terms of sections 15 and 16 of this License; or\n        \n            b) Requiring preservation of specified reasonable legal notices or\n            author attributions in that material or in the Appropriate Legal\n            Notices displayed by works containing it; or\n        \n            c) Prohibiting misrepresentation of the origin of that material, or\n            requiring that modified versions of such material be marked in\n            reasonable ways as different from the original version; or\n        \n            d) Limiting the use for publicity purposes of names of licensors or\n            authors of the material; or\n        \n            e) Declining to grant rights under trademark law for use of some\n            trade names, trademarks, or service marks; or\n        \n            f) Requiring indemnification of licensors and authors of that\n            material by anyone who conveys the material (or modified versions of\n            it) with contractual assumptions of liability to the recipient, for\n            any liability that these contractual assumptions directly impose on\n            those licensors and authors.\n        \n          All other non-permissive additional terms are considered \"further\n        restrictions\" within the meaning of section 10.  If the Program as you\n        received it, or any part of it, contains a notice stating that it is\n        governed by this License along with a term that is a further\n        restriction, you may remove that term.  If a license document contains\n        a further restriction but permits relicensing or conveying under this\n        License, you may add to a covered work material governed by the terms\n        of that license document, provided that the further restriction does\n        not survive such relicensing or conveying.\n        \n          If you add terms to a covered work in accord with this section, you\n        must place, in the relevant source files, a statement of the\n        additional terms that apply to those files, or a notice indicating\n        where to find the applicable terms.\n        \n          Additional terms, permissive or non-permissive, may be stated in the\n        form of a separately written license, or stated as exceptions;\n        the above requirements apply either way.\n        \n          8. Termination.\n        \n          You may not propagate or modify a covered work except as expressly\n        provided under this License.  Any attempt otherwise to propagate or\n        modify it is void, and will automatically terminate your rights under\n        this License (including any patent licenses granted under the third\n        paragraph of section 11).\n        \n          However, if you cease all violation of this License, then your\n        license from a particular copyright holder is reinstated (a)\n        provisionally, unless and until the copyright holder explicitly and\n        finally terminates your license, and (b) permanently, if the copyright\n        holder fails to notify you of the violation by some reasonable means\n        prior to 60 days after the cessation.\n        \n          Moreover, your license from a particular copyright holder is\n        reinstated permanently if the copyright holder notifies you of the\n        violation by some reasonable means, this is the first time you have\n        received notice of violation of this License (for any work) from that\n        copyright holder, and you cure the violation prior to 30 days after\n        your receipt of the notice.\n        \n          Termination of your rights under this section does not terminate the\n        licenses of parties who have received copies or rights from you under\n        this License.  If your rights have been terminated and not permanently\n        reinstated, you do not qualify to receive new licenses for the same\n        material under section 10.\n        \n          9. Acceptance Not Required for Having Copies.\n        \n          You are not required to accept this License in order to receive or\n        run a copy of the Program.  Ancillary propagation of a covered work\n        occurring solely as a consequence of using peer-to-peer transmission\n        to receive a copy likewise does not require acceptance.  However,\n        nothing other than this License grants you permission to propagate or\n        modify any covered work.  These actions infringe copyright if you do\n        not accept this License.  Therefore, by modifying or propagating a\n        covered work, you indicate your acceptance of this License to do so.\n        \n          10. Automatic Licensing of Downstream Recipients.\n        \n          Each time you convey a covered work, the recipient automatically\n        receives a license from the original licensors, to run, modify and\n        propagate that work, subject to this License.  You are not responsible\n        for enforcing compliance by third parties with this License.\n        \n          An \"entity transaction\" is a transaction transferring control of an\n        organization, or substantially all assets of one, or subdividing an\n        organization, or merging organizations.  If propagation of a covered\n        work results from an entity transaction, each party to that\n        transaction who receives a copy of the work also receives whatever\n        licenses to the work the party's predecessor in interest had or could\n        give under the previous paragraph, plus a right to possession of the\n        Corresponding Source of the work from the predecessor in interest, if\n        the predecessor has it or can get it with reasonable efforts.\n        \n          You may not impose any further restrictions on the exercise of the\n        rights granted or affirmed under this License.  For example, you may\n        not impose a license fee, royalty, or other charge for exercise of\n        rights granted under this License, and you may not initiate litigation\n        (including a cross-claim or counterclaim in a lawsuit) alleging that\n        any patent claim is infringed by making, using, selling, offering for\n        sale, or importing the Program or any portion of it.\n        \n          11. Patents.\n        \n          A \"contributor\" is a copyright holder who authorizes use under this\n        License of the Program or a work on which the Program is based.  The\n        work thus licensed is called the contributor's \"contributor version\".\n        \n          A contributor's \"essential patent claims\" are all patent claims\n        owned or controlled by the contributor, whether already acquired or\n        hereafter acquired, that would be infringed by some manner, permitted\n        by this License, of making, using, or selling its contributor version,\n        but do not include claims that would be infringed only as a\n        consequence of further modification of the contributor version.  For\n        purposes of this definition, \"control\" includes the right to grant\n        patent sublicenses in a manner consistent with the requirements of\n        this License.\n        \n          Each contributor grants you a non-exclusive, worldwide, royalty-free\n        patent license under the contributor's essential patent claims, to\n        make, use, sell, offer for sale, import and otherwise run, modify and\n        propagate the contents of its contributor version.\n        \n          In the following three paragraphs, a \"patent license\" is any express\n        agreement or commitment, however denominated, not to enforce a patent\n        (such as an express permission to practice a patent or covenant not to\n        sue for patent infringement).  To \"grant\" such a patent license to a\n        party means to make such an agreement or commitment not to enforce a\n        patent against the party.\n        \n          If you convey a covered work, knowingly relying on a patent license,\n        and the Corresponding Source of the work is not available for anyone\n        to copy, free of charge and under the terms of this License, through a\n        publicly available network server or other readily accessible means,\n        then you must either (1) cause the Corresponding Source to be so\n        available, or (2) arrange to deprive yourself of the benefit of the\n        patent license for this particular work, or (3) arrange, in a manner\n        consistent with the requirements of this License, to extend the patent\n        license to downstream recipients.  \"Knowingly relying\" means you have\n        actual knowledge that, but for the patent license, your conveying the\n        covered work in a country, or your recipient's use of the covered work\n        in a country, would infringe one or more identifiable patents in that\n        country that you have reason to believe are valid.\n        \n          If, pursuant to or in connection with a single transaction or\n        arrangement, you convey, or propagate by procuring conveyance of, a\n        covered work, and grant a patent license to some of the parties\n        receiving the covered work authorizing them to use, propagate, modify\n        or convey a specific copy of the covered work, then the patent license\n        you grant is automatically extended to all recipients of the covered\n        work and works based on it.\n        \n          A patent license is \"discriminatory\" if it does not include within\n        the scope of its coverage, prohibits the exercise of, or is\n        conditioned on the non-exercise of one or more of the rights that are\n        specifically granted under this License.  You may not convey a covered\n        work if you are a party to an arrangement with a third party that is\n        in the business of distributing software, under which you make payment\n        to the third party based on the extent of your activity of conveying\n        the work, and under which the third party grants, to any of the\n        parties who would receive the covered work from you, a discriminatory\n        patent license (a) in connection with copies of the covered work\n        conveyed by you (or copies made from those copies), or (b) primarily\n        for and in connection with specific products or compilations that\n        contain the covered work, unless you entered into that arrangement,\n        or that patent license was granted, prior to 28 March 2007.\n        \n          Nothing in this License shall be construed as excluding or limiting\n        any implied license or other defenses to infringement that may\n        otherwise be available to you under applicable patent law.\n        \n          12. No Surrender of Others' Freedom.\n        \n          If conditions are imposed on you (whether by court order, agreement or\n        otherwise) that contradict the conditions of this License, they do not\n        excuse you from the conditions of this License.  If you cannot convey a\n        covered work so as to satisfy simultaneously your obligations under this\n        License and any other pertinent obligations, then as a consequence you may\n        not convey it at all.  For example, if you agree to terms that obligate you\n        to collect a royalty for further conveying from those to whom you convey\n        the Program, the only way you could satisfy both those terms and this\n        License would be to refrain entirely from conveying the Program.\n        \n          13. Use with the GNU Affero General Public License.\n        \n          Notwithstanding any other provision of this License, you have\n        permission to link or combine any covered work with a work licensed\n        under version 3 of the GNU Affero General Public License into a single\n        combined work, and to convey the resulting work.  The terms of this\n        License will continue to apply to the part which is the covered work,\n        but the special requirements of the GNU Affero General Public License,\n        section 13, concerning interaction through a network will apply to the\n        combination as such.\n        \n          14. Revised Versions of this License.\n        \n          The Free Software Foundation may publish revised and/or new versions of\n        the GNU General Public License from time to time.  Such new versions will\n        be similar in spirit to the present version, but may differ in detail to\n        address new problems or concerns.\n        \n          Each version is given a distinguishing version number.  If the\n        Program specifies that a certain numbered version of the GNU General\n        Public License \"or any later version\" applies to it, you have the\n        option of following the terms and conditions either of that numbered\n        version or of any later version published by the Free Software\n        Foundation.  If the Program does not specify a version number of the\n        GNU General Public License, you may choose any version ever published\n        by the Free Software Foundation.\n        \n          If the Program specifies that a proxy can decide which future\n        versions of the GNU General Public License can be used, that proxy's\n        public statement of acceptance of a version permanently authorizes you\n        to choose that version for the Program.\n        \n          Later license versions may give you additional or different\n        permissions.  However, no additional obligations are imposed on any\n        author or copyright holder as a result of your choosing to follow a\n        later version.\n        \n          15. Disclaimer of Warranty.\n        \n          THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY\n        APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT\n        HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM \"AS IS\" WITHOUT WARRANTY\n        OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO,\n        THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR\n        PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM\n        IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF\n        ALL NECESSARY SERVICING, REPAIR OR CORRECTION.\n        \n          16. Limitation of Liability.\n        \n          IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING\n        WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS\n        THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY\n        GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE\n        USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF\n        DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD\n        PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS),\n        EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF\n        SUCH DAMAGES.\n        \n          17. Interpretation of Sections 15 and 16.\n        \n          If the disclaimer of warranty and limitation of liability provided\n        above cannot be given local legal effect according to their terms,\n        reviewing courts shall apply local law that most closely approximates\n        an absolute waiver of all civil liability in connection with the\n        Program, unless a warranty or assumption of liability accompanies a\n        copy of the Program in return for a fee.\n        \n                             END OF TERMS AND CONDITIONS\n        \n                    How to Apply These Terms to Your New Programs\n        \n          If you develop a new program, and you want it to be of the greatest\n        possible use to the public, the best way to achieve this is to make it\n        free software which everyone can redistribute and change under these terms.\n        \n          To do so, attach the following notices to the program.  It is safest\n        to attach them to the start of each source file to most effectively\n        state the exclusion of warranty; and each file should have at least\n        the \"copyright\" line and a pointer to where the full notice is found.\n        \n            <one line to give the program's name and a brief idea of what it does.>\n            Copyright (C) <year>  <name of author>\n        \n            This program is free software: you can redistribute it and/or modify\n            it under the terms of the GNU General Public License as published by\n            the Free Software Foundation, either version 3 of the License, or\n            (at your option) any later version.\n        \n            This program is distributed in the hope that it will be useful,\n            but WITHOUT ANY WARRANTY; without even the implied warranty of\n            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n            GNU General Public License for more details.\n        \n            You should have received a copy of the GNU General Public License\n            along with this program.  If not, see <http://www.gnu.org/licenses/>.\n        \n        Also add information on how to contact you by electronic and paper mail.\n        \n          If the program does terminal interaction, make it output a short\n        notice like this when it starts in an interactive mode:\n        \n            <program>  Copyright (C) <year>  <name of author>\n            This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.\n            This is free software, and you are welcome to redistribute it\n            under certain conditions; type `show c' for details.\n        \n        The hypothetical commands `show w' and `show c' should show the appropriate\n        parts of the General Public License.  Of course, your program's commands\n        might be different; for a GUI interface, you would use an \"about box\".\n        \n          You should also get your employer (if you work as a programmer) or school,\n        if any, to sign a \"copyright disclaimer\" for the program, if necessary.\n        For more information on this, and how to apply and follow the GNU GPL, see\n        <http://www.gnu.org/licenses/>.\n        \n          The GNU General Public License does not permit incorporating your program\n        into proprietary programs.  If your program is a subroutine library, you\n        may consider it more useful to permit linking proprietary applications with\n        the library.  If this is what you want to do, use the GNU Lesser General\n        Public License instead of this License.  But first, please read\n        <http://www.gnu.org/philosophy/why-not-lgpl.html>.\nLocation: /usr/local/lib/python3.10/dist-packages\nRequires: numpy\nRequired-by: albumentations, arviz, bayesian-optimization, Boruta, catboost, category_encoders, cesium, clarabel, cuml-cu12, cvxpy, datascience, datashader, dipy, easyocr, eli5, fastai, featuretools, fury, gensim, hep_ml, hyperopt, ImageHash, imbalanced-learn, imgaug, jax, jaxlib, kaggle-environments, libpysal, librosa, lightgbm, lime, matplotlib-venn, missingno, mizani, mlxtend, mne, nilearn, osqp, phik, plotly-express, plotnine, pyLDAvis, pymc, pymc3, pytensor, qdldl, scikit-image, scikit-learn, scikit-optimize, scikit-plot, scikit-surprise, scs, sentence-transformers, shap, sklearn-pandas, statsmodels, stumpy, Theano, Theano-PyMC, TPOT, treelite, tsfresh, woodwork, xarray-einstats, xgboost, ydata-profiling, yellowbrick\n\nName: numpy\nVersion: 1.24.3\nSummary: Fundamental package for array computing in Python\nHome-page: https://www.numpy.org\nAuthor: Travis E. Oliphant et al.\nAuthor-email: \nLicense: BSD-3-Clause\nLocation: /usr/local/lib/python3.10/dist-packages\nRequires: \nRequired-by: accelerate, albucore, albumentations, arviz, astropy, autograd, bayesian-optimization, bigframes, blis, blosc2, bokeh, Boruta, Bottleneck, bqplot, Cartopy, catboost, category_encoders, cesium, chex, clarabel, cmdstanpy, contourpy, cudf-cu12, cufflinks, cuml-cu12, cupy-cuda12x, cuvs-cu12, cvxpy, dask-cuda, dask-cudf-cu12, datascience, datasets, datashader, db-dtypes, deap, diffusers, dipy, dopamine_rl, easyocr, eli5, fasttext, featuretools, flax, folium, fury, geemap, gensim, geopandas, gym, gymnasium, h5py, hep_ml, holoviews, hyperopt, ibis-framework, ImageHash, imageio, imbalanced-learn, imgaug, ipympl, jax, jaxlib, kaggle-environments, keras, keras-core, keras-hub, langchain, langid, libpysal, librosa, lightgbm, lime, matplotlib, matplotlib-venn, missingno, mizani, mkl-fft, mkl-random, mkl-umath, ml-dtypes, mlcrate, mlxtend, mne, moviepy, music21, nibabel, nilearn, numba, numexpr, nx-cugraph-cu12, onnx, opencv-contrib-python, opencv-python, opencv-python-headless, optax, optuna, orbax-checkpoint, osqp, pandas, pandas-gbq, pandas-stubs, pandasql, patsy, peft, pettingzoo, phik, plotly-express, plotnine, prophet, pycocotools, pydegensac, pyemd, pyerfa, pyLDAvis, pylibraft-cu12, pymc, pymc3, pyogrio, pytensor, python-louvain, PyUpSet, PyWavelets, qdldl, rmm-cu12, rna-predict, scikit-image, scikit-learn, scikit-learn-intelex, scikit-optimize, scikit-surprise, scipy, scs, seaborn, shap, shapely, Shimmy, sklearn-pandas, soxr, spacy, stable-baselines3, stanio, statsmodels, stumpy, tables, tensorboard, tensorflow, tensorflow-datasets, tensorflow-hub, tensorflow-probability, tensorflow_decision_forests, tensorstore, Theano, Theano-PyMC, thinc, tifffile, torchmetrics, torchtune, TPOT, transformers, treelite, trx-python, tsfresh, ucx-py-cu12, ucxx-cu12, visions, wavio, woodwork, wordcloud, xarray, xarray-einstats, xgboost, ydata-profiling, ydf, yellowbrick, yfinance\n\nName: seaborn\nVersion: 0.12.2\nSummary: Statistical data visualization\nHome-page: \nAuthor: \nAuthor-email: Michael Waskom <mwaskom@gmail.com>\nLicense: \nLocation: /usr/local/lib/python3.10/dist-packages\nRequires: matplotlib, numpy, pandas\nRequired-by: missingno, ydata-profiling\n\n=== [End of Final Package Versions] ===\n\n[INFO] Done. If Kaggle forcibly reverts any packages, you may need to re-run this cell.\n"
    }
   ],
   "source": [
    "# ---\n",
    "# Cell: ALL-IN-ONE Environment Setup\n",
    "# ---\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import subprocess\n",
    "import sys\n",
    "\n",
    "\n",
    "def run_and_print(cmd):\n",
    "    result = subprocess.run(cmd, capture_output=True, text=True)\n",
    "    print(result.stdout)\n",
    "    if result.stderr:\n",
    "        print(result.stderr)\n",
    "\n",
    "\n",
    "# ====== Gather Detailed System Information ======\n",
    "print(\"\\n=== [System Information] ===\")\n",
    "\n",
    "# Python version\n",
    "print(\"\\n[Python Version]\")\n",
    "print(sys.version)\n",
    "\n",
    "# Kernel and OS details\n",
    "print(\"\\n[Kernel and OS Information]\")\n",
    "run_and_print([\"uname\", \"-a\"])\n",
    "\n",
    "# CPU details\n",
    "print(\"\\n[CPU Information]\")\n",
    "run_and_print([\"lscpu\"])\n",
    "\n",
    "# Memory details\n",
    "print(\"\\n[Memory Information]\")\n",
    "run_and_print([\"free\", \"-mh\"])\n",
    "\n",
    "# Disk and partition details\n",
    "print(\"\\n[Disk Information]\")\n",
    "run_and_print([\"lsblk\"])\n",
    "\n",
    "print(\"\\n=== [End of System Information] ===\\n\")\n",
    "\n",
    "# ====== USER CONFIG ======\n",
    "RNA_PREDICT_VERSION = \"0.1.3\"\n",
    "BLOCK_SPARSE_WHEEL = (\n",
    "    \"block_sparse_attn-0.0.1cu118torch2.0cxx11abiTRUE-cp310-cp310-linux_x86_64.whl\"\n",
    ")\n",
    "\n",
    "TORCH_2_0_CU118_SPEC = \"torch==2.0.1+cu118\"  # Torch pip spec\n",
    "TORCH_INDEX_URL = \"https://download.pytorch.org/whl/cu118\"\n",
    "\n",
    "\n",
    "def fix_environment():\n",
    "    \"\"\"\n",
    "    1) Uninstall Torch + rna_predict + scikit-learn + SciPy + NumPy + Seaborn\n",
    "    2) Install Torch 2.0.1+cu118 (with dependencies, so pip can unify them)\n",
    "    3) Copy+rename and install block_sparse_attn\n",
    "    4) Install rna_predict\n",
    "    5) Force reinstall pinned scikit-learn, SciPy, NumPy, Seaborn\n",
    "    6) Show final versions\n",
    "    \"\"\"\n",
    "\n",
    "    # 1) Uninstall everything relevant\n",
    "    print(\n",
    "        \"[INFO] Uninstalling Torch, rna_predict, scikit-learn, SciPy, NumPy, Seaborn...\\n\"\n",
    "    )\n",
    "    packages_to_remove = [\n",
    "        \"torch\",\n",
    "        \"torchvision\",\n",
    "        \"torchaudio\",\n",
    "        \"rna-predict\",\n",
    "        \"scikit-learn\",\n",
    "        \"scipy\",\n",
    "        \"numpy\",\n",
    "        \"seaborn\",\n",
    "    ]\n",
    "    run_and_print([\"pip\", \"uninstall\", \"-y\"] + packages_to_remove)\n",
    "\n",
    "    # 2) Install Torch 2.0.1+cu118 from PyPI\n",
    "    print(f\"\\n[INFO] Installing Torch: {TORCH_2_0_CU118_SPEC} from {TORCH_INDEX_URL}\\n\")\n",
    "    run_and_print(\n",
    "        [\n",
    "            \"pip\",\n",
    "            \"install\",\n",
    "            TORCH_2_0_CU118_SPEC,\n",
    "            \"--index-url\",\n",
    "            TORCH_INDEX_URL,\n",
    "            \"--no-cache-dir\",\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    # 3) Copy & rename the block_sparse_attn wheel, then install it\n",
    "    original_whl_path = f\"/kaggle/input/block-sparse-wheels/{BLOCK_SPARSE_WHEEL}\"\n",
    "    renamed_whl_name = (\n",
    "        \"block_sparse_attn-0.0.1+cu118torch2.0-cp310-cp310-linux_x86_64.whl\"\n",
    "    )\n",
    "    renamed_whl_path = f\"/kaggle/working/{renamed_whl_name}\"\n",
    "\n",
    "    try:\n",
    "        shutil.copyfile(original_whl_path, renamed_whl_path)\n",
    "        print(\n",
    "            \"\\n[INFO] Copied and renamed block-sparse-attn wheel for PEP 440 compliance.\\n\"\n",
    "        )\n",
    "    except Exception as e:\n",
    "        print(f\"\\n[ERROR] Could not copy the wheel file: {e}\")\n",
    "        print(\"Verify the input file path or rename manually.\")\n",
    "\n",
    "    print(f\"\\n[INFO] Installing block-sparse-attn from: {renamed_whl_path}\")\n",
    "    run_and_print([\"pip\", \"install\", \"--no-cache-dir\", \"--no-deps\", renamed_whl_path])\n",
    "\n",
    "    # 4) Install rna_predict with --no-deps\n",
    "    rnapred_whl_path = f\"/kaggle/input/rna-structure-predict/rna_predict-{RNA_PREDICT_VERSION}-py3-none-any.whl\"\n",
    "    print(f\"\\n[INFO] Installing rna_predict from: {rnapred_whl_path}\")\n",
    "    run_and_print([\"pip\", \"install\", \"--no-cache-dir\", \"--no-deps\", rnapred_whl_path])\n",
    "\n",
    "    # 5) Force reinstall pinned scikit-learn, SciPy, NumPy, Seaborn\n",
    "    fix_packages = [\n",
    "        \"numpy==1.24.3\",\n",
    "        \"scipy==1.10.1\",\n",
    "        \"scikit-learn==1.2.2\",\n",
    "        \"seaborn==0.12.2\",\n",
    "    ]\n",
    "    print(\n",
    "        \"\\n[INFO] Reinstalling scikit-learn, SciPy, NumPy, Seaborn to fix version conflicts...\"\n",
    "    )\n",
    "    run_and_print(\n",
    "        [\"pip\", \"install\", \"--upgrade\", \"--force-reinstall\", \"--no-cache-dir\"]\n",
    "        + fix_packages\n",
    "    )\n",
    "\n",
    "    # 6) Show final versions\n",
    "    print(\"\\n=== [Final Package Versions] ===\")\n",
    "    for pkg in [\"torch\", \"scikit-learn\", \"scipy\", \"numpy\", \"seaborn\"]:\n",
    "        run_and_print([\"pip\", \"show\", pkg])\n",
    "    print(\"=== [End of Final Package Versions] ===\\n\")\n",
    "\n",
    "    print(\n",
    "        \"[INFO] Done. If Kaggle forcibly reverts any packages, you may need to re-run this cell.\"\n",
    "    )\n",
    "\n",
    "\n",
    "# === Run everything ===\n",
    "fix_environment()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:17:51.676759Z",
     "iopub.status.busy": "2025-03-29T01:17:51.676424Z",
     "iopub.status.idle": "2025-03-29T01:17:52.574376Z",
     "shell.execute_reply": "2025-03-29T01:17:52.573791Z",
     "shell.execute_reply.started": "2025-03-29T01:17:51.676727Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Cell 1: ENVIRONMENT SETUP & LOGGING\n",
    "-----------------------------------\n",
    "\"\"\"\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Machine Learning Libraries\n",
    "from sklearn.impute import SimpleImputer\n",
    "from xgboost import XGBRegressor\n",
    "\n",
    "# Logging\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO,\n",
    "    format=\"%(asctime)s [%(levelname)s] %(message)s\",\n",
    "    handlers=[logging.StreamHandler(sys.stdout)],\n",
    ")\n",
    "logging.info(\"Cell 1 complete: Libraries imported and logging initialized.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:17:52.575303Z",
     "iopub.status.busy": "2025-03-29T01:17:52.575104Z",
     "iopub.status.idle": "2025-03-29T01:17:52.974227Z",
     "shell.execute_reply": "2025-03-29T01:17:52.973591Z",
     "shell.execute_reply.started": "2025-03-29T01:17:52.575286Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 2: DATA IMPORT\n",
    "-------------------\n",
    "Here, we read in train/validation/test CSVs and a sample submission from the Kaggle environment.\n",
    "Adjust the paths if needed for your environment.\n",
    "\"\"\"\n",
    "\n",
    "# Example file paths\n",
    "TRAIN_SEQUENCES_PATH = \"/kaggle/input/stanford-rna-3d-folding/train_sequences.csv\"\n",
    "TRAIN_LABELS_PATH = \"/kaggle/input/stanford-rna-3d-folding/train_labels.csv\"\n",
    "VALID_SEQUENCES_PATH = \"/kaggle/input/stanford-rna-3d-folding/validation_sequences.csv\"\n",
    "VALID_LABELS_PATH = \"/kaggle/input/stanford-rna-3d-folding/validation_labels.csv\"\n",
    "TEST_SEQUENCES_PATH = \"/kaggle/input/stanford-rna-3d-folding/test_sequences.csv\"\n",
    "SAMPLE_SUB_PATH = \"/kaggle/input/stanford-rna-3d-folding/sample_submission.csv\"\n",
    "\n",
    "try:\n",
    "    train_sequences = pd.read_csv(TRAIN_SEQUENCES_PATH)\n",
    "    train_labels = pd.read_csv(TRAIN_LABELS_PATH)\n",
    "    validation_sequences = pd.read_csv(VALID_SEQUENCES_PATH)\n",
    "    validation_labels = pd.read_csv(VALID_LABELS_PATH)\n",
    "    test_sequences = pd.read_csv(TEST_SEQUENCES_PATH)\n",
    "    sample_submission = pd.read_csv(SAMPLE_SUB_PATH)\n",
    "\n",
    "    logging.info(\"Cell 2 complete: Data loaded successfully.\")\n",
    "except Exception as e:\n",
    "    logging.error(f\"Error loading data: {e}\")\n",
    "    sys.exit(1)\n",
    "\n",
    "logging.info(\n",
    "    f\"train_sequences: {train_sequences.shape}, train_labels: {train_labels.shape}\"\n",
    ")\n",
    "logging.info(\n",
    "    f\"validation_sequences: {validation_sequences.shape}, validation_labels: {validation_labels.shape}\"\n",
    ")\n",
    "logging.info(\n",
    "    f\"test_sequences: {test_sequences.shape}, sample_submission: {sample_submission.shape}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:17:52.975304Z",
     "iopub.status.busy": "2025-03-29T01:17:52.975016Z",
     "iopub.status.idle": "2025-03-29T01:17:53.261853Z",
     "shell.execute_reply": "2025-03-29T01:17:53.261096Z",
     "shell.execute_reply.started": "2025-03-29T01:17:52.975276Z"
    },
    "trusted": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAx8AAAGJCAYAAADrBI7SAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuNSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/xnp5ZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA/UElEQVR4nO3dd3wU1cL/8e+m7IZ0FBKI9N5EFJQblGYCQXLVcBEbKhErYm+AioB6QRDhKirXq5dyuQKPiiKKKAlFEStIQKpU9aEFKYFQkpCc3x/8dp5s6m5IJiR83q9XXrAzZ2bOOXtmM9+dEocxxggAAAAAKphfZVcAAAAAwPmB8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAVRDDodDo0ePruxqePjpp5/UpUsXhYSEyOFwKC0trbKrhCpk9OjRcjgc+vPPP89qPX379tU999xTTrXyjrvu54sePXqoR48e1utdu3bJ4XBoxowZpS6bnJysRo0alWt9ZsyYIYfDoV27dpXrevPbuHGjAgICtH79+grbBlBdED4AH7h/ieX/iYqKUs+ePbVo0aLKrt5Z27hxo0aPHl3uv6RzcnI0YMAAHTp0SJMnT9asWbPUsGHDYsvv2rVLd955p5o2baqgoCDVqVNH3bp106hRo8q1Xucb90HgxIkTK7sqxRo7dqzmz59fIeteuXKlFi9erGHDhkmSGjVqVGh/LurHm4Pmquijjz6Sw+HQu+++W2yZlJQUORwOvf766zbWrGwqcuyUpk2bNkpMTNTzzz9fKdsHqpKAyq4AUBW98MILaty4sYwx2r9/v2bMmKG+ffvq008/1V//+tfKrl6Zbdy4UWPGjFGPHj3K9dvH7du367ffftM777yju+++u8Sy27Zt0+WXX64aNWpo8ODBatSokfbu3auff/5Z48eP15gxY8qtXjj3jB07VjfccIOSkpLKfd2vvPKK4uLi1KxZM0nSP/7xD2VmZlrzP//8c82ZM0eTJ09WrVq1rOldunQ5q+0+99xzGj58+FmtoyIkJiYqIiJCs2fPLna/nD17tvz9/XXzzTeXeTsNGzbUyZMnFRgYWOZ1eKO4sXP77bfr5ptvlsvlqtDt33///erbt6+2b9+upk2bVui2gKqM8AGUwTXXXKNOnTpZr++66y5FR0drzpw5VTp8VJT09HRJUmRkZKllJ0+erMzMTKWlpRU6O+JeD+Cr9PR0LVy4UP/85z+taQUPUvft26c5c+YoKSmpxPB9/PhxhYSEeL3tgIAABQRU7K/b0aNHa8aMGT6dtXS5XLrhhhs0ffp07dmzRzExMR7zT506pY8//li9evVSVFRUmevmcDgUFBRU5uXPlr+/v/z9/St8O/Hx8apZs6ZmzpypF154ocK3B1RVXHYFlIPIyEjVqFGj0AHG8ePH9cQTT6h+/fpyuVxq2bKlJk6cKGOMJOnkyZNq1aqVWrVqpZMnT1rLHTp0SHXr1lWXLl2Um5sr6cy10KGhodqxY4cSEhIUEhKimJgYvfDCC9b6SrJmzRpdc801Cg8PV2hoqOLi4vT9999b82fMmKEBAwZIknr27GldcrJ8+fIS17t06VJ17dpVISEhioyM1PXXX69NmzZZ85OTk9W9e3dJ0oABA+RwODyuBy9o+/btqlevXpGXZRV1ALRo0SJr+2FhYUpMTNSGDRsKlZs/f77atWunoKAgtWvXTh9//HGh68uXL19eZJuLu2Z98+bNuuGGG3TBBRcoKChInTp10oIFCzzKuC/VW7lypR5//HHVrl1bISEh6tevnw4cOFBke7p3766wsDCFh4fr8ssv1+zZsz3K/PDDD+rTp48iIiIUHBys7t27a+XKlYXWVVZZWVkaNWqUmjVrJpfLpfr16+vpp59WVlaWRzmHw6EHH3zQ6luXy6W2bdvqiy++KLTO5cuXq1OnTgoKClLTpk319ttvF7oXwuFw6Pjx45o5c6Y1/pKTkz3Wc+TIESUnJysyMlIRERG68847deLEiVLbtHDhQp0+fVrx8fE+9YV7v9u+fbv69u2rsLAwDRw4UJK0YsUKDRgwQA0aNLD66bHHHvPYl6Wi7/nwpe8q0m233aa8vDzNnTu30LyFCxcqIyPDau/06dN19dVXKyoqSi6XS23atNHUqVNL3UZx+09R+2RRJk6cqC5duujCCy9UjRo11LFjR3344YceZUoaO8Xd8/HWW2+pbdu2crlciomJ0dChQ3XkyBGPMj169FC7du20ceNG9ezZU8HBwbrooos0YcKEQvUMDAxUjx499Mknn5TaJ8D5jPABlEFGRob+/PNPHThwQBs2bNCQIUOUmZmp2267zSpjjNF1112nyZMnq0+fPpo0aZJatmypp556So8//rgkqUaNGpo5c6a2bdumZ5991lp26NChysjI0IwZMzy+scvNzVWfPn0UHR2tCRMmqGPHjho1alSp90Js2LBBXbt21dq1a/X0009r5MiR2rlzp3r06KEffvhBktStWzc9/PDDkqRnnnlGs2bN0qxZs9S6deti15uamqqEhASlp6dr9OjRevzxx/Xtt9/qyiuvtH7R33fffXrmmWckSQ8//LBmzZrl0daCGjZsqD/++ENLly4tsU2SNGvWLCUmJio0NFTjx4/XyJEjtXHjRl111VUeBxqLFy9W//795XA4NG7cOCUlJenOO+/UqlWrSt1GcTZs2KC//OUv2rRpk4YPH65XX31VISEhSkpKKvIg6qGHHtLatWs1atQoDRkyRJ9++qkefPBBjzIzZsxQYmKiDh06pBEjRujll19Whw4dPA5Ily5dqm7duuno0aMaNWqUxo4dqyNHjujqq6/Wjz/+WOb2uOXl5em6667TxIkTde2112rKlClKSkrS5MmTddNNNxUq/8033+iBBx7QzTffrAkTJujUqVPq37+/Dh48aJVZs2aN+vTpo4MHD2rMmDG666679MILLxS6Pn/WrFlyuVzq2rWrNf7uu+8+jzI33nijjh07pnHjxunGG2/UjBkzvLoU79tvv9WFF15Y4r1GxTl9+rQSEhIUFRWliRMnqn///pKkDz74QCdOnNCQIUM0ZcoUJSQkaMqUKbrjjju8Wq83fVfRunXrpnr16hUKuNKZS66Cg4OtM0RTp05Vw4YN9cwzz+jVV19V/fr19cADD+jNN9/0ebu+7JOvvfaaLr30Ur3wwgsaO3asAgICNGDAAC1cuNAq483YyW/06NEaOnSoYmJi9Oqrr6p///56++231bt3b+Xk5HiUPXz4sPr06aNLLrlEr776qlq1aqVhw4YVeZ9fx44dtX79eh09etTnPgHOGwaA16ZPn24kFfpxuVxmxowZHmXnz59vJJmXXnrJY/oNN9xgHA6H2bZtmzVtxIgRxs/Pz3z99dfmgw8+MJLMP/7xD4/lBg0aZCSZhx56yJqWl5dnEhMTjdPpNAcOHLCmSzKjRo2yXiclJRmn02m2b99uTduzZ48JCwsz3bp1s6a5t71s2TKv+qNDhw4mKirKHDx40Jq2du1a4+fnZ+644w5r2rJly4wk88EHH5S6zvXr15saNWoYSaZDhw7mkUceMfPnzzfHjx/3KHfs2DETGRlp7rnnHo/p+/btMxERER7TO3ToYOrWrWuOHDliTVu8eLGRZBo2bFiongXbv3PnTiPJTJ8+3ZoWFxdnLr74YnPq1ClrWl5enunSpYtp3ry5Nc09ZuLj401eXp41/bHHHjP+/v5WnY4cOWLCwsJM586dzcmTJz22714uLy/PNG/e3CQkJHis68SJE6Zx48amV69eRfZpwXa88sorxZaZNWuW8fPzMytWrPCY/s9//tNIMitXrrSmSTJOp9NjLK9du9ZIMlOmTLGmXXvttSY4ONjs3r3bmrZ161YTEBBgCv4aCgkJMYMGDSpUr1GjRhlJZvDgwR7T+/XrZy688MIS222MMVdddZXp2LFjiWVeeeUVI8ns3LnTmube74YPH16o/IkTJwpNGzdunHE4HOa3334rVPf8vO07b40aNcpjLPviqaeeMpLMli1brGkZGRkmKCjI3HLLLda0otqbkJBgmjRp4jGte/fupnv37tbrovYfb/fJorabnZ1t2rVrZ66++mqP6cWNHfc+6H5f09PTjdPpNL179za5ublWuTfeeMNIMtOmTfNoiyTzn//8x5qWlZVl6tSpY/r3719oW7NnzzaSzA8//FBoHoAzOPMBlMGbb76plJQUpaSk6L///a969uypu+++Wx999JFV5vPPP5e/v791NsHtiSeekDHG41uz0aNHq23btho0aJAeeOABde/evdBybvm/LXdfupGdna3U1NQiy+fm5mrx4sVKSkpSkyZNrOl169bVrbfeqm+++aZM39Lt3btXaWlpSk5O1gUXXGBNb9++vXr16qXPP//c53VKUtu2bZWWlqbbbrtNu3bt0muvvaakpCRFR0frnXfescqlpKToyJEjuuWWW/Tnn39aP/7+/urcubOWLVvmUc9BgwYpIiLCWr5Xr15q06ZNmep46NAhLV261PoW3r3tgwcPKiEhQVu3btXu3bs9lrn33ns9Lr3p2rWrcnNz9dtvv1ntOXbsmIYPH17o+nj3cmlpadq6datuvfVWHTx40Nru8ePHFRcXp6+//lp5eXllapPbBx98oNatW6tVq1Ye/Xr11VdLktWvbvHx8R4317Zv317h4eHasWOHpDPjLzU1VUlJSR73FDRr1kzXXHONz/W7//77PV537dpVBw8eLHUMHzx4UDVr1vR5e25DhgwpNK1GjRrW/48fP64///xTXbp0kTFGa9asKXWdpfVdSfK/N3/++adOnDihvLy8QtMLXipXFPcZ2/xnP+bNm6dTp05Zl1wVbK/77G/37t21Y8cOZWRklLodN1/3yfzbPXz4sDIyMtS1a1f9/PPPXm8zv9TUVGVnZ+vRRx+Vn9//HQbdc889Cg8P9zijIkmhoaEeZ7WdTqeuuOKKIt8n9xg720dCA9UZN5wDZXDFFVd43HB+yy236NJLL9WDDz6ov/71r3I6nfrtt98UExOjsLAwj2XdlzG5DzqlM7/Mpk2bpssvv1xBQUGaPn16kX8XwM/PzyNASFKLFi0kqdgbTQ8cOKATJ06oZcuWhea1bt1aeXl5+uOPP9S2bVvvGv//uetf3Hq//PJLn2/MdWvRooVmzZql3Nxcbdy4UZ999pkmTJige++9V40bN1Z8fLy2bt0qSdZBcUHh4eEe9WzevHmhMi1btizTAcy2bdtkjNHIkSM1cuTIIsukp6froosusl43aNDAY777IOXw4cOSztzrIknt2rUrdrvuNg8aNKjYMhkZGWd1kL1161Zt2rRJtWvXLnJ+wZv+C7ZLOtM2d7vS09N18uRJ6wlT+RU1rTQl9aP7PS+O8eLeqKIEBASoXr16hab//vvvev7557VgwQKrvW7eHIyX1nclKe79KTh9+vTphe6bKah9+/Zq166d5syZY/19oNmzZ6tWrVpKSEiwyq1cuVKjRo3Sd999V+g+m4yMDI8gURJf98nPPvtML730ktLS0jzCVFn/dkpxn11Op1NNmjTx+GyWpHr16hXaVs2aNbVu3bpC63aPsfPp77oAviJ8AOXAz89PPXv21GuvvaatW7f6fCAvSV9++aWkM0+Y2bp1qxo3blze1axy/P39dfHFF+viiy9WbGysevbsqffee0/x8fHWN/yzZs1SnTp1Ci1blqcLFXfA4L7p38297SeffNLj4Cy/ggfWxT1tx5cDYvd2X3nlFXXo0KHIMqGhoV6vr7htXHzxxZo0aVKR8+vXr+/xujza5Yuybu/CCy/06qC+KC6Xy+MbcunMmOjVq5cOHTqkYcOGqVWrVgoJCdHu3buVnJzs1Rmos+m7lJQUj9f/+c9/tHjxYv33v//1mO7tZ9Ftt92m4cOHa9WqVapXr56WLVum++67z9qPtm/frri4OLVq1UqTJk1S/fr15XQ69fnnn2vy5MlnfcatOCtWrNB1112nbt266a233lLdunUVGBio6dOnF3mfSkXw5X1yj7H8j2oG4InwAZST06dPS5L1dwMaNmyo1NRUHTt2zOPsx+bNm635buvWrdMLL7ygO++8U2lpabr77rv1yy+/FPomMS8vTzt27LDOdkjSr7/+KknFPhq0du3aCg4O1pYtWwrN27x5s/z8/KwDSl++rXPXv7j11qpVq0xnPYrjPtO0d+9eSbIuV4mKiirxCUbuerrPGuRXsO7ub9ELPvGm4Deh7rNPgYGBPj89qTju9qxfv77YMwLuMuHh4eW23aK2sXbtWsXFxZXLt7dRUVEKCgrStm3bCs0ralpFfWPcqlUrzZs3r9zW98svv+jXX3/VzJkzPW4wLxgKKkrB9/+bb75RUFBQmcfFLbfcohEjRmj27Nlq2LChcnNzPS65+vTTT5WVlaUFCxZ4nLEpeBmeN3zZJ+fNm6egoCB9+eWXHn+nY/r06YWW9Xbs5P/syn8mOTs7Wzt37jyrfWvnzp3y8/Pz+IwG4Il7PoBykJOTo8WLF8vpdFqXVfXt21e5ubl64403PMpOnjxZDofDut49JydHycnJiomJ0WuvvaYZM2Zo//79euyxx4rcVv71GWP0xhtvKDAwUHFxcUWW9/f3V+/evfXJJ594XJq1f/9+zZ49W1dddZV1uYo7LBQ8+C5K3bp11aFDB82cOdOj/Pr167V48WL17du31HUUZcWKFYWeNiPJuofEfalEQkKCwsPDNXbs2CLLux9jm7+e+S+FSUlJ0caNGz2Wadiwofz9/fX11197TH/rrbc8XkdFRalHjx56++23rTBU1LZ90bt3b4WFhWncuHE6deqUxzz3N6wdO3ZU06ZNNXHiRI8/jnc22y3oxhtv1O7duz3ur3E7efKkjh8/7tP6/P39FR8fr/nz52vPnj3W9G3bthX5tKCQkBCvxp+vYmNjdfjwYa/up/CG+9vw/N9+G2P02muvlcv67dagQQN17dpV//M//6P//ve/aty4sccfVyyqvRkZGUWGgNL4sk/6+/vL4XB4nH3ctWtXkX/J3NuxEx8fL6fTqddff92jPf/+97+VkZGhxMREn9vktnr1arVt29brS9CA8xFnPoAyWLRokXUGIz09XbNnz9bWrVs1fPhw60D+2muvVc+ePfXss89q165duuSSS7R48WJ98sknevTRR61vsd3XMi9ZskRhYWFq3769nn/+eT333HO64YYbPA7ig4KC9MUXX2jQoEHq3LmzFi1apIULF+qZZ54p9hpw9zZSUlJ01VVX6YEHHlBAQIDefvttZWVleTyvvkOHDvL399f48eOVkZEhl8tlPde/KK+88oquueYaxcbG6q677tLJkyc1ZcoURUREWNeO+2r8+PFavXq1/va3v6l9+/aSpJ9//ln/+c9/dMEFF+jRRx+VdObb/6lTp+r222/XZZddpptvvlm1a9fW77//roULF+rKK6+0gtq4ceOUmJioq666SoMHD9ahQ4c0ZcoUtW3b1uMgPiIiQgMGDNCUKVPkcDjUtGlTffbZZ0X+ccM333xTV111lS6++GLdc889atKkifbv36/vvvtO//u//6u1a9f61O7w8HBNnjxZd999ty6//HLdeuutqlmzptauXasTJ05o5syZ8vPz07vvvqtrrrlGbdu21Z133qmLLrpIu3fv1rJlyxQeHq5PP/201G0tWbKkUMCRzvzRvdtvv13vv/++7r//fi1btkxXXnmlcnNztXnzZr3//vv68ssvPe538sbo0aO1ePFiXXnllRoyZIgVytu1a6e0tDSPsh07dlRqaqomTZqkmJgYNW7cWJ07d/Zpe0VJTExUQECAUlNTde+99571+lq1aqWmTZvqySef1O7duxUeHq558+aV+dKuc8Ftt92me++9V3v27Cn0OOzevXvL6XTq2muv1X333afMzEy98847ioqKKjKAl8bbfTIxMVGTJk1Snz59dOuttyo9PV1vvvmmmjVrVuieC2/HTu3atTVixAiNGTNGffr00XXXXactW7borbfe0uWXX+5xc7kvcnJy9NVXX+mBBx4o0/LAecP+B2wBVVdRj9oNCgoyHTp0MFOnTvV4/KkxZx4H+9hjj5mYmBgTGBhomjdvbl555RWr3OrVq01AQIDH43ONMeb06dPm8ssvNzExMebw4cPGmDOP/AwJCTHbt283vXv3NsHBwSY6OtqMGjXK43GRxhR+1K4xxvz8888mISHBhIaGmuDgYNOzZ0/z7bffFmrjO++8Y5o0aWL8/f29euxuamqqufLKK02NGjVMeHi4ufbaa83GjRs9yvjyqN2VK1eaoUOHmnbt2pmIiAgTGBhoGjRoYJKTkz0eFZx/3QkJCSYiIsIEBQWZpk2bmuTkZLNq1SqPcvPmzTOtW7c2LpfLtGnTxnz00Udm0KBBhR7reeDAAdO/f38THBxsatasae677z6zfv36Qo8KNcaY7du3mzvuuMPUqVPHBAYGmosuusj89a9/NR9++KFVxj1mfvrppyL7pGD/LliwwHTp0sXqzyuuuMLMmTPHo8yaNWvM3/72N3PhhRcal8tlGjZsaG688UazZMmSEvvW/cjT4n5mzZpljDnzKNPx48ebtm3bGpfLZWrWrGk6duxoxowZYzIyMqz1STJDhw4ttJ2GDRsWeuTpkiVLzKWXXmqcTqdp2rSpeffdd80TTzxhgoKCPMpt3rzZdOvWzXrcsns97sfV5n+kdP7+zf943OJcd911Ji4urtj5xT1qNyQkpMjyGzduNPHx8SY0NNTUqlXL3HPPPdbjcvOPleIetett33njbB6163bo0CHjcrmMpEL7sDFnxmb79u1NUFCQadSokRk/fryZNm1aoT7z5lG7xni/T/773/82zZs3Ny6Xy7Rq1cpMnz69yD4tbuwUN0beeOMN06pVKxMYGGiio6PNkCFDrM/b/G1p27Ztob4oqp6LFi0ykszWrVsLlQfwfxzGVNBdgQDKVXJysj788MMiL7dB2SQnJ2v58uXFPikMFSspKUkbNmwo8tr/irBixQr16NFDmzdvLvJJS8DZSEpKksPhKPYvtQM4g3s+AAAV7uTJkx6vt27dqs8//1w9evSwrQ5du3ZV7969PS41BMrDpk2b9Nlnn+nFF1+s7KoA5zzu+QAAVLgmTZooOTnZ+jsKU6dOldPp1NNPP21rPYq6yR04W61bt7aeeAigZIQPAECF69Onj+bMmaN9+/bJ5XIpNjZWY8eO5fInADjPcM8HAAAAAFtwzwcAAAAAWxA+AAAAANiizPd85OXlac+ePQoLC5PD4SjPOgEAAACoQowxOnbsmGJiYuTnV/z5jTKHjz179qh+/fplXRwAAABANfPHH3+oXr16xc4vc/gICwuzNhAeHl7W1QAAAACo4o4ePar69etbGaE4ZQ4f7kutwsPDCR8AAAAASr0dgxvOAQAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFgGVXYHyZIxRdna29f+cnBxJUmBgoBwOR7HLOZ3OEucDAAAAOHvVKnxkZ2dr2LBhPi83fvx4uVyuCqgRAAAAADcuuwIAAABgi2p15iO/Btfdqd8XTJckNfrbvfILCPSYn3c6R7s++ldlVA0AAAA4L1Xb8JE/bPgFBBYKHwAAAADsxWVXAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtAiq7AmfLGKPs7Gzr/3Zv0+l0yuFw2LJdAAAAoCqr8mc+srOzNWzYMA0bNkw5OTm2b9MdQgAAAACUrMqHDwAAAABVA+EDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALBFQGVXoKp74YUXdPr0aWVnZ8sY4zEvMDBQxhidPn1aLpdLrVu31rp16xQXF6fExESv1r9+/XrNnTtXWVlZysnJkb+/v3Jzc+Xv76+8vDxFRUVp//796tChg5KTk61l5s2bp/79+6tdu3Y+t6mo5devX685c+ZY2+7SpYtWrVpV5m14Y+HChVqyZIlP/VUeSus/X/r3bN+L6sib/p07d65ycnIUGBio2NhYffvtt8rNzZUxRnl5eTp9+rTi4+NtHRcF65i/DcW1ydv3371/ORwO3XzzzT6PFcZZ9eXN52BFvf/u9Xbq1EmrVq1Sp06d9PXXXys7O9ur/a8s9aruY7m6t89X+ftDUrkfu1Ql3tS/uOOzqtZuznycpePHjysrK6tQ8JCknJwcnT59WpKUlZWltLQ05eXlKTU1VZmZmaWuOzs7W++//74yMzOVk5MjScrNzbX+NcZo//79kqS0tDQdOnRI2dnZ+uCDD3T48GF98MEHys7O9qk9RS3vrsfx48d16tQpHT9+XCkpKWXehjcyMzOVmprqU3+Vh9L6z5f+Pdv3ojrypn/dYz4rK0uZmZlKSUmxxp47hBtjbB0XJbUhMzOzyDZ5+/7n378yMzP1/vvv+zRWGGfVlzefgxX1/udfb2pqqg4fPqyUlBTr911p+19Z6lXdx3J1b5+v8vfH+++/r/fff79cj12qEm/qX9zxWVVsN+GjEhhjNG3atFLLpaam6ujRo16v9/XXX/dY5ujRo0pNTfWpbkUtX1I9yrINb0ybNs0KdN72V3korf986d+zfS+qI1/6tzR2jov8CrZh2rRpRbbJ2/e/YJt9HSuMs+rLm8/Binr/86+3qC/XStv/ylKv6j6Wq3v7fFWwP8r72KUq8ab+pR2fVaV2V/nwkf9DMX/iK+rDsqTlsrKyvP7Zu3fvWdd7x44d2rJlS7HzDxw44PMgOnLkiBYvXuzxy2rJkiU6cOCAV8u7t5l/+dTUVC1evLjYZXzdhje2bNmiHTt2eEwrrb/KQ1Htz9+20ub7sq7zkbf96ws7xkV+RbVhx44dhdq0ZcsWr97/AwcOKCUlpdB2UlNTvRorjLPqy5vPwYp6/wuutzjF7X9lqVd1H8vVvX2+KmmMlcexS1XqW2/qX9zxWUpKSpVst9fhIysry0qm+RNqZXNfjiRJL774ovV/k3u6xOXyzx85cqSGDRvm9c8//vGPcqn7zJkzlZeXV7huxujDDz8scp6vjDGaN2+eV2Fs3rx5haZ7U4e8vDyvtuGNvLw8zZw5s8h5xfVXeSiu/e7p7jYWNz9/20tbV3n0U1XjTf+WdcxX5LjIr7g2FFTcGC74/rv386LGg7s/ShorjLPqy5vPwYp6/70d5wXrU9ryJdWruo/l6t4+X3kzxs722KWq9K039S/p+Kxg+6pKu70OH+PGjVNERIT1U79+/Yqs13nhxIkT2rhxY6Hp+/fvL7dvc/Py8rR582br3pDi7N+/X5s3by7TQZwxxqtteGPjxo06ceJEkfOK66/yUFz73f23cePGEufnb3tp6yqPfqpqvOnfso75ihwX+Xm7jxhjdOLEiVLf/9L28y1btpQ4Vhhn1Zc3n4MV9f77+rug4P5XlnpV97Fc3dvnK2/G2Nkeu1SVvvWm/r7sk1Wl3V4/7WrEiBF6/PHHrddHjx49JwJIYGCg9f+RI0daZz8c/iU3Lf/8F198UU6n06vtGWP07rvvatu2bWWoraeQkBC1adOm0PTo6Gi1bNmyXAKIn5+fWrRooejo6BLLRUdHq1WrVvr11199DiAOh0MtW7YsdRveaNOmjYKDg4v8xVtcf5WH4trv7r82bdqUOD9/20tbV3n0U1XjTf+WdcxX5LjIz9t9xOFwqEaNGjp16lSJ739p+3mrVq1KHCuMs+rLm89Bh8NRIe+/r78LCu5/ZRmX1X0sV/f2+cqbMXa2xy5VpW+9rb+3+2RVabfXZz5cLpfCw8M9fs4FDofD+n/+AJF/ujfLuVwur36CgoLUr1+/cqn7oEGD5OdX+C1wOBy64YYbipznK/e6vOkP96Pu8vOmDn5+fl5twxt+fn4aNGhQkfOK66/yUFz7878XJc3P3/bS1lUe/VTVeNO/ZR3zFTku8iuuDQX5+flZj70uuHz+97+k8eDNPsU4q768+RysqPff23FesD6lLV9Svar7WK7u7fOVN2PsbI9dqkrfelP/ko7PCravqrS7yt9wXhlq1ap11uto0qSJWrRoUez82rVrKz4+3qd1RkZGqnfv3h4HN3FxcV7X173N/MvHx8erd+/exS7j6za80bJlSzVp0sRjWmn9VR6Kan/+tpU235d1nY+87V9f2DEu8iuqDU2aNCnUphYtWnj1/teuXVu9evUqtJ34+HivxgrjrPry5nOwot7/gustTnH7X1nqVd3HcnVvn69KGmPlcexSlfrWm/oXd3zWq1evKtluwkclcDgcGjx4cKnl4uPjfTrD9PDDD3ssExER4fPBXFHLl1SPsmzDG4MHD/bYobzpr/JQWv/50r9n+15UR770b2nsHBf5FWzD4MGDi2yTt+9/wTb7OlYYZ9WXN5+DFfX+519vcQeIJe1/ZalXdR/L1b19virYH+V97FKVeFP/0o7PqlK7CR9nKSQkRC6Xq8gP58DAQAUEnLm3xOVyqUOHDvLz81N8fLxCQ0NLXbfT6dSNN96o0NBQ694Wf39/61+Hw2Fd19ehQwddcMEFcjqdGjBggGrWrKkbbrjB63tZ8m+z4PLueoSEhCgoKEghISHq1atXmbfhjdDQUMXHx/vUX+WhtP7zpX/P9r2ojrzpX/eYd7lcCg0NVa9evayx53K5FBgYaH3rY9e4KKkNoaGhRbbJ2/c///7lXpcvY4VxVn158zlYUe9//vXGx8erZs2a6tWrl/X7rrT9ryz1qu5jubq3z1f5+2PAgAG68cYby/XYpSrxpv7FHZ9VxXY7TBmfx3X06FFFREQoIyOjUu//yMrK0rBhwySduXF85MiRkqRGf7tXuz76lySpyY1D5RcQ6LFc3ukc7Xj/TUnS+PHj5XK5yrRNX5cFAAAAqhtvswFnPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGCLgMquwNlyOp0aP368JMkYY/s2nU6nLdsEAAAAqroqHz4cDodcLpckKSsry/ZtAgAAAPAOl10BAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwRUBlV6Ci5J3OKfL/JU0DAAAAUHGqbfj4fcF06/+7PvpXJdYEAAAAgMRlVwAAAABsUq3OfDidTo0fP16SZIxRTs6ZS6sCAwPlcDhKXA4AAABAxapW4cPhcMjlclmvg4KCKrE2AAAAAPLjsisAAAAAtiB8AAAAALAF4QMAAACALQgfAAAAAGxB+AAAAABgC8IHAAAAAFsQPgAAAADYgvABAAAAwBaEDwAAAAC2IHwAAAAAsAXhAwAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAAAAtiB8AAAAALAF4QMAAACALQLKuqAxRpJ09OjRcqsMAAAAgKrHnQncGaE4ZQ4fx44dkyTVr1+/rKsAAAAAUI0cO3ZMERERxc53mNLiSTHy8vK0Z88ehYWFyeFwlLmC5eHo0aOqX7++/vjjD4WHh1dqXQDGI84ljEecSxiPOJcwHsuXMUbHjh1TTEyM/PyKv7OjzGc+/Pz8VK9evbIuXiHCw8MZPDhnMB5xLmE84lzCeMS5hPFYfko64+HGDecAAAAAbEH4AAAAAGCLahE+XC6XRo0aJZfLVdlVARiPOKcwHnEuYTziXMJ4rBxlvuEcAAAAAHxRLc58AAAAADj3ET4AAAAA2ILwAQAAAMAWhA8AAAAAtqjy4ePNN99Uo0aNFBQUpM6dO+vHH3+s7CqhGvj666917bXXKiYmRg6HQ/Pnz/eYb4zR888/r7p166pGjRqKj4/X1q1bPcocOnRIAwcOVHh4uCIjI3XXXXcpMzPTo8y6devUtWtXBQUFqX79+powYUJFNw1VzLhx43T55ZcrLCxMUVFRSkpK0pYtWzzKnDp1SkOHDtWFF16o0NBQ9e/fX/v37/co8/vvvysxMVHBwcGKiorSU089pdOnT3uUWb58uS677DK5XC41a9ZMM2bMqOjmoYqZOnWq2rdvb/1RttjYWC1atMiaz1hEZXr55ZflcDj06KOPWtMYk+cgU4XNnTvXOJ1OM23aNLNhwwZzzz33mMjISLN///7KrhqquM8//9w8++yz5qOPPjKSzMcff+wx/+WXXzYRERFm/vz5Zu3atea6664zjRs3NidPnrTK9OnTx1xyySXm+++/NytWrDDNmjUzt9xyizU/IyPDREdHm4EDB5r169ebOXPmmBo1api3337brmaiCkhISDDTp08369evN2lpaaZv376mQYMGJjMz0ypz//33m/r165slS5aYVatWmb/85S+mS5cu1vzTp0+bdu3amfj4eLNmzRrz+eefm1q1apkRI0ZYZXbs2GGCg4PN448/bjZu3GimTJli/P39zRdffGFre3FuW7BggVm4cKH59ddfzZYtW8wzzzxjAgMDzfr1640xjEVUnh9//NE0atTItG/f3jzyyCPWdMbkuadKh48rrrjCDB061Hqdm5trYmJizLhx4yqxVqhuCoaPvLw8U6dOHfPKK69Y044cOWJcLpeZM2eOMcaYjRs3Gknmp59+ssosWrTIOBwOs3v3bmOMMW+99ZapWbOmycrKssoMGzbMtGzZsoJbhKosPT3dSDJfffWVMebM2AsMDDQffPCBVWbTpk1Gkvnuu++MMWfCtJ+fn9m3b59VZurUqSY8PNwaf08//bRp27atx7Zuuukmk5CQUNFNQhVXs2ZN8+677zIWUWmOHTtmmjdvblJSUkz37t2t8MGYPDdV2cuusrOztXr1asXHx1vT/Pz8FB8fr++++64Sa4bqbufOndq3b5/H2IuIiFDnzp2tsffdd98pMjJSnTp1ssrEx8fLz89PP/zwg1WmW7ducjqdVpmEhARt2bJFhw8ftqk1qGoyMjIkSRdccIEkafXq1crJyfEYj61atVKDBg08xuPFF1+s6Ohoq0xCQoKOHj2qDRs2WGXyr8Ndhs9TFCc3N1dz587V8ePHFRsby1hEpRk6dKgSExMLjRvG5LkpoLIrUFZ//vmncnNzPQaLJEVHR2vz5s2VVCucD/bt2ydJRY4997x9+/YpKirKY35AQIAuuOACjzKNGzcutA73vJo1a1ZI/VF15eXl6dFHH9WVV16pdu3aSTozVpxOpyIjIz3KFhyPRY1X97ySyhw9elQnT55UjRo1KqJJqIJ++eUXxcbG6tSpUwoNDdXHH3+sNm3aKC0tjbEI282dO1c///yzfvrpp0Lz+Hw8N1XZ8AEA55uhQ4dq/fr1+uabbyq7KjiPtWzZUmlpacrIyNCHH36oQYMG6auvvqrsauE89Mcff+iRRx5RSkqKgoKCKrs68FKVveyqVq1a8vf3L/TEgv3796tOnTqVVCucD9zjq6SxV6dOHaWnp3vMP336tA4dOuRRpqh15N8G4Pbggw/qs88+07Jly1SvXj1rep06dZSdna0jR454lC84Hksba8WVCQ8P51s9eHA6nWrWrJk6duyocePG6ZJLLtFrr73GWITtVq9erfT0dF122WUKCAhQQECAvvrqK73++usKCAhQdHQ0Y/IcVGXDh9PpVMeOHbVkyRJrWl5enpYsWaLY2NhKrBmqu8aNG6tOnToeY+/o0aP64YcfrLEXGxurI0eOaPXq1VaZpUuXKi8vT507d7bKfP3118rJybHKpKSkqGXLllxyBYsxRg8++KA+/vhjLV26tNCleh07dlRgYKDHeNyyZYt+//13j/H4yy+/eATilJQUhYeHq02bNlaZ/Otwl+HzFKXJy8tTVlYWYxG2i4uL0y+//KK0tDTrp1OnTho4cKD1f8bkOaiy73g/G3PnzjUul8vMmDHDbNy40dx7770mMjLS44kFQFkcO3bMrFmzxqxZs8ZIMpMmTTJr1qwxv/32mzHmzKN2IyMjzSeffGLWrVtnrr/++iIftXvppZeaH374wXzzzTemefPmHo/aPXLkiImOjja33367Wb9+vZk7d64JDg7mUbvwMGTIEBMREWGWL19u9u7da/2cOHHCKnP//febBg0amKVLl5pVq1aZ2NhYExsba813P0qyd+/eJi0tzXzxxRemdu3aRT5K8qmnnjKbNm0yb775Jo+SRCHDhw83X331ldm5c6dZt26dGT58uHE4HGbx4sXGGMYiKl/+p10Zw5g8F1Xp8GGMMVOmTDENGjQwTqfTXHHFFeb777+v7CqhGli2bJmRVOhn0KBBxpgzj9sdOXKkiY6ONi6Xy8TFxZktW7Z4rOPgwYPmlltuMaGhoSY8PNzceeed5tixYx5l1q5da6666irjcrnMRRddZF5++WW7mogqoqhxKMlMnz7dKnPy5EnzwAMPmJo1a5rg4GDTr18/s3fvXo/17Nq1y1xzzTWmRo0aplatWuaJJ54wOTk5HmWWLVtmOnToYJxOp2nSpInHNgBjjBk8eLBp2LChcTqdpnbt2iYuLs4KHsYwFlH5CoYPxuS5x2GMMZVzzgUAAADA+aTK3vMBAAAAoGohfAAAAACwBeEDAAAAgC0IHwAAAABsQfgAAAAAYAvCBwAAAABbED4AAAAA2ILwAQAAAMAWhA8AAApwOByaP39+ZVcDAKodwgcAVIADBw5oyJAhatCggVwul+rUqaOEhAStXLmysqt2zjgXDvBHjx6tDh06VGodAOB8ElDZFQCA6qh///7Kzs7WzJkz1aRJE+3fv19LlizRwYMHK7tqAABUGs58AEA5O3LkiFasWKHx48erZ8+eatiwoa644gqNGDFC1113nUe5u+++W7Vr11Z4eLiuvvpqrV271mNdL7/8sqKjoxUWFqa77rpLw4cP9/imvkePHnr00Uc9lklKSlJycrL1OisrS08++aQuuugihYSEqHPnzlq+fLk1f8aMGYqMjNSXX36p1q1bKzQ0VH369NHevXs91jtt2jS1bdtWLpdLdevW1YMPPuhTW3z17rvvqnXr1goKClKrVq301ltvWfN27dolh8Ohjz76SD179lRwcLAuueQSfffddx7reOedd1S/fn0FBwerX79+mjRpkiIjI612jxkzRmvXrpXD4ZDD4dCMGTOsZf/880/169dPwcHBat68uRYsWHBW7QEAED4AoNyFhoYqNDRU8+fPV1ZWVrHlBgwYoPT0dC1atEirV6/WZZddpri4OB06dEiS9P7772v06NEaO3asVq1apbp163ocgHvrwQcf1Hfffae5c+dq3bp1GjBggPr06aOtW7daZU6cOKGJEydq1qxZ+vrrr/X777/rySeftOZPnTpVQ4cO1b333qtffvlFCxYsULNmzbxui6/ee+89Pf/88/r73/+uTZs2aezYsRo5cqRmzpzpUe7ZZ5/Vk08+qbS0NLVo0UK33HKLTp8+LUlauXKl7r//fj3yyCNKS0tTr1699Pe//91a9qabbtITTzyhtm3bau/evdq7d69uuukma/6YMWN04403at26derbt68GDhxY5vYAAP4/AwAodx9++KGpWbOmCQoKMl26dDEjRowwa9euteavWLHChIeHm1OnTnks17RpU/P2228bY4yJjY01DzzwgMf8zp07m0suucR63b17d/PII494lLn++uvNoEGDjDHG/Pbbb8bf39/s3r3bo0xcXJwZMWKEMcaY6dOnG0lm27Zt1vw333zTREdHW69jYmLMs88+W2RbvWlLUSSZjz/+uMh5TZs2NbNnz/aY9uKLL5rY2FhjjDE7d+40ksy7775rzd+wYYORZDZt2mSMMeamm24yiYmJHusYOHCgiYiIsF6PGjXKoz/z1+25556zXmdmZhpJZtGiRcW2BwBQOs58AEAF6N+/v/bs2aMFCxaoT58+Wr58uS677DLrsp61a9cqMzNTF154oXWmJDQ0VDt37tT27dslSZs2bVLnzp091hsbG+tTPX755Rfl5uaqRYsWHtv56quvrO1IUnBwsJo2bWq9rlu3rtLT0yVJ6enp2rNnj+Li4orchjdt8cXx48e1fft23XXXXR7re+mllwqtr3379h51dtdXkrZs2aIrrrjCo3zB1yXJv+6QkBCFh4db6wYAlA03nANABQkKClKvXr3Uq1cvjRw5UnfffbdGjRql5ORkZWZmqm7duh73Xri570nwhp+fn4wxHtNycnKs/2dmZsrf31+rV6+Wv7+/R7nQ0FDr/4GBgR7zHA6Htd4aNWqUWIfyakv+9Uln7tcoGL4KtiF/vR0OhyQpLy/P520Wpag+Ka91A8D5ivABADZp06aN9WjZyy67TPv27VNAQIAaNWpUZPnWrVvrhx9+0B133GFN+/777z3K1K5d2+PG8NzcXK1fv149e/aUJF166aXKzc1Venq6unbtWqZ6h4WFqVGjRlqyZIm13vy8aYsvoqOjFRMTox07dmjgwIFlXk/Lli31008/eUwr+NrpdCo3N7fM2wAA+IbwAQDl7ODBgxowYIAGDx6s9u3bKywsTKtWrdKECRN0/fXXS5Li4+MVGxurpKQkTZgwQS1atNCePXu0cOFC9evXT506ddIjjzyi5ORkderUSVdeeaXee+89bdiwQU2aNLG2dfXVV+vxxx/XwoUL1bRpU02aNElHjhyx5rdo0UIDBw7UHXfcoVdffVWXXnqpDhw4oCVLlqh9+/ZKTEz0qk2jR4/W/fffr6ioKF1zzTU6duyYVq5cqYceesirthRn586dSktL85jWvHlzjRkzRg8//LAiIiLUp08fZWVladWqVTp8+LAef/xxr+r80EMPqVu3bpo0aZKuvfZaLV26VIsWLbLOkEhSo0aNrDrUq1dPYWFhcrlcXq0fAFAGlX3TCQBUN6dOnTLDhw83l112mYmIiDDBwcGmZcuW5rnnnjMnTpywyh09etQ89NBDJiYmxgQGBpr69eubgQMHmt9//90q8/e//93UqlXLhIaGmkGDBpmnn37a4wbp7OxsM2TIEHPBBReYqKgoM27cOI8bzt1lnn/+edOoUSMTGBho6tata/r162fWrVtnjDlzw3n+m7CNMebjjz82BX9F/POf/zQtW7a01vHQQw/51JaCJBX5s2LFCmOMMe+9957p0KGDcTqdpmbNmqZbt27mo48+Msb83w3na9assdZ3+PBhI8ksW7bMmvavf/3LXHTRRaZGjRomKSnJvPTSS6ZOnToe71X//v1NZGSkkWSmT59u1a3gzfARERHWfABA2TiMKXCxMADgnDV69GjNnz+/0NkCeOeee+7R5s2btWLFisquCgCcl7jsCgBQbU2cOFG9evVSSEiIFi1apJkzZ5bpb6UAAMoH4QMAUG39+OOPmjBhgo4dO6YmTZro9ddf1913313Z1QKA8xaXXQEAAACwBX9kEAAAAIAtCB8AAAAAbEH4AAAAAGALwgcAAAAAWxA+AAAAANiC8AEAAADAFoQPAAAAALYgfAAAAACwxf8DhS/aZUnxTL8AAAAASUVORK5CYII=\n",
      "text/plain": "<Figure size 1000x400 with 1 Axes>"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Cell 3: COMBINE TRAIN + VALIDATION & BASIC EDA\n",
    "----------------------------------------------\n",
    "We concatenate the train and validation sets to maximize data. \n",
    "Then we do a quick EDA check on shapes, missingness, etc.\n",
    "\"\"\"\n",
    "\n",
    "# Combine sequences and labels\n",
    "trainval_sequences = pd.concat(\n",
    "    [train_sequences, validation_sequences], ignore_index=True\n",
    ")\n",
    "trainval_labels = pd.concat([train_labels, validation_labels], ignore_index=True)\n",
    "\n",
    "logging.info(\n",
    "    f\"Combined train+validation sequences: {trainval_sequences.shape}, labels: {trainval_labels.shape}\"\n",
    ")\n",
    "\n",
    "# Quick check for missing\n",
    "logging.info(\n",
    "    \"Missing in combined sequences:\\n\" + str(trainval_sequences.isnull().sum())\n",
    ")\n",
    "logging.info(\"Missing in combined labels:\\n\" + str(trainval_labels.isnull().sum()))\n",
    "\n",
    "# Example EDA: sequence length distribution\n",
    "trainval_sequences[\"sequence_length\"] = trainval_sequences[\"sequence\"].str.len()\n",
    "\n",
    "plt.figure(figsize=(10, 4))\n",
    "sns.boxplot(x=trainval_sequences[\"sequence_length\"], color=\"skyblue\")\n",
    "plt.title(\"Boxplot of Sequence Length (Train + Validation)\")\n",
    "plt.xlabel(\"Sequence Length\")\n",
    "plt.show()\n",
    "\n",
    "logging.info(\"Cell 3 complete: Basic EDA finished.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:17:53.263922Z",
     "iopub.status.busy": "2025-03-29T01:17:53.263707Z",
     "iopub.status.idle": "2025-03-29T01:17:53.594535Z",
     "shell.execute_reply": "2025-03-29T01:17:53.593907Z",
     "shell.execute_reply.started": "2025-03-29T01:17:53.263904Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 4: HANDLE MISSING COORDINATES & MERGE\n",
    "------------------------------------------\n",
    "We replace '-1e18' with np.nan, then merge sequences with labels on target_id.\n",
    "\"\"\"\n",
    "\n",
    "# Replace -1e18 with np.nan in the labels\n",
    "for col in [\"x_1\", \"y_1\", \"z_1\"]:\n",
    "    trainval_labels[col] = trainval_labels[col].replace(-1e18, np.nan)\n",
    "\n",
    "logging.info(\"Replaced -1e18 with NaN in trainval_labels for x_1, y_1, z_1.\")\n",
    "\n",
    "# Extract pdb_id, chain_id from ID\n",
    "trainval_labels[\"pdb_id\"] = trainval_labels[\"ID\"].apply(lambda x: x.split(\"_\")[0])\n",
    "trainval_labels[\"chain_id\"] = trainval_labels[\"ID\"].apply(lambda x: x.split(\"_\")[1])\n",
    "trainval_labels[\"target_id\"] = (\n",
    "    trainval_labels[\"pdb_id\"] + \"_\" + trainval_labels[\"chain_id\"]\n",
    ")\n",
    "\n",
    "# Merge\n",
    "train_data = pd.merge(trainval_labels, trainval_sequences, on=\"target_id\", how=\"left\")\n",
    "logging.info(f\"Merged train_data shape: {train_data.shape}\")\n",
    "\n",
    "# Quick check\n",
    "logging.info(\n",
    "    f\"Missing in x_1: {train_data['x_1'].isnull().sum()}, \"\n",
    "    f\"y_1: {train_data['y_1'].isnull().sum()}, \"\n",
    "    f\"z_1: {train_data['z_1'].isnull().sum()}\"\n",
    ")\n",
    "\n",
    "logging.info(\"Cell 4 complete: Merged train_data, ready for group-based imputation.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:17:53.595726Z",
     "iopub.status.busy": "2025-03-29T01:17:53.595492Z",
     "iopub.status.idle": "2025-03-29T01:18:26.167036Z",
     "shell.execute_reply": "2025-03-29T01:18:26.166387Z",
     "shell.execute_reply.started": "2025-03-29T01:17:53.595706Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 5: FEATURE ENGINEERING\n",
    "---------------------------\n",
    "Create numerical/categorical features from the 'sequence'.\n",
    "We'll keep 'resname' from the labels as a valuable feature.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def engineer_features(df):\n",
    "    \"\"\"\n",
    "    Create numerical & (some) categorical features from raw RNA sequence data.\n",
    "    \"\"\"\n",
    "    df = df.copy()\n",
    "    # Sequence-based\n",
    "    df[\"seq_length\"] = df[\"sequence\"].str.len()\n",
    "    df[\"A_cnt\"] = df[\"sequence\"].str.count(\"A\")\n",
    "    df[\"C_cnt\"] = df[\"sequence\"].str.count(\"C\")\n",
    "    df[\"G_cnt\"] = df[\"sequence\"].str.count(\"G\")\n",
    "    df[\"U_cnt\"] = df[\"sequence\"].str.count(\"U\")\n",
    "    df[\"begin_seq\"] = df[\"sequence\"].str[0]\n",
    "    df[\"end_seq\"] = df[\"sequence\"].str[-1]\n",
    "\n",
    "    # Di-nucleotide counts (example set)\n",
    "    for pair in [\n",
    "        \"AC\",\n",
    "        \"AG\",\n",
    "        \"AU\",\n",
    "        \"CA\",\n",
    "        \"CG\",\n",
    "        \"CU\",\n",
    "        \"GA\",\n",
    "        \"GC\",\n",
    "        \"GU\",\n",
    "        \"UA\",\n",
    "        \"UC\",\n",
    "        \"UG\",\n",
    "        \"AA\",\n",
    "        \"CC\",\n",
    "        \"GG\",\n",
    "        \"UU\",\n",
    "    ]:\n",
    "        df[f\"{pair}_cnt\"] = df[\"sequence\"].str.count(pair)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "# Apply feature engineering\n",
    "train_data = engineer_features(train_data)\n",
    "\n",
    "logging.info(\"Feature engineering applied to merged train_data.\")\n",
    "\n",
    "# We'll show an example of newly added columns\n",
    "example_cols = [\n",
    "    \"seq_length\",\n",
    "    \"A_cnt\",\n",
    "    \"C_cnt\",\n",
    "    \"G_cnt\",\n",
    "    \"U_cnt\",\n",
    "    \"begin_seq\",\n",
    "    \"end_seq\",\n",
    "    \"AC_cnt\",\n",
    "    \"AA_cnt\",\n",
    "]\n",
    "logging.info(f\"Columns after FE sample:\\n{train_data[example_cols].head(3)}\")\n",
    "\n",
    "logging.info(\"Cell 5 complete: Feature engineering done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:18:26.167997Z",
     "iopub.status.busy": "2025-03-29T01:18:26.167781Z",
     "iopub.status.idle": "2025-03-29T01:18:33.837453Z",
     "shell.execute_reply": "2025-03-29T01:18:33.836860Z",
     "shell.execute_reply.started": "2025-03-29T01:18:26.167977Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 6: GROUP-BASED IMPUTATION\n",
    "------------------------------\n",
    "We impute missing x_1, y_1, z_1 within each (target_id, resname) group.\n",
    "Finally, if any NAs remain, we fill them with a global median or drop them.\n",
    "\"\"\"\n",
    "\n",
    "# Perform group-based fill for x_1, y_1, z_1\n",
    "train_data[[\"x_1\", \"y_1\", \"z_1\"]] = (\n",
    "    train_data.groupby([\"target_id\", \"resname\"])[[\"x_1\", \"y_1\", \"z_1\"]]\n",
    "    .apply(lambda grp: grp.fillna(grp.mean()))\n",
    "    .reset_index(level=[\"target_id\", \"resname\"], drop=True)\n",
    ")\n",
    "\n",
    "# In case any remain after group-based mean fill (e.g. group is all NaN), do a global fill\n",
    "num_cols = [\"x_1\", \"y_1\", \"z_1\"]\n",
    "global_imputer = SimpleImputer(strategy=\"median\")\n",
    "train_data[num_cols] = global_imputer.fit_transform(train_data[num_cols])\n",
    "\n",
    "# If you'd prefer to drop any leftover NAs instead:\n",
    "# train_data.dropna(subset=['x_1','y_1','z_1'], inplace=True)\n",
    "\n",
    "logging.info(\"Group-based imputation + global median fallback complete.\")\n",
    "\n",
    "# Confirm missing values\n",
    "logging.info(\n",
    "    f\"Remaining missing x_1: {train_data['x_1'].isna().sum()}, \"\n",
    "    f\"y_1: {train_data['y_1'].isna().sum()}, z_1: {train_data['z_1'].isna().sum()}\"\n",
    ")\n",
    "\n",
    "logging.info(\"Cell 6 complete: Group-based imputation finished.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:18:33.838368Z",
     "iopub.status.busy": "2025-03-29T01:18:33.838176Z",
     "iopub.status.idle": "2025-03-29T01:18:34.286356Z",
     "shell.execute_reply": "2025-03-29T01:18:34.285740Z",
     "shell.execute_reply.started": "2025-03-29T01:18:33.838351Z"
    },
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 7: PREPARE DATA FOR MODELING\n",
    "---------------------------------\n",
    "We'll define the columns we won't use, set up X and y for x_1, y_1, z_1, \n",
    "and one-hot encode any relevant categorical columns (including resname).\n",
    "\"\"\"\n",
    "\n",
    "# Unused columns\n",
    "unused_cols = [\n",
    "    \"ID\",\n",
    "    \"pdb_id\",\n",
    "    \"chain_id\",\n",
    "    \"resid\",\n",
    "    \"x_1\",\n",
    "    \"y_1\",\n",
    "    \"z_1\",\n",
    "    \"sequence\",\n",
    "    \"description\",\n",
    "    \"temporal_cutoff\",\n",
    "    \"all_sequences\",\n",
    "    \"target_id\",  # key used for merges\n",
    "]\n",
    "\n",
    "# We'll keep resname, begin_seq, end_seq as features this time\n",
    "feature_cols = [col for col in train_data.columns if col not in unused_cols]\n",
    "\n",
    "# Make a copy\n",
    "train_df = train_data.copy()\n",
    "\n",
    "# Convert to categories\n",
    "for cat_col in [\"resname\", \"begin_seq\", \"end_seq\"]:\n",
    "    if cat_col in feature_cols:\n",
    "        train_df[cat_col] = train_df[cat_col].astype(\"category\")\n",
    "\n",
    "# One-hot encode\n",
    "train_df = pd.get_dummies(\n",
    "    train_df, columns=[\"resname\", \"begin_seq\", \"end_seq\"], drop_first=True\n",
    ")\n",
    "\n",
    "# Our final set of features\n",
    "X_cols = [col for col in train_df.columns if col not in unused_cols]\n",
    "\n",
    "X_full = train_df[X_cols]\n",
    "y_x_full = train_df[\"x_1\"]\n",
    "y_y_full = train_df[\"y_1\"]\n",
    "y_z_full = train_df[\"z_1\"]\n",
    "\n",
    "logging.info(f\"Feature matrix shape: {X_full.shape}\")\n",
    "logging.info(\"Cell 7 complete: Prepared data for modeling.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-03-29T01:18:34.287240Z",
     "iopub.status.busy": "2025-03-29T01:18:34.287041Z"
    },
    "trusted": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": "Fitting 3 folds for each of 5 candidates, totalling 15 fits\n"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": "/usr/local/lib/python3.10/dist-packages/xgboost/core.py:160: UserWarning: [01:18:59] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\nPotential solutions:\n- Use a data structure that matches the device ordinal in the booster.\n- Set the device for booster before call to inplace_predict.\n\nThis warning will only be shown once.\n\n  warnings.warn(smsg, UserWarning)\n"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Cell 8: KFold CV for X, Y, Z & Hyperparam Search\n",
    "------------------------------------------------\n",
    "We'll do a simplified KFold cross-validation for each coordinate \n",
    "to get a sense of good hyperparams, then train final models.\n",
    "\"\"\"\n",
    "\n",
    "import numpy as np\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "\n",
    "# Example hyperparameter grid (you can expand as needed)\n",
    "param_dist = {\n",
    "    \"learning_rate\": [0.03, 0.05, 0.1],\n",
    "    \"max_depth\": [6, 10, 15],\n",
    "    \"n_estimators\": [500, 800, 1000],\n",
    "    \"subsample\": [0.7, 0.9, 1.0],\n",
    "    \"colsample_bytree\": [0.7, 0.9, 1.0],\n",
    "}\n",
    "\n",
    "\n",
    "def run_random_search(X, y, param_dist, n_iter=5, cv_splits=3):\n",
    "    \"\"\"Simple RandomizedSearchCV for an XGBRegressor using GPU in XGBoost >= 2.0.\"\"\"\n",
    "    xgb = XGBRegressor(tree_method=\"hist\", device=\"cuda\", random_state=42)\n",
    "    rsearch = RandomizedSearchCV(\n",
    "        estimator=xgb,\n",
    "        param_distributions=param_dist,\n",
    "        n_iter=n_iter,\n",
    "        scoring=\"neg_mean_squared_error\",\n",
    "        cv=cv_splits,\n",
    "        verbose=1,\n",
    "        random_state=42,\n",
    "    )\n",
    "    rsearch.fit(X, y)\n",
    "    best_model = rsearch.best_estimator_\n",
    "    logging.info(\n",
    "        f\"Best params: {rsearch.best_params_}, Best CV Score: {rsearch.best_score_}\"\n",
    "    )\n",
    "    return best_model, rsearch.best_params_\n",
    "\n",
    "\n",
    "logging.info(\"Starting hyperparam search for X coordinate.\")\n",
    "best_model_x, best_params_x = run_random_search(\n",
    "    X_full, y_x_full, param_dist, n_iter=5, cv_splits=3\n",
    ")\n",
    "\n",
    "logging.info(\"Starting hyperparam search for Y coordinate.\")\n",
    "best_model_y, best_params_y = run_random_search(\n",
    "    X_full, y_y_full, param_dist, n_iter=5, cv_splits=3\n",
    ")\n",
    "\n",
    "logging.info(\"Starting hyperparam search for Z coordinate.\")\n",
    "best_model_z, best_params_z = run_random_search(\n",
    "    X_full, y_z_full, param_dist, n_iter=5, cv_splits=3\n",
    ")\n",
    "\n",
    "logging.info(\"Cell 8 complete: RandomizedSearchCV best params found.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 9: FINAL TRAINING ON FULL DATA\n",
    "-----------------------------------\n",
    "Use the best hyperparams for each coordinate found in CV. \n",
    "Retrain each coordinate model on all data (X_full, y_*_full).\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def get_best_xgb(params):\n",
    "    \"\"\"Return an XGBRegressor with the given params, using GPU.\"\"\"\n",
    "    # Here we override or add 'tree_method' to ensure GPU usage\n",
    "    # We can also specify predictor='gpu_predictor' to accelerate inference on GPU\n",
    "    model = XGBRegressor(\n",
    "        **params,\n",
    "        tree_method=\"hist\",  # or tree_method=params.get('tree_method', 'hist')\n",
    "        device=\"cuda\",  # ensures GPU usage\n",
    "        random_state=42,\n",
    "    )\n",
    "    return model\n",
    "\n",
    "\n",
    "logging.info(\"Retraining final model for X coordinate...\")\n",
    "model_x = get_best_xgb(best_params_x)\n",
    "model_x.fit(X_full, y_x_full)\n",
    "\n",
    "logging.info(\"Retraining final model for Y coordinate...\")\n",
    "model_y = get_best_xgb(best_params_y)\n",
    "model_y.fit(X_full, y_y_full)\n",
    "\n",
    "logging.info(\"Retraining final model for Z coordinate...\")\n",
    "model_z = get_best_xgb(best_params_z)\n",
    "model_z.fit(X_full, y_z_full)\n",
    "\n",
    "logging.info(\"Cell 9 complete: Final models trained.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 10: PREPARE & ENGINEER TEST DATA\n",
    "-------------------------------------\n",
    "We expand test_sequences into (ID, resname, resid), merge with engineered features, \n",
    "and align columns with training's X_full. \n",
    "\"\"\"\n",
    "\n",
    "# Expand test data row-by-row\n",
    "test_expanded = []\n",
    "for idx, row in test_sequences.iterrows():\n",
    "    sequence_id = row[\"target_id\"]\n",
    "    seq_str = row[\"sequence\"]\n",
    "    for i, nucleotide in enumerate(seq_str, start=1):\n",
    "        test_expanded.append([sequence_id, nucleotide, i])\n",
    "\n",
    "test_clean_df = pd.DataFrame(test_expanded, columns=[\"ID\", \"resname\", \"resid\"])\n",
    "logging.info(f\"test_clean_df shape: {test_clean_df.shape} (expanded test sequences)\")\n",
    "\n",
    "# We'll rename test_sequences -> test_feats for clarity, and engineer features\n",
    "test_feats = engineer_features(test_sequences)\n",
    "\n",
    "# FIX: use 'seq_length' instead of 'sequence_length' in drop() or remove the drop entirely\n",
    "test_merged = pd.merge(\n",
    "    test_clean_df,\n",
    "    # drop 'seq_length' if you don't need it in final test features\n",
    "    test_feats.drop(columns=[\"seq_length\"]),\n",
    "    left_on=\"ID\",\n",
    "    right_on=\"target_id\",\n",
    "    how=\"left\",\n",
    ")\n",
    "\n",
    "logging.info(f\"test_merged shape after merging: {test_merged.shape}\")\n",
    "\n",
    "# Convert any -1e18 to np.nan if present\n",
    "for col in [\"x_1\", \"y_1\", \"z_1\"]:\n",
    "    if col in test_merged.columns:\n",
    "        test_merged[col] = test_merged[col].replace(-1e18, np.nan)\n",
    "\n",
    "# Drop unneeded columns if present\n",
    "drop_cols = [\"sequence\", \"description\", \"temporal_cutoff\", \"all_sequences\", \"target_id\"]\n",
    "test_merged = test_merged.drop(\n",
    "    columns=[c for c in drop_cols if c in test_merged.columns], errors=\"ignore\"\n",
    ")\n",
    "\n",
    "logging.info(\"Now we align test_merged columns with training X.\")\n",
    "\n",
    "# Convert to category for 'resname','begin_seq','end_seq'\n",
    "cat_cols = set([\"resname\", \"begin_seq\", \"end_seq\"]) & set(test_merged.columns)\n",
    "for cat_col in cat_cols:\n",
    "    test_merged[cat_col] = test_merged[cat_col].astype(\"category\")\n",
    "\n",
    "# One-hot encode\n",
    "test_merged = pd.get_dummies(test_merged, columns=cat_cols, drop_first=True)\n",
    "\n",
    "# Add missing columns from X_full\n",
    "missing_cols = set(X_full.columns) - set(test_merged.columns)\n",
    "for c in missing_cols:\n",
    "    test_merged[c] = 0\n",
    "\n",
    "# Drop any extra columns not in X_full\n",
    "extra_cols = set(test_merged.columns) - set(X_full.columns)\n",
    "if extra_cols:\n",
    "    test_merged.drop(columns=extra_cols, inplace=True)\n",
    "\n",
    "# Reorder columns\n",
    "test_merged = test_merged[X_full.columns]\n",
    "\n",
    "# Fill final missing numeric features with global_imputer\n",
    "test_merged_imputed = pd.DataFrame(\n",
    "    global_imputer.transform(test_merged), columns=test_merged.columns\n",
    ")\n",
    "\n",
    "logging.info(\"Cell 10 complete: Test data prepared & aligned.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 11: GENERATE PREDICTIONS & BUILD SUBMISSION\n",
    "------------------------------------------------\n",
    "We'll predict (x_1, y_1, z_1) for each residue, \n",
    "then replicate those coordinates for structures x_2..z_5.\n",
    "Finally, we'll align with sample_submission and save submission.csv.\n",
    "\"\"\"\n",
    "\n",
    "# Predict x_1, y_1, z_1\n",
    "test_pred_x = model_x.predict(test_merged_imputed)\n",
    "test_pred_y = model_y.predict(test_merged_imputed)\n",
    "test_pred_z = model_z.predict(test_merged_imputed)\n",
    "\n",
    "# Build submission from test_clean_df\n",
    "submission = test_clean_df.copy()\n",
    "\n",
    "# Add predicted coords for structure 1\n",
    "submission[\"x_1\"] = test_pred_x\n",
    "submission[\"y_1\"] = test_pred_y\n",
    "submission[\"z_1\"] = test_pred_z\n",
    "\n",
    "# For simplicity, replicate for structures 2..5\n",
    "for i in [2, 3, 4, 5]:\n",
    "    submission[f\"x_{i}\"] = test_pred_x\n",
    "    submission[f\"y_{i}\"] = test_pred_y\n",
    "    submission[f\"z_{i}\"] = test_pred_z\n",
    "\n",
    "# Adjust ID format: ID + \"_\" + resid\n",
    "submission[\"ID\"] = submission[\"ID\"] + \"_\" + submission[\"resid\"].astype(str)\n",
    "\n",
    "# Reorder columns to match sample_submission\n",
    "final_cols = list(sample_submission.columns)  # ID, resname, resid, x_1..z_5\n",
    "submission = submission[\n",
    "    [\n",
    "        \"ID\",\n",
    "        \"resname\",\n",
    "        \"resid\",\n",
    "        \"x_1\",\n",
    "        \"y_1\",\n",
    "        \"z_1\",\n",
    "        \"x_2\",\n",
    "        \"y_2\",\n",
    "        \"z_2\",\n",
    "        \"x_3\",\n",
    "        \"y_3\",\n",
    "        \"z_3\",\n",
    "        \"x_4\",\n",
    "        \"y_4\",\n",
    "        \"z_4\",\n",
    "        \"x_5\",\n",
    "        \"y_5\",\n",
    "        \"z_5\",\n",
    "    ]\n",
    "]\n",
    "\n",
    "# Merge with sample_submission to match row order\n",
    "sample_submission[\"sort_order\"] = range(len(sample_submission))\n",
    "submission_merged = (\n",
    "    pd.merge(submission, sample_submission[[\"ID\", \"sort_order\"]], on=\"ID\", how=\"left\")\n",
    "    .sort_values(\"sort_order\")\n",
    "    .drop(columns=\"sort_order\")\n",
    ")\n",
    "\n",
    "# This is our final submission dataframe\n",
    "submission_df = submission_merged.copy()\n",
    "\n",
    "# Save to CSV\n",
    "submission_df.to_csv(\"submission.csv\", index=False)\n",
    "logging.info(\"submission.csv created successfully.\")\n",
    "\n",
    "print(\"Cell 11 complete: Submission file saved. Ready to submit!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "trusted": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Cell 12: CONCLUSIONS & NEXT STEPS\n",
    "---------------------------------\n",
    "We've done:\n",
    "- Group-based imputation\n",
    "- Preserved resname\n",
    "- Hyperparameter tuning via RandomizedSearchCV\n",
    "- Final training on full combined data\n",
    "- Test predictions with the same coordinate repeated across 5 structures\n",
    "\n",
    "Suggestions for further improvement:\n",
    "- Fine-tune hyperparameters with a broader search or Bayesian optimization\n",
    "- Explore more advanced RNA 3D features\n",
    "- Generate truly distinct 5 structures instead of repeating the same coordinates\n",
    "\"\"\"\n",
    "logging.info(\"Notebook complete. Good luck on the leaderboard!\")\n",
    "print(\"All done! Submit 'submission.csv' to the competition.\")"
   ]
  }
 ],
 "metadata": {
  "kaggle": {
   "accelerator": "gpu",
   "dataSources": [
    {
     "databundleVersionId": 11403143,
     "sourceId": 87793,
     "sourceType": "competition"
    },
    {
     "datasetId": 6866703,
     "sourceId": 11026565,
     "sourceType": "datasetVersion"
    },
    {
     "datasetId": 6866398,
     "sourceId": 11026927,
     "sourceType": "datasetVersion"
    }
   ],
   "dockerImageVersionId": 30918,
   "isGpuEnabled": true,
   "isInternetEnabled": true,
   "language": "python",
   "sourceType": "notebook"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}