
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../Stage_B/">
      
      
        <link rel="next" href="../torsionbert_code/">
      
      
        
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Overview - rna_predict</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#b1-torsionbert" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="rna_predict" class="md-header__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            rna_predict
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Overview
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="rna_predict" class="md-nav__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    rna_predict
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Guides
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Guides
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" >
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Getting Started
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Getting Started
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/windows_compatibility_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Windows Compatibility
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Best Practices
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/code_quality_best_practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code Quality
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2_2" id="__nav_2_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_coverage_strategy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage Strategy
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/progressive_coverage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progressive Coverage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_generation_prompt/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Test Generation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_2_3" id="__nav_2_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Debugging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Debugging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/debugging/comprehensive_debugging_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Guide
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Pipeline
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Overview
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/core_framework/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Core Framework
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Multi_Stage_Implementation_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Multi-Stage Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Hydra Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Hydra Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_gap_analysis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Gap Analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_master_document/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Master Document
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3_4" id="__nav_3_2_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Components
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_2_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Components
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageA/StageA_2D_Adjacency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageB/StageB_Torsion_Pairwise/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageC/StageC_3D_Reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageD/StageD_Diffusion_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage D
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/unified_latent/UnifiedLatentMerger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage A
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/StageA_RFold/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_code/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Code
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/rfold-demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/stage_a_extra/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Extra Notes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" checked>
        
          
          <label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage B
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Stage_B/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_3_4_2" id="__nav_3_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    TorsionBERT
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    TorsionBERT
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../torsionbert_code/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../torsionBert_full_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" >
        
          
          <label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage C
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Stage_C/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Integrated_RNA_Geometry_and_3D_Reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Geometry & Reconstruction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Unified, Comprehensive Plan for Integrating MP-NeRF into Stage C.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Integration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/mp_nerf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Details
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_6" >
        
          
          <label class="md-nav__link" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Unified Latent
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unified_latent/Perceiver_IO/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Perceiver IO
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Energy_Minimization_%26_Molecular_Dynamics_%28MD%29_for_RNA_Structure_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Energy & MD
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_8" >
        
          
          <label class="md-nav__link" for="__nav_3_8" id="__nav_3_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Kaggle
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Kaggle
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/kaggle_competition/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Competition
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/M2_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    M2 Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Advanced Methods
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Advanced Methods
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    AlphaFold 3
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    AlphaFold 3
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AF3_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AlphaFold3_progress/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progress
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/Pairwise_Distance_Based_Prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Distance Prediction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Diffusion
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Diffusion
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/s4_diffusion/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    S4 Diffusion
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/isosteric_substitutions/RNA_isostericity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Isosteric Substitutions
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    External Literature
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    External Literature
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/2d_structure_prediction_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2D Structure
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/RNA_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA Papers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Residue Atom Bridging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Residue Atom Bridging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/audit_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Audit Report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/design_spec/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design Spec
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/documentation_draft/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Documentation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/implementation_notes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Implementation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/refactoring_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Refactoring
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/bridging_caveats_guidelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Guidelines
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Torsion Calculations
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Torsion Calculations
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angles_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Standard_Bond_Lengths_and_Angles_in_RNA_Nucleotides/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Bond Lengths & Angles
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Torsion_Angles_3D_Coordinates_RNA_Structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    3D Coordinates
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angle_Latent_Manifold_Representation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Latent Manifold
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<p>: Below is a systematic comparecontrast and assessment of each repository (THEIRS and OURS), focusing on strengths, weaknesses, creativity, depth, ease of implementation, and performance. Afterward, I summarize actionable takeaways for anyone deciding which approach (or hybrid) to adopt.</p>
<p>[emoji]: For clarity, I will designate the code from THEIRS (the rna_torsionBERT/ code directory that uses PyTorch, huggingface transformers, and MCQ metrics) as RNA-TorsionBERT. The code from OURS (the docs/ plus rna_predict/ pipeline with multi-stage approaches and alphafold-style local attention) as RNA_PREDICT Pipeline.</p>
<p></p>
<ol>
<li>High-Level Purposes &amp; Scopes</li>
</ol>
<p>THEIRS (RNA-TorsionBERT)
       Scope: A specialized library for predicting RNA torsion angles (and pseudo-torsion angles) directly from sequence using a BERT-based model.
       Goal: Provide a pre-trained or fine-tunable model (a language model for RNA torsions), plus a scoring function (TB-MCQ) to evaluate predicted angles.</p>
<p>Strengths
    1.  Clear Focus: Very specific to angle prediction from sequence.
    2.  Machine-Learning Depth: Incorporates advanced huggingface transformers, sin/cos outputs for torsions, and integrated scoring with MCQ.
    3.  Performance: Empirical tests (MAE on angles) show strong results on real PDB structures.</p>
<p>Weaknesses
    1.  Limited to Angles: Not a complete pipeline for 3D structure, beyond the angles (no 3D builder except partial re-conversion, no 2D structure integration).
    2.  Harder to Extend: If you want to incorporate MSA, base-pair adjacency, or advanced geometry features, thats mostly outside the TorsionBERT scope.
    3.  Data Requirements: TorsionBERT is trained on known structures, so one must have curated data or use the pre-trained weights.</p>
<p></p>
<p>OURS (RNA_PREDICT Pipeline)
       Scope: A broader multi-stage approach:
    1.  StageA: 2D structure from sequence (via external or built-in folding methods).
    2.  StageB: Torsion angles from adjacency + advanced GNN or Transformers.
    3.  StageC: Forward-kinematics to build 3D from angles.
    4.  (Optional) Additional AF3-inspired trunk &amp; diffusion refinements.
       Goal: Provide a modular pipeline that can transform an RNA sequence  2D structure  torsion angles  full 3D coordinates, with room for advanced geometric or diffusion-based refinement.</p>
<p>Strengths
    1.  End-to-End Pipeline: Goes from raw sequence (with or without external 2D folding) all the way to 3D cartesian output.
    2.  Creativity &amp; Extensibility: Encourages a mix of GNN, local attention, and modular refinement. Clear synergy with advanced methods (Liquid-S4 or AF3-like).
    3.  Modular Stages: Each stage can be replaced (e.g., StageA with different 2D folding algorithms, StageB with different torsion-angle nets).</p>
<p>Weaknesses
    1.  Less Off-the-Shelf: Its not a single integrated model you can just pip install and call for predict angles. Instead, its a set of documented modules that must be orchestrated.
    2.  Performance Still Unclear: The repositorys code includes partial benchmarks or placeholders, but a robust comparison to existing SOTA 3D methods (like Rosetta or TorsionBERT) is not fully documented yet.
    3.  Implementation Overhead: Because it covers multiple stages (2D, angles, 3D), a user must gather more external dependencies or data to run the entire pipeline effectively.</p>
<p></p>
<ol>
<li>Code &amp; Directory Structure</li>
</ol>
<p>Aspect  THEIRS (RNA-TorsionBERT)    OURS (RNA_PREDICT Pipeline)
Directory Layout    - Minimalistic src/ with enums, helper, metrics, CLIs, plus data/ and requirements.txt.- Focused on angles &amp; TB-MCQ scoring.  - Larger, more design docheavy approach: docs/advanced_methods/, docs/pipeline/, rna_predict/ with subfolders for models/, scripts/, utils/.
Modularity  - Single-level src/* structure: rna_torsionBERT_helper, mcq.py, extractor, etc. - Each script is specialized for angles &amp; MCQ.    - Multi-stage design with separate stageA, stageB, stageC modules - Additional advanced diffusion/s4_diffusion references, test scripts, utilities.
Ease of Navigation  - Straightforward: one top-level src folder, each file is fairly small.  - Good for focusing on torsion angle inference.    - More folders: it might feel more scattered, but also thoroughly documented in docs/. - Developer can find extensive textual guides for each stage.
Documentation   - README + docstrings, mentions usage python -m src.rna_torsionBERT_cli  - Some references to MCQ.   - Very large textual docs (like Stage_C.md, core_framework.md, etc.). - Each pipeline aspect is explained with pseudo-code and rationale.
Creative Extras - Dockerfile, small example CLI usage (run TorsionBERT, run TB-MCQ).    - Analysis of Liquid-S4, advanced alphaFold3, isostericity design docs - Potential synergy for not just angle prediction but entire 3D structure pipeline.</p>
<p></p>
<ol>
<li>Installation &amp; Ease of Implementation</li>
</ol>
<p>RNA-TorsionBERT
       Installation:
       pip install -r requirements.txt and done.
       Mostly standard Python libraries + huggingface Transformers + PyTorch + biopython.
       Ease:
       If you simply want Predict angles from an RNA sequence, its very direct: python -m src.rna_torsionBERT_cli --in_seq GGGAAAUCC  etc.
       tb_mcq_cli.py also provides a convenient command to score predicted angles from a .pdb.</p>
<p>RNA_PREDICT Pipeline
       Installation:
       Has a setup.py, but also depends on external tools for 2D folding (ViennaRNA or RFold), and for 3D analysis (MDAnalysis).
       The docs/ mention possible Docker usage but not quite as succinct.
       Ease:
       Because its a multi-stage approach, you must piece everything together: Stage A  Stage B  Stage C  (Optional Stage D).
       The user can do step-by-step or partial usage (just the Stage B torsion predictor, etc.).</p>
<p>Verdict:
       If you want a single-step pipeline to get angles from sequence, TorsionBERT is simpler.
       If you want an entire workflow from 2D adjacency to final 3D, RNA_PREDICT covers more ground but is more involved to set up.</p>
<p></p>
<ol>
<li>Algorithmic Depth &amp; Creativity</li>
</ol>
<p>Aspect  THEIRS (RNA-TorsionBERT)    OURS (RNA_PREDICT Pipeline)
Core Method - BERT-based language model approach, uses tokenization to predict each angles (sin, cos).- Integrates MCQ to measure angle error.   - Multi-step GNN/transformer approach for adjacency-based angle prediction - Potential to incorporate advanced AF3 trunk or diffusion-based refinement.
Breadth - Specifically addresses angle prediction and a TB-MCQ scoring - Not trying to handle 2D or final 3D coords.  - Broader design with references to sugar pucker, forward kinematics, optional local minimization. - Full pipeline from sequence to 3D.
Creativity  - Innovative usage of DNABERT-like embedding adapted for RNA angles. - Crisp integration of pLDDT-like heads for angles.    - Combines multiple approaches (2D adjacency, local S4 diffusion, isosteric design) - Encourages a very flexible the user can pick a method for each stage.
Potential   - Could be extended to consider 3D if integrated with a geometry module. - Possibly adapt the TB-MCQ for advanced metrics.  - Ready for all sorts of expansions: isosteric substitutions, AlphaFold3 trunk, etc. - Potentially complicated for a single user wanting angles alone.</p>
<p></p>
<ol>
<li>Performance &amp; Benchmarking</li>
</ol>
<p>THEIRS (RNA-TorsionBERT)
       Angle MAE: They demonstrate systematic comparisons on multiple PDB sets (Test sets, e.g. RNA-Puzzles, CASP-RNA). Achieves strong angle accuracy.
       TB-MCQ: They show how well the final angles match MCQ metrics.
       Inference Speed: BERT inference is fairly quick on a GPU (time depends on sequence length).</p>
<p>OURS (RNA_PREDICT Pipeline)
       Performance: The code includes benchmark.py for timing local block-sparse attention, plus partial references to memory usage.
       No explicit angle MAE: The pipelines docs mention the possibility to test or compare but do not show final numeric results on standard sets.
       Large-scale: The pipeline can scale, but the actual speed might be slower or faster depending on the GNN approach. The doc references local block-sparse attention for efficiency.</p>
<p>Summary:
       TorsionBERT has more published/explicit performance metrics for angle errors.
       RNA_PREDICT is broader but less fully validated in the docs. The user might need to do extra steps to see final performance on standard benchmarks (like RMSD or MCQ).</p>
<p></p>
<ol>
<li>Strengths vs. Weaknesses Overview</li>
</ol>
<p>(A) THEIRS: RNA-TorsionBERT
       Strengths
    1.  Straight-to-the-point solution for angle prediction from sequence.
    2.  Well-defined huggingface framework; easy to fine-tune or do inference.
    3.  Good docs for installation and usage (CLI scripts for rna_torsionBERT_cli.py, tb_mcq_cli.py).
       Weaknesses
    1.  Limited to anglesno direct 3D building or advanced structural constraints.
    2.  Less modular pipeline. Harder to integrate if you want to incorporate a custom 2D adjacency or ring-closure logic.
    3.  Some parts revolve around that specific model architecture; less pluggable with other GNN or advanced alphaFold-like modules.</p>
<p>(B) OURS: RNA_PREDICT
       Strengths
    1.  End-to-end pipeline from sequence  2D  angles  3D, plus potential for advanced trunk or diffusion.
    2.  Highly modular, multiple documented stages (A/B/C) for easy partial usage or extension.
    3.  Rich design docs, referencing next-gen ideas (Liquid-S4, alphaFold3 trunk, isosteric substitution, etc.).
       Weaknesses
    1.  More complex to set up and orchestrate. Possibly overkill if you only need angles.
    2.  Real performance/accuracy on final 3D is not pre-benchmarked in a standardized manner.
    3.  Dependent on external folding tools (StageA) and external library for partial 3D checks, so not as turnkey for a single purpose.</p>
<p></p>
<ol>
<li>Which To Choose? Or Combine?</li>
</ol>
<p>Choose RNA-TorsionBERT (THEIRS) if:
       You specifically want RNA torsion angles from the sequence in a single shot, with minimal fuss.
       You like huggingface/transformers pipeline or want to adapt a BERT-based approach.
       You want TB-MCQ scoring of angles.</p>
<p>Choose RNA_PREDICT (OURS) if:
       You need a full pipeline from sequence to final 3D coords (or you want to integrate your own 2D  torsion steps).
       You want the option to experiment with GNN, local block-sparse attention, or advanced alphaFold3-like diffusion modules.
       You prefer a more research-lab style codebase with rich docs for potential expansions (like isosteric design, sugar pucker modeling).</p>
<p>Combine or Hybrid:
    1.  Potentially use TorsionBERT as the StageB in the OURS pipeline:
       StageA = 2D adjacency from external folder.
       StageB (torsion) = TorsionBERT.
       Then feed those angles into StageC forward-kinematics.
    2.  Or after TorsionBERT outputs angles, you run the final 3D building from the OURS forward-kinematics code.</p>
<p></p>
<ol>
<li>Concluding Assessment</li>
</ol>
<p>(a) Creativity &amp; Depth
       TorsionBERT: Creative usage of a BERT-based masked approach for predicting angles. Very neat for language-model lovers.
       RNA_PREDICT: Provides a deeper end-to-end pipeline with more potential for experimental or advanced methods (diffusion, s4, etc.).</p>
<p>(b) Ease of Implementation
       TorsionBERT: Easiest for predict angles from sequence. They have a single script and pretrained weights.
       RNA_PREDICT: More elaborate. Possibly tricky for new users but extremely flexible for bigger RNA 3D projects.</p>
<p>(c) Performance
       TorsionBERT: Published angle MAE results, using the MCQ and real test sets.
       RNA_PREDICT: Has block-sparse benchmarks for attention, but the final 3D performance is still up to the user to measure.</p>
<p>(d) Future Extensions
       TorsionBERT: Could eventually incorporate a 3D building step or a trunk module.
       RNA_PREDICT: Already built for expansions (pairformer, confidence heads, isosteric design, local geometry checks).</p>
<p></p>
<p>Final Takeaways
    1.  If your immediate need is a well-tested, single-step angle predictor with a straightforward CLI, RNA-TorsionBERT out-of-the-box is the simpler solution.
    2.  If you want a thorough pipeline that can incorporate 2D folding info, do angle predictions (from your own or external model), and build 3D coordinates with optional advanced refinements, RNA_PREDICT is the more comprehensive framework.
    3.  Hybrid usage is feasible: run TorsionBERT for angles, then feed them into StageC of RNA_PREDICT for 3D reconstruction.</p>
<p>In short, RNA-TorsionBERT is the better choice for specialized angle predictions, while RNA_PREDICT is a creative, more extensive pipeline for full 3D builds and advanced custom expansions. Both projects excel in different ways and can even be combined if desired.</p>
<p>: Great question. RNA-TorsionBERT (the THEIRS approach) was originally built to predict torsion angles just from sequence, whereas your StageA output (an adjacency or basepair matrix) is not something TorsionBERT naturally expects. Below are a few ways to incorporate StageAs 2D data if you still want to use TorsionBERT in StageB.</p>
<p>[emoji]: Ill outline four main approaches, from simplest (ignore adjacency) to more advanced (modify TorsionBERTs model to fuse adjacency). Choose whichever best fits your timeframe and coding constraints.</p>
<p></p>
<p>1) Ignore StageA adjacency inside TorsionBERT (simple fallback)</p>
<p>Idea
    1.  StageA still produces adjacency (2D structure), but TorsionBERT doesnt use it.
    2.  TorsionBERT just takes the raw sequence, outputs torsion angles.
    3.  You pass these angles to StageC for forward kinematics.</p>
<p>Pros &amp; Cons
       Pros:
       Easiest to implement. TorsionBERT remains untouchedjust feed it the sequence.
       You still have adjacency from StageA in case you need it for later validations or a small post-processing step.
       Cons:
       You lose any advantage that adjacency might bring. TorsionBERTs predictions wont directly reflect the 2D structure from StageA.
       Might give you less accurate angles if the base-pair info was important.</p>
<p></p>
<p>2) Use TorsionBERT as is, then refine with adjacency in a small GNN step</p>
<p>Idea
    1.  StageA outputs adjacency.
    2.  StageB is TorsionBERT: it predicts torsion angles from sequence alone.
    3.  StageB (an extra mini-stage) takes TorsionBERTs angles + adjacency to do a short correction or refinement.
       For instance, you can design a small GNN/regressor that sees the adjacency + TorsionBERT angles + sequence to produce slightly adjusted angles.</p>
<p>Sketch</p>
<h1 id="b1-torsionbert">B1: TorsionBERT</h1>
<p>angles_raw = torsionBertModel.predict(sequence)</p>
<h1 id="b2-gnn-refiner">B2: GNN Refiner</h1>
<p>angles_refined = smallGNNrefiner(adjacency, angles_raw, sequence)</p>
<h1 id="then-stage-c-forward_kinematicsangles_refined-3d">Then Stage C: forward_kinematics(angles_refined) -&gt; 3D</h1>
<p>Pros &amp; Cons
       Pros:
       Minimal changes to TorsionBERT itself.
       You can incorporate adjacency as a post-processing refiner.
       Cons:
       Requires training a new GNN or MLP that sits on top of TorsionBERT output.
       More steps = more complexity.</p>
<p></p>
<p>3) Fine-tune TorsionBERT with adjacency as an auxiliary embedding (medium complexity)</p>
<p>Idea</p>
<p>If you have source code or the model architecture for TorsionBERT, you can extend its input to incorporate adjacency or base-pair probability in addition to the raw sequence tokens. For instance:
    1.  Modify TorsionBERTs tokenizer/embedding step.
       Normally it just sees k-mer tokens from the sequence.
       You could pass an extra channel (like a 2D feature) for each token i, summarizing how i is paired or not.
    2.  The models first layers might combine (token embedding + adjacency hints)  produce the same final angle outputs.</p>
<p>Practical Steps
       Concatenate adjacency features to each residues token embedding. For example, you can embed # of pairs for residue i, or some local base-pair feature, or the sum of adjacency row i, etc.
       Retrain or fine-tune TorsionBERT so it learns sequence + adjacency.</p>
<p>Pros &amp; Cons
       Pros:
       TorsionBERT can then directly leverage StageA data.
       A single model, no post-refinement.
       Cons:
       TorsionBERTs code might need moderate refactoring.
       Must re-train or at least heavily fine-tune with adjacency-labeled training data (which you must gather).</p>
<p></p>
<p>4) Rewrite TorsionBERT to fully integrate adjacency (most advanced)</p>
<p>Idea</p>
<p>Take TorsionBERTs concept (masking angles, sin/cos predictions, BERT-like architecture) but replace or enhance its attention layers with adjacency-based attention or gating. This is more akin to a Graph-BERT design.
       Youd embed each residues token the usual way, but in the attention layers, incorporate adjacency-based attention masks or pairwise bias (similar to how some alphaFold modules do).</p>
<p>Pros &amp; Cons
       Pros:
       Possibly best synergy: TorsionBERT language model approach + direct adjacency constraints.
       Cons:
       Major engineering joblike building a new model that merges TorsionBERTs masked-language concept with a GNN adjacency approach.
       Then youd have to re-train from scratch or from TorsionBERTs weights, which is not trivial.</p>
<p></p>
<p>Which Approach to Pick?
    1.  You have little time to integrate adjacency  Option #1 (ignore adjacency in TorsionBERT) is the simplest.
    2.  You can do a small neural refiner easily  Option #2 is a good compromise. Let TorsionBERT do the heavy lifting, then a GNN or MLP adjusts angles to be consistent with StageAs adjacency.
    3.  You want a single final model but can handle some TorsionBERT modifications  Option #3. Youd add adjacency features as an extra input channel, then re-train or fine-tune.
    4.  Youre comfortable rewriting or thoroughly customizing TorsionBERT  Option #4 is potentially the best, but also the largest time investment.</p>
<p></p>
<p>Summary</p>
<p>TorsionBERT wasnt originally designed to take adjacency from StageA, so theres no built-in hook to incorporate that 2D structure. To do it anyway, you must:
       Either (A) keep TorsionBERT as is (and skip adjacency or do a post-processor), or
       (B) adapt TorsionBERTs input layers or attention layers to handle adjacency.</p>
<p>Hence: If youre short on time, you can simply use TorsionBERT as StageB ignoring adjacencybut you lose the synergy from StageA. For a more thorough integration, youd need to do some form of post-processing or direct re-training that merges adjacency into TorsionBERTs pipeline.</p>
<p>[emoji]: Thats it. The take-home message is: TorsionBERT doesnt natively incorporate your StageA adjacency, so you either do a second refinement step or fork TorsionBERTs architecture to fuse adjacency. Both are feasible, but they require a bit of custom engineering.
: In terms of pure synergy and best possible results (i.e., fully leveraging StageA adjacency inside a powerful language-like model for torsion angles), the most direct route is to rewrite or extend TorsionBERT so that it natively handles adjacency (Option #4 in the previous list). However, thats also the most laborintensive. Depending on timelines and willingness to refactor code, you may prefer Option #3 (finetune TorsionBERT with some adjacencybased embeddings) as a still-powerful compromise.</p>
<p>Below is a stepbystep reasoning for which approach yields best synergy with StageABC, factoring in your GNN/geometry data, your anglediffusion design, and your Kaggle timeline.</p>
<p></p>
<p>1) Quick Recap of the Options
    1.  (Option #1) Use TorsionBERT as is, ignoring adjacency.
    2.  (Option #2) TorsionBERT + a small adjacency-based refiner network.
    3.  (Option #3) Fine-tune TorsionBERT with adjacency as an extra input channel.
    4.  (Option #4) Rewrite TorsionBERT to incorporate adjacency in its internal architecture (attention layers).</p>
<p></p>
<p>2) Criteria if We Want The Best Integration in a Larger Pipeline
    1.  Exploiting StageA adjacency: If your StageA is itself carefully curated (like a highconfidence adjacency from RFold or a GNN), we want a StageB that definitely uses that adjacency to produce angles consistent with basepairs.
    2.  AngleBased Diffusion: If you do an AF3style diffusion (StageD) after StageB, it helps if StageBs angles are already consistent with basepairs; you wont have to re-fix them in diffusion.
    3.  Extensibility: Possibly youll add MSA or 2.5D paired row attention. If TorsionBERT is a black box (Option #1), that synergy is lost.</p>
<p></p>
<p>3) Which Option Is Best for Highest Accuracy and Consistency?
    1.  Option #1 (Ignore adjacency)
       Easiest, but does not incorporate the StageA output at all. If your adjacency is accurate, youre missing out on the biggest synergy.
       Typically not the best result for a pipeline that invests in StageA.
    2.  Option #2 (Post-hoc refiner)
       Slight synergy: TorsionBERT gives angles from sequence; a GNN sees adjacency + angles, corrects them.
       You do keep TorsionBERTs pre-trained power. But two separate models can be more finickyand the refiner must be trained on 3D or angle data anyway.
       Can yield decent results if well done, and is simpler than a full TorsionBERT rewrite.
    3.  Option #3 (Fine-tune TorsionBERT with adjacency as an aux input)
       Less rewriting of the model architecture than Option #4, because youre mostly hooking adjacency somewhere in the TorsionBERT embedding or first few layers.
       Youll get a single model that sees sequence + adjacency and outputs angles.
       The model can still leverage TorsionBERTs pre-trained weights. You only add adjacency channels in the embedding stage or hidden states, then do a partial re-train.
       Likely a good balance of synergy vs. dev time.
    4.  Option #4 (Rewrite TorsionBERT to fully integrate adjacency in attention)
       This is the most advanced approach: e.g., replacing TorsionBERTs self-attention with a pairbias approach, or adding adjacency gating to QK^T.
       Potentially best synergy if well-implemented: the entire TorsionBERT architecture is graph-aware, so every layer can factor in the StageA adjacency constraints.
       Downside: High engineering overhead. Might require re-training from scratch or near scratch, which can be time-consuming.</p>
<p>Hence:
       If you want the absolute best synergy (and can handle re-coding TorsionBERTs attentions), Option #4.
       If you want still integrated with less overhead, Option #3.</p>
<p></p>
<p>4) Hybrid Approaches?</p>
<p>You also mentioned hybrid. One interesting approach is:
       Fine-tune TorsionBERT with adjacency as an input embedding (Option #3).
       Then optionally do a small diffusion or GNN refiner that also sees adjacency, bridging you and letting your StageD do global corrections.</p>
<p>But typically for the pipeline to remain cleaner, you either do one integrated TorsionBERT that already sees adjacency (Option #3 or #4) or do TorsionBERT + separate adjacency-based refiner (Option #2).</p>
<p></p>
<p>5) Final Recommendation</p>
<div class="highlight"><pre><span></span><code>Short Answer: If your priority is truly best synergy + best final accuracy for a multi-stage system, the most powerful approach is:
    1.  Deeply integrate adjacency into TorsionBERTs architecture or attention (Option #4).
2.  Retrain (or at least strongly finetune) the model so that the entire angleprediction process is adjacencyaware from the start.
3.  Then feed those adjacencyconsistent angles into your angle-based diffusion or StageC forward kinematics, leading to a cohesive pipeline.
</code></pre></div>
<p>If you do not have enough time or resources for a partial rewrite, then:</p>
<div class="highlight"><pre><span></span><code>Practical Middle Ground: Option #3  add adjacency as an auxiliary feature in TorsionBERTs embedding or first layer, finetune the model, and skip rewriting the deeper selfattention. Youll still get fairly strong synergy with StageA, with significantly less re-coding.
</code></pre></div>
<p>[emoji]: Thats the trade-off: rewriting the entire TorsionBERT attention is the best for performance, but the fastest path is just injecting adjacency signals into TorsionBERTs existing architecture (Option #3).</p>
<p>Below is a concise check of how well that paper excerpt (the Bioinformatics article text you pasted) answers each of the key questions we had about RNA-TorsionBERTnamely the dataset size, model-parameter size, compute used, maximum nucleotide length, and final TM-score.</p>
<p></p>
<p>1) Dataset Size
       The excerpt does indeed mention a Training set of 4267 structures** filtered by max sequence length 512, plus a Validation set of 29 structures and a Test set of 34 (RNA-Puzzles + CASP-RNA).
       So, yes, it does confirm that ~42004267 structures were used in pre-training or training.</p>
<p></p>
<p>2) Model Parameter Size
       In the excerpt, the authors say the DNABERT has a size of ~328MB, while RNABERT is around 2MB. Then they pick the DNABERT-3mer variant and call that RNA-TorsionBERT.
       So from that text, their final TorsionBERT is ~328MB of model weights. (Previously, a separate reference had 86.9MB, but this paper specifically states ~328MB for the DNABERT-based approach.)
       They do not give an exact million parameters count. But that 300+MB checkpoint typically corresponds to hundreds of millions of parameters.</p>
<p></p>
<p>3) Compute Used
       The excerpt never explicitly says how many GPUs or what HPC environment was used to train.
       So it does not really answer the compute used question, beyond referencing normal PyTorch/Hugging Face usage.</p>
<p></p>
<p>4) Maximum Nucleotide Length
       The article states that they truncated sequences to a maximum of 512 nucleotides, matching DNABERTs limit.
       Hence it does confirm a 512-nt max input length.</p>
<p></p>
<p>5) TM-score for TorsionBERT
       The snippet does not provide a final TM-score. Instead, they measure angle errors (via MCQ) and compare with other methods.
       They do mention that, if you wanted to fully rebuild 3D structures from TorsionBERT angles, you could measure a TM-score, but they did not do so in the paper.</p>
<p></p>
<p>Conclusion</p>
<p>Yes, the excerpt does clarify:
    1.  Dataset: ~4267 structures for training, plus smaller sets for validation &amp; test.
    2.  Model size: ~328MB (DNABERT-based).
    3.  Max length: 512 nucleotides.
    4.  Compute: Not stated.
    5.  No direct TM-score is given; they use MCQ as an anglebased metric instead.</p>
<p>So it largely answers the questions about dataset size, model size, and max sequence length; but not the compute details or a TorsionBERT TM-score.
Below is a rough Fermistyle reasoning to estimate the likely compute cost (how many GPUdays?) and a plausible TMscore range if TorsionBERTs angles were actually used to rebuild 3D structures and then aligned. These are not official numbersjust an informed guess based on the model size, typical BERTlike training, and known compounding errors when going from angles  3D.</p>
<p></p>
<p>1) Model Parameter &amp; Checkpoint Size
       The paper references ~328MB for the DNABERT-3mer version they adapted into RNA-TorsionBERT.
       In float32 (4 bytes/param), 1MB \approx 250k parameters.
       Thus, 328MB \approx 82million parameters (very approximate).
       Another older reference says ~86.9MB, but in the new text we have ~328MB; presumably its a larger version or includes embeddings, some optimization states, etc.</p>
<p>So we can guess ~80100M parameters total.</p>
<p></p>
<p>2) Fermi Estimate of Training Compute</p>
<p>A typical BERTBase (110M parameters) can be trained on text with a single midrange GPU (like an NVIDIA V100) in roughly 24 weeks (depending on batch size, sequence length, and number of epochs). More industrial setups use 48 GPUs to cut that time to just a few days.</p>
<p>Given TorsionBERT for RNA torsion angles:
    1.  They first do a pretraining on ~4200 RNA structures or more. Thats smaller data than typical big text corpora, so you might see fewer total steps or smaller batch sizes.
    2.  They do a finetuning pass on ~200300 RNAs in a supervised angleprediction setting. That is fairly quick.</p>
<p>Hence a ballpark guess:
       Pretraining might have taken on the order of a few days on, say, a single or dual GPU, since the dataset (4200) is not huge but the model is quite large. Possibly 12 weeks if done with small batch sizes.
       Finetuning on a set of a few hundred RNAs is typically just hours or at most a day on one GPU.</p>
<p>Conclusion: Probably \sima few GPUdays to a couple of GPUweeks total training time, not a monstrous HPC requirement like protein-scale data.</p>
<p></p>
<p>3) Probable TMScore if Rebuilding 3D</p>
<p>They do not provide a final TMscore for TorsionBERT. But we can guess:
    1.  Angle  3D compounding error: If you reconstruct coordinates from predicted angles (without an advanced energy minimization), small angle deviations can accumulate. This can degrade RMSD or TMscore quicklyespecially on longer RNAs.
    2.  In practice, nearnative reconstructions with perfect angles can yield high TM (&gt;0.8). But with predicted angles, 3D folds can drift.</p>
<p>A typical guess:
       For short RNAs (&lt;100 nt), the reconstructed TM might easily be ~0.50.7 if the angles are fairly accurate.
       For longer RNAs (200300 nt), the compounding error might push average TM down to ~0.30.5 if no additional relaxation is done.
       If one introduced a light refinement (e.g., a local MD step or partial minimization), you might nudge that up to ~0.50.6 overall or better.</p>
<p>Hence a Fermi guess for average TM across varied RNAs could be ~0.40.6. That is still not officialjust a plausible range if you rely on pure forwardkinematics from TorsionBERTs angles, with minimal geometry correction.</p>
<p></p>
<p>4) Summary of the Fermi Estimates
    1.  Model Size: ~80100 million parameters, given a 300+MB checkpoint.
    2.  Compute: Possibly a few GPUdays to 2 GPUweeks for the entire pretrain/finetune cycle, on a standard HPC environment.
    3.  TMScore: If you rebuilt 3D from TorsionBERT angles, a typical no advanced refinement guess might be 0.40.6 range overall, varying by RNA length. With a short local minimization, you might push that higher.</p>
<p>These are all besteffort guesses based on typical BERT training times and the known difficulty of angle3D reconstructions for RNA.</p>
<p>===</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>