
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../torsionBert/">
      
      
        <link rel="next" href="../torsionBert_full_paper/">
      
      
        
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Code - rna_predict</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#file-srcenumsinitpy" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="rna_predict" class="md-header__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            rna_predict
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Code
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="rna_predict" class="md-nav__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    rna_predict
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Guides
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Guides
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" >
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Getting Started
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Getting Started
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/windows_compatibility_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Windows Compatibility
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Best Practices
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/code_quality_best_practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code Quality
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2_2" id="__nav_2_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_coverage_strategy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage Strategy
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/progressive_coverage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progressive Coverage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_generation_prompt/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Test Generation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_2_3" id="__nav_2_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Debugging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Debugging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/debugging/comprehensive_debugging_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Guide
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Pipeline
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Overview
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/core_framework/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Core Framework
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Multi_Stage_Implementation_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Multi-Stage Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Hydra Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Hydra Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_gap_analysis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Gap Analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_master_document/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Master Document
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3_4" id="__nav_3_2_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Components
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_2_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Components
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageA/StageA_2D_Adjacency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageB/StageB_Torsion_Pairwise/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageC/StageC_3D_Reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageD/StageD_Diffusion_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage D
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/unified_latent/UnifiedLatentMerger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage A
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/StageA_RFold/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_code/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Code
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/rfold-demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/stage_a_extra/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Extra Notes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" checked>
        
          
          <label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage B
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Stage_B/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_3_4_2" id="__nav_3_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    TorsionBERT
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    TorsionBERT
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../torsionBert/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Code
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Code
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#file-srcenumsinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/enums/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcenumsatomspy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/enums/atoms.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelpercomputation_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/computation_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperextractor_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/extractor_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperrna_torsionbert_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/rna_torsionBERT_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcmetricsinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/metrics/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcmetricsmcqpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/metrics/mcq.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcrna_torsionbert_clipy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/rna_torsionBERT_cli.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srctb_mcq_clipy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/tb_mcq_cli.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcutilspy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/utils.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-readmemd" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /README.md
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rna-torsionbert" class="md-nav__link">
    <span class="md-ellipsis">
      
        RNA-TorsionBERT
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RNA-TorsionBERT">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rna-torsionbert-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        RNA-TorsionBERT usage
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tb-mcq" class="md-nav__link">
    <span class="md-ellipsis">
      
        TB-MCQ
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#docker" class="md-nav__link">
    <span class="md-ellipsis">
      
        Docker
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../torsionBert_full_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" >
        
          
          <label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage C
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Stage_C/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Integrated_RNA_Geometry_and_3D_Reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Geometry & Reconstruction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/Unified, Comprehensive Plan for Integrating MP-NeRF into Stage C.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Integration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageC/mp_nerf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Details
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_6" >
        
          
          <label class="md-nav__link" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Unified Latent
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unified_latent/Perceiver_IO/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Perceiver IO
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Energy_Minimization_%26_Molecular_Dynamics_%28MD%29_for_RNA_Structure_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Energy & MD
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_8" >
        
          
          <label class="md-nav__link" for="__nav_3_8" id="__nav_3_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Kaggle
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Kaggle
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/kaggle_competition/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Competition
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/M2_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    M2 Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Advanced Methods
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Advanced Methods
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    AlphaFold 3
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    AlphaFold 3
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AF3_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AlphaFold3_progress/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progress
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/Pairwise_Distance_Based_Prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Distance Prediction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Diffusion
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Diffusion
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/s4_diffusion/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    S4 Diffusion
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/isosteric_substitutions/RNA_isostericity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Isosteric Substitutions
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    External Literature
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    External Literature
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/2d_structure_prediction_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2D Structure
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/RNA_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA Papers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Residue Atom Bridging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Residue Atom Bridging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/audit_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Audit Report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/design_spec/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design Spec
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/documentation_draft/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Documentation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/implementation_notes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Implementation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/refactoring_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Refactoring
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/bridging_caveats_guidelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Guidelines
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Torsion Calculations
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Torsion Calculations
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angles_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Standard_Bond_Lengths_and_Angles_in_RNA_Nucleotides/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Bond Lengths & Angles
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Torsion_Angles_3D_Coordinates_RNA_Structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    3D Coordinates
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angle_Latent_Manifold_Representation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Latent Manifold
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#file-srcenumsinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/enums/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcenumsatomspy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/enums/atoms.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelpercomputation_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/computation_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperextractor_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/extractor_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srchelperrna_torsionbert_helperpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/helper/rna_torsionBERT_helper.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcmetricsinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/metrics/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcmetricsmcqpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/metrics/mcq.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcinitpy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/init.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcrna_torsionbert_clipy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/rna_torsionBERT_cli.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srctb_mcq_clipy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/tb_mcq_cli.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-srcutilspy" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /src/utils.py
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#file-readmemd" class="md-nav__link">
    <span class="md-ellipsis">
      
        File: /README.md
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rna-torsionbert" class="md-nav__link">
    <span class="md-ellipsis">
      
        RNA-TorsionBERT
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="RNA-TorsionBERT">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Installation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#rna-torsionbert-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        RNA-TorsionBERT usage
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tb-mcq" class="md-nav__link">
    <span class="md-ellipsis">
      
        TB-MCQ
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#docker" class="md-nav__link">
    <span class="md-ellipsis">
      
        Docker
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#citation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Citation
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<p>Directory Structure:</p>
<p> ./
     src
        enums
           <strong>init</strong>.py
           atoms.py
        helper
           <strong>init</strong>.py
           computation_helper.py
           extractor_helper.py
           rna_torsionBERT_helper.py
        metrics
           <strong>init</strong>.py
           mcq.py
        <strong>init</strong>.py
        rna_torsionBERT_cli.py
        tb_mcq_cli.py
        utils.py
     README.md</p>
<hr />
<h2 id="file-srcenumsinitpy">File: /src/enums/<strong>init</strong>.py</h2>
<hr />
<h2 id="file-srcenumsatomspy">File: /src/enums/atoms.py</h2>
<p>import numpy as np
import os</p>
<p>ALL_ATOMS = [
    "P",
    "OP1",
    "OP2",
    "O5'",
    "C5'",
    "C4'",
    "O4'",
    "C3'",
    "O3'",
    "C1'",
    "C2'",
    "N1",
    "C2",
    "N9",
    "C4",
]</p>
<p>ANGLES = {
    "alpha": {"atoms": ["O3'", "P", "O5'", "C5'"], "index": [-1, 0, 0, 0]},
    "beta": {"atoms": ["P", "O5'", "C5'", "C4'"], "index": [0, 0, 0, 0]},
    "gamma": {"atoms": ["O5'", "C5'", "C4'", "C3'"], "index": [0, 0, 0, 0]},
    "delta": {"atoms": ["C5'", "C4'", "C3'", "O3'"], "index": [0, 0, 0, 0]},
    "epsilon": {"atoms": ["C4'", "C3'", "O3'", "P"], "index": [0, 0, 0, 1]},
    "zeta": {"atoms": ["C3'", "O3'", "P", "O5'"], "index": [0, 0, 1, 1]},
    "chi": {"atoms": ["O4'", "C1'", "N1", "C2"], "index": [0, 0, 0, 0]},
    "eta": {"atoms": ["C4'", "P", "C4'", "P"], "index": [-1, 0, 0, 1]},
    "theta": {"atoms": ["P", "C4'", "P", "C4'"], "index": [0, 0, 1, 1]},
    "eta'": {"atoms": ["C1'", "P", "C1'", "P"], "index": [-1, 0, 0, 1]},
    "theta'": {"atoms": ["P", "C1'", "P", "C1'"], "index": [0, 0, 1, 1]},
    "v0": {"atoms": ["C4'", "O4'", "C1'", "C2'"], "index": [0, 0, 0, 0]},
    "v1": {"atoms": ["O4'", "C1'", "C2'", "C3'"], "index": [0, 0, 0, 0]},
    "v2": {"atoms": ["C1'", "C2'", "C3'", "C4'"], "index": [0, 0, 0, 0]},
    "v3": {"atoms": ["C2'", "C3'", "C4'", "O4'"], "index": [0, 0, 0, 0]},
    "v4": {"atoms": ["C3'", "C4'", "O4'", "C1'"], "index": [0, 0, 0, 0]},
}</p>
<hr />
<h2 id="file-srchelperinitpy">File: /src/helper/<strong>init</strong>.py</h2>
<hr />
<h2 id="file-srchelpercomputation_helperpy">File: /src/helper/computation_helper.py</h2>
<p>import numpy as np
from typing import List</p>
<p>from src.enums.atoms import ANGLES, ALL_ATOMS
from src.utils import compute_torsion_angle</p>
<p>class ComputationHelper:
    def <strong>init</strong>(self, matrix: np.ndarray, sequence: str):
        self.matrix = matrix
        self.sequence = sequence</p>
<div class="highlight"><pre><span></span><code>def compute_angles(self, angle_name: str) -&gt; List:
    &quot;&quot;&quot;
    Compute all the angles for the given structure.
    :param angle_name: the angle to compute values from
    :return: a list with the angle values
    &quot;&quot;&quot;
    c_angle_dict = ANGLES.get(angle_name, {})
    atoms = c_angle_dict.get(&quot;atoms&quot;, [])
    atoms_position = [ALL_ATOMS.index(atom) for atom in atoms]
    indexes = c_angle_dict.get(&quot;index&quot;, [])
    angles_out = []
    for i, c_atoms in enumerate(self.matrix):
        if angle_name == &quot;chi&quot; and self.sequence[i] in [&quot;A&quot;, &quot;G&quot;]:
            atoms_position = [
                ALL_ATOMS.index(atom) for atom in [&quot;O4&#39;&quot;, &quot;C1&#39;&quot;, &quot;N9&quot;, &quot;C4&quot;]
            ]
        if angle_name == &quot;chi&quot; and self.sequence[i] in [&quot;C&quot;, &quot;U&quot;]:
            atoms_position = [
                ALL_ATOMS.index(atom) for atom in [&quot;O4&#39;&quot;, &quot;C1&#39;&quot;, &quot;N1&quot;, &quot;C2&quot;]
            ]
        specific_atoms = [
            self.matrix[i + offset, atom_pos]
            for offset, atom_pos in zip(indexes, atoms_position)
            if i + offset &lt; len(self.matrix) and i + offset &gt;= 0
        ]
        angle = (
            compute_torsion_angle(*specific_atoms)
            if len(specific_atoms) == 4
            else np.nan
        )
        angles_out.append(angle)
    return angles_out
</code></pre></div>
<hr />
<h2 id="file-srchelperextractor_helperpy">File: /src/helper/extractor_helper.py</h2>
<p>from typing import Dict, List, Optional
import numpy as np</p>
<p>import pandas as pd</p>
<p>from src.enums.atoms import ALL_ATOMS, ANGLES
from src.helper.computation_helper import ComputationHelper
from src.utils import read_all_atoms, get_sequence</p>
<p>class ExtractorHelper:
    def <strong>init</strong>(self, all_atoms: List = ALL_ATOMS):
        self.all_atoms = all_atoms</p>
<div class="highlight"><pre><span></span><code>def extract_all(
    self, in_pdb: str, save_to_path: Optional[str] = None
) -&gt; pd.DataFrame:
    &quot;&quot;&quot;
    Extract all the torsional angles and bond angles from the pdb file
    :param in_pdb: path to a .pdb file
    :param save_to_path: path where to save the output
    :return: a .csv file with the torsional and bond angles.
    &quot;&quot;&quot;
    all_atoms = read_all_atoms(in_pdb)
    matrix = self.convert_atoms_to_matrix(all_atoms)
    sequence = [element for element in get_sequence(in_pdb)]
    computation_helper = ComputationHelper(matrix, sequence)
    torsion_angles = {
        angle: computation_helper.compute_angles(angle) for angle in ANGLES
    }
    sequence = [element for element in get_sequence(in_pdb)]
    df = pd.DataFrame(
        {**{&quot;sequence&quot;: sequence}, **torsion_angles},
        index=range(1, len(sequence) + 1),
    )
    if save_to_path:
        df.to_csv(save_to_path)
    return df

def convert_atoms_to_matrix(self, all_atoms: Dict) -&gt; np.ndarray:
    &quot;&quot;&quot;
    Convert the different atoms into a matrix of size (L, N, 3) where:
        L: the number of nucleotides
        N: the number of atoms per nucleotide
        3: the x,y,z coordinates
    :param all_atoms: list of atoms with their coordinates
    :return: a np.array matrix
    &quot;&quot;&quot;
    output = np.nan * np.ones((len(all_atoms), len(self.all_atoms), 3))
    for index, atoms in enumerate(all_atoms):
        for atom in atoms:
            if atom in self.all_atoms:
                output[index, self.all_atoms.index(atom)] = np.array(atoms[atom])
    return output
</code></pre></div>
<hr />
<h2 id="file-srchelperrna_torsionbert_helperpy">File: /src/helper/rna_torsionBERT_helper.py</h2>
<p>import transformers
from transformers import AutoModel, AutoTokenizer
import numpy as np
import pandas as pd
from typing import Optional, Dict
import os</p>
<p>os.environ["TOKENIZERS_PARALLELISM"] = "false"</p>
<p>transformers.logging.set_verbosity_error()</p>
<p>BACKBONE = [
    "alpha",
    "beta",
    "gamma",
    "delta",
    "epsilon",
    "zeta",
    "chi",
    "eta",
    "theta",
    "eta'",
    "theta'",
    "v0",
    "v1",
    "v2",
    "v3",
    "v4",
]</p>
<p>class RNATorsionBERTHelper:
    def <strong>init</strong>(self):
        self.model_name = "sayby/rna_torsionbert"
        self.tokenizer = AutoTokenizer.from_pretrained(
            self.model_name, trust_remote_code=True
        )
        self.params_tokenizer = {
            "return_tensors": "pt",
            "padding": "max_length",
            "max_length": 512,
            "truncation": True,
        }
        self.model = AutoModel.from_pretrained(self.model_name, trust_remote_code=True)</p>
<div class="highlight"><pre><span></span><code>def predict(self, sequence: str):
    sequence_tok = self.convert_raw_sequence_to_k_mers(sequence)
    inputs = self.tokenizer(sequence_tok, **self.params_tokenizer)
    outputs = self.model(inputs)[&quot;logits&quot;]
    outputs = self.convert_sin_cos_to_angles(
        outputs.cpu().detach().numpy(), inputs[&quot;input_ids&quot;]
    )
    output_angles = self.convert_logits_to_dict(
        outputs[0, :], inputs[&quot;input_ids&quot;][0, :].cpu().detach().numpy()
    )
    output_angles.index = list(sequence)[:-2]  # Because of the 3-mer representation
    return output_angles

def convert_raw_sequence_to_k_mers(self, sequence: str, k_mers: int = 3):
    &quot;&quot;&quot;
    Convert a raw RNA sequence into sequence readable for the tokenizer.
    It converts the sequence into k-mers, and replace U by T
    :return: input readable by the tokenizer
    &quot;&quot;&quot;
    sequence = sequence.upper().replace(&quot;U&quot;, &quot;T&quot;)
    k_mers_sequence = [
        sequence[i : i + k_mers]
        for i in range(len(sequence))
        if len(sequence[i : i + k_mers]) == k_mers
    ]
    return &quot; &quot;.join(k_mers_sequence)

def convert_sin_cos_to_angles(
    self, output: np.ndarray, input_ids: Optional[np.ndarray] = None
):
    &quot;&quot;&quot;
    Convert the raw predictions of the RNA-TorsionBERT into angles.
    It converts the cos and sinus into angles using:
        alpha = arctan(sin(alpha)/cos(alpha))
    :param output: Dictionary with the predictions of the RNA-TorsionBERT per angle
    :param input_ids: the input_ids of the RNA-TorsionBERT. It allows to only select the of the sequence,
        and not the special tokens.
    :return: a np.ndarray with the angles for the sequence
    &quot;&quot;&quot;
    if input_ids is not None:
        output[
            (input_ids == 0)
            | (input_ids == 2)
            | (input_ids == 3)
            | (input_ids == 4)
        ] = np.nan
    pair_indexes, impair_indexes = np.arange(0, output.shape[-1], 2), np.arange(
        1, output.shape[-1], 2
    )
    sin, cos = output[:, :, impair_indexes], output[:, :, pair_indexes]
    tan = np.arctan2(sin, cos)
    angles = np.degrees(tan)
    return angles

def convert_logits_to_dict(self, output: np.ndarray, input_ids: np.ndarray) -&gt; Dict:
    &quot;&quot;&quot;
    Convert the raw predictions into dictionary format.
    It removes the special tokens and only keeps the predictions for the sequence.
    :param output: predictions from the models in angles
    :param input_ids: input ids from the tokenizer
    :return: a dictionary with the predictions for each angle
    &quot;&quot;&quot;
    index_start, index_end = (
        np.where(input_ids == 2)[0][0],
        np.where(input_ids == 3)[0][0],
    )
    output_non_pad = output[index_start + 1 : index_end, :]
    output_angles = {
        angle: output_non_pad[:, angle_index]
        for angle_index, angle in enumerate(BACKBONE)
    }
    out = pd.DataFrame(output_angles)
    return out
</code></pre></div>
<hr />
<h2 id="file-srcmetricsinitpy">File: /src/metrics/<strong>init</strong>.py</h2>
<hr />
<h2 id="file-srcmetricsmcqpy">File: /src/metrics/mcq.py</h2>
<p>from typing import Union, List
import numpy as np
import pandas as pd</p>
<p>import warnings</p>
<p>warnings.filterwarnings("ignore")</p>
<p>TORSION_TO_ANGLES = {
    "BACKBONE": ["alpha", "beta", "gamma", "delta", "epsilon", "zeta", "chi", "phase"],
    "PSEUDO": ["eta", "theta"],
}</p>
<p>class MCQ:
    """
    MCQ Helper. Reproduce the MCQ computation from https://github.com/tzok/mcq4structures.
    """</p>
<div class="highlight"><pre><span></span><code>def mod(self, values: Union[np.ndarray, float]) -&gt; Union[np.ndarray, float]:
    &quot;&quot;&quot;
    Compute the mod(t) = t + 2pi % 2pi
    Computation is done in degree
    :param values:
    :return:
    &quot;&quot;&quot;
    return (values + 360) % 360

def difference(self, x: float, y: float):
    &quot;&quot;&quot;
    Compute the distance between two angles
    :param x: the true angle
    :param x: the predicted atom
    :return: the distance based on MCQ computation
    &quot;&quot;&quot;
    if np.isnan(x) and np.isnan(y):
        return 0
    elif np.isnan(x) or np.isnan(y):
        return 180
    else:
        return min(
            abs(self.mod(x) - self.mod(y)), 360 - abs(self.mod(x) - self.mod(y))
        )

def get_phase(self, values: np.ndarray):
    &quot;&quot;&quot;
    Compute the phase P = arctan(v1 + v4 - v0 - v3, 2v2(sin 36 + sin72))
    :param values:
    :return:
    &quot;&quot;&quot;
    try:
        riboses = np.radians(values[[&quot;v0&quot;, &quot;v1&quot;, &quot;v2&quot;, &quot;v3&quot;, &quot;v4&quot;]])
        P = np.arctan(
            riboses[&quot;v1&quot;] + riboses[&quot;v4&quot;] - riboses[&quot;v0&quot;] - riboses[&quot;v3&quot;],
            2 * riboses[&quot;v2&quot;] * (np.sin(np.radians(36)) + np.sin(np.radians(72))),
        )
    except KeyError:
        P = np.nan
    return P

def compute_mcq(
    self,
    true_values: np.ndarray,
    pred_values: np.ndarray,
    torsion: str = &quot;BACKBONE&quot;,
):
    &quot;&quot;&quot;
    Compute the MCQ between two sets of angles.
    :param true_values: experimental inferred angles from the native structure
    :param pred_values: predicted angles
    :param torsion: the type of angles to use. Default to BACKBONE.
    :return:
    &quot;&quot;&quot;
    diff, angles = self._get_diff_angles(true_values, pred_values, torsion)
    sin_mod, cos_mod = (
        np.sin(np.radians(diff)).sum(),
        np.cos(np.radians(diff)).sum(),
    )
    mcq = np.arctan2(sin_mod, cos_mod)
    mcq = np.degrees(mcq)
    return mcq

def get_current_angles(self, torsion: str, pred_cols: List):
    &quot;&quot;&quot;
    Get the current angles values from all the angles available.
    &quot;&quot;&quot;
    angles = []
    for angle in TORSION_TO_ANGLES[torsion]:
        if angle == &quot;phase&quot; and &quot;v0&quot; in pred_cols:
            angles.append(angle)
        elif angle in pred_cols:
            angles.append(angle)
    return angles

def _get_diff_angles(
    self, true_values: pd.DataFrame, pred_values: pd.DataFrame, torsion: str
):
    &quot;&quot;&quot;
    Compute the differences to be used for the MCQ computation
    :return:
    &quot;&quot;&quot;
    if len(pred_values) &lt; len(true_values):
        true_values = true_values[: len(pred_values)]
    else:
        pred_values = pred_values[: len(true_values)]
    angles = self.get_current_angles(torsion, pred_values.columns)
    if torsion == &quot;BACKBONE&quot;:
        phase_true = np.degrees(self.get_phase(true_values))
        phase_pred = np.degrees(self.get_phase(pred_values))
        true_values[&quot;phase&quot;] = phase_true
        pred_values[&quot;phase&quot;] = phase_pred
    true_angles = true_values[angles].values
    pred_angles = pred_values[angles].values
    diff_fn = np.vectorize(self.difference)
    diff = diff_fn(true_angles, pred_angles)
    return diff, angles

def compute_mcq_per_angle(self, true_values, pred_values, torsion: str):
    &quot;&quot;&quot;
    Compute the MCQ for a given angle.
    :param true_values: experimental inferred angles from the native structure
    :param pred_values: predicted angles
    :param torsion: the type of angles to use. Default to BACKBONE.
    :return: MCQ per angle
    &quot;&quot;&quot;
    diff, angles = self._get_diff_angles(true_values, pred_values, torsion)
    sin_mod, cos_mod = np.sin(np.radians(diff)).sum(axis=0), np.cos(
        np.radians(diff)
    ).sum(axis=0)
    mcq = np.arctan2(sin_mod, cos_mod)
    mcq = np.degrees(mcq)
    output = {angle: mcq[i] for i, angle in enumerate(angles)}
    return output

def compute_mcq_per_sequence(self, true_values, pred_values, torsion: str):
    &quot;&quot;&quot;
    Compute the MCQ for a given position.
    :return:
    &quot;&quot;&quot;
    diff, angles = self._get_diff_angles(true_values, pred_values, torsion)
    sin_mod, cos_mod = np.sin(np.radians(diff)).sum(axis=1), np.cos(
        np.radians(diff)
    ).sum(axis=1)
    mcq = np.arctan2(sin_mod, cos_mod)
    mcq = np.degrees(mcq)
    return mcq.tolist()
</code></pre></div>
<hr />
<h2 id="file-srcinitpy">File: /src/<strong>init</strong>.py</h2>
<hr />
<h2 id="file-srcrna_torsionbert_clipy">File: /src/rna_torsionBERT_cli.py</h2>
<p>import argparse
from typing import Optional</p>
<p>from src.helper.rna_torsionBERT_helper import RNATorsionBERTHelper
from src.utils import read_fasta
from loguru import logger</p>
<p>class RNATorsionBERTCLI:
    def <strong>init</strong>(
        self,
        in_seq: Optional[str],
        in_fasta: Optional[str],
        out_path: Optional[str],
        <em>args,
        </em>*kwargs,
    ):
        self.sequence = self._init_inputs(in_seq, in_fasta)
        self.out_path = out_path</p>
<div class="highlight"><pre><span></span><code>def _init_inputs(self, in_seq: Optional[str], in_fasta: Optional[str]) -&gt; str:
    &quot;&quot;&quot;
    Initialise the inputs given the arguments
    :return: the sequence
    &quot;&quot;&quot;
    if in_seq is None and in_fasta is None:
        raise ValueError(&quot;You must provide either a sequence or a fasta file.&quot;)
    if in_seq is not None and in_fasta is not None:
        raise ValueError(
            &quot;Please provide only the sequence or the fasta file, not both.&quot;
        )
    if in_seq is not None:
        sequence = in_seq
    elif in_fasta is not None:
        sequence = read_fasta(in_fasta)
    return sequence

def run(self):
    output = RNATorsionBERTHelper().predict(self.sequence)
    if self.out_path is not None:
        output.to_csv(self.out_path)
        logger.info(f&quot;Saved the output to {self.out_path}&quot;)
    return output

@staticmethod
def get_args():
    parser = argparse.ArgumentParser(
        description=&quot;Prediction of Torsional angles for RNA structures&quot;
    )
    # Add command line arguments
    parser.add_argument(
        &quot;--in_seq&quot;,
        dest=&quot;in_seq&quot;,
        type=str,
        help=&quot;RNA Input sequence.&quot;,
        default=None,
    )
    parser.add_argument(
        &quot;--in_fasta&quot;,
        dest=&quot;in_fasta&quot;,
        type=str,
        help=&quot;Path to a fasta file.&quot;,
        default=None,
    )
    parser.add_argument(
        &quot;--out_path&quot;,
        dest=&quot;out_path&quot;,
        type=str,
        help=&quot;Path to a .csv file to save the prediction&quot;,
        default=None,
    )
    # Parse the command line arguments
    args = parser.parse_args()
    return args
</code></pre></div>
<p>if <strong>name</strong> == "<strong>main</strong>":
    args = RNATorsionBERTCLI.get_args()
    rna_torsionBERT_cli = RNATorsionBERTCLI(**vars(args))
    rna_torsionBERT_cli.run()</p>
<hr />
<h2 id="file-srctb_mcq_clipy">File: /src/tb_mcq_cli.py</h2>
<p>import argparse
import os</p>
<p>from loguru import logger
import tqdm
import pandas as pd
from typing import Optional, List</p>
<p>from src.helper.extractor_helper import ExtractorHelper
from src.helper.rna_torsionBERT_helper import RNATorsionBERTHelper
from src.metrics.mcq import MCQ</p>
<p>class TBMCQCLI:
    def <strong>init</strong>(self, in_pdb: str, out_path: Optional[str], <em>args, </em>*kwargs):
        self.list_files = self._init_pdb(in_pdb)
        self.out_path = out_path</p>
<div class="highlight"><pre><span></span><code>def _init_pdb(self, in_pdb: Optional[str]) -&gt; List:
    &quot;&quot;&quot;
    Initialise the inputs structures.
    :param in_pdb: a path to either a .pdb file or a directory of .pdb files
    :return: a list of path to .pdb files
    &quot;&quot;&quot;
    if os.path.isdir(in_pdb):
        list_files = os.listdir(in_pdb)
        list_files = [os.path.join(in_pdb, file_) for file_ in list_files]
    elif os.path.isfile(in_pdb):
        list_files = [in_pdb]
    else:
        logger.info(f&quot;NO INPUTS FOUND FOR INPUT .PDB: {in_pdb}&quot;)
    return list_files

def run(self):
    all_scores = {&quot;RNA&quot;: [], &quot;TB-MCQ&quot;: []}
    for in_path in tqdm.tqdm(self.list_files):
        score = self.compute_tb_mcq(in_path)
        all_scores[&quot;RNA&quot;].append(os.path.basename(in_path))
        all_scores[&quot;TB-MCQ&quot;].append(score)
    all_scores = pd.DataFrame(all_scores, columns=[&quot;TB-MCQ&quot;, &quot;RNA&quot;]).set_index(
        &quot;RNA&quot;
    )
    if self.out_path is not None:
        logger.info(f&quot;Saved the output to {self.out_path}&quot;)
        all_scores.to_csv(self.out_path, index=True)
    return all_scores

def compute_tb_mcq(self, pred_path: str) -&gt; float:
    &quot;&quot;&quot;
    Compute the TB-MCQ with RNA-TorsionBERT model
    :param pred_path: the path to the .pdb file of a prediction.
            It could be a native or a predicted structure.
    &quot;&quot;&quot;
    experimental_angles = ExtractorHelper().extract_all(pred_path)
    sequence = &quot;&quot;.join(experimental_angles[&quot;sequence&quot;].values)
    torsionBERT_helper = RNATorsionBERTHelper()
    torsionBERT_output = torsionBERT_helper.predict(sequence)
    mcq = MCQ().compute_mcq(experimental_angles, torsionBERT_output)
    return mcq

@staticmethod
def get_args():
    parser = argparse.ArgumentParser(
        description=&quot;Prediction of Torsional angles for RNA structures&quot;
    )
    # Add command line arguments
    parser.add_argument(
        &quot;--in_pdb&quot;,
        dest=&quot;in_pdb&quot;,
        type=str,
        help=&quot;Path a .pdb file or a directory of .pdb files.&quot;,
        default=None,
    )
    parser.add_argument(
        &quot;--out_path&quot;,
        dest=&quot;out_path&quot;,
        type=str,
        help=&quot;Path to a .csv file to save the predictions.&quot;,
        default=None,
    )
    # Parse the command line arguments
    args = parser.parse_args()
    return args
</code></pre></div>
<p>if <strong>name</strong> == "<strong>main</strong>":
    args = TBMCQCLI.get_args()
    tb_mcq_cli = TBMCQCLI(**vars(args))
    tb_mcq_cli.run()</p>
<hr />
<h2 id="file-srcutilspy">File: /src/utils.py</h2>
<p>from typing import Any, List, Dict
from Bio.PDB import Atom, Model, Chain, Residue, Structure, PDBIO
import Bio
import numpy as np
from Bio.PDB import Atom, Residue, PDBParser
import warnings</p>
<p>warnings.filterwarnings("ignore")</p>
<p>def read_fasta(in_path: str) -&gt; str:
    """
    Read a fasta file to get the sequence
    :param in_path: path to a .fasta file
    :return: the RNA sequence
    """
    with open(in_path, "r") as f:
        lines = f.readlines()
    sequence = "".join([line.strip() for line in lines[1:]])
    return sequence</p>
<p>def read_all_atoms(in_pdb: str) -&gt; Any:
    """
    Read and return the coordinates of all the  atoms from a pdb file
    :param in_pdb: path to a .pdb file
    """
    parser = PDBParser()
    all_atoms = []
    structure = parser.get_structure("", in_pdb)
    for model in structure:
        for chain in model:
            for residue in chain:
                res = residue.get_resname().replace(" ", "")
                if res in ["A", "C", "G", "U"]:
                    atoms = get_atoms_torsion(residue)
                    c_atom = {
                        atom.get_name(): atom.get_coord().tolist() for atom in atoms
                    }
                    all_atoms.append(c_atom)
    return all_atoms</p>
<p>def get_atoms_torsion(residue: Bio.PDB.Residue.Residue):
    """
    Return the atoms coordinates for a given residue.
    :param residue: the residue to get the atoms from
    """
    atoms = []
    for atom in residue:
        atoms.append(atom)
    return atoms</p>
<p>def compute_torsion_angle(
    atom1: np.ndarray, atom2: np.ndarray, atom3: np.ndarray, atom4: np.ndarray
) -&gt; float:
    """
    Compute torsional angles between 4 atoms
    :return: the torsional angles between the atoms
    """
    v12 = atom1 - atom2
    v23 = atom2 - atom3
    v34 = atom3 - atom4
    e1 = np.cross(v12, v23)
    e2 = np.cross(v23, v34)
    sign = +1 if np.dot(v23, np.cross(e1, e2)) &lt; 0 else -1
    angle_in_radians = np.arccos(
        np.dot(e1, e2) / (np.linalg.norm(e1) * np.linalg.norm(e2))
    )
    angle_in_degrees = sign * np.degrees(angle_in_radians)
    return angle_in_degrees</p>
<p>def get_sequence(in_pdb: str) -&gt; str:
    """
    Return the RNA sequence from a .pdb file
    :param in_pdb: path to a pdb file
    :return: RNA sequence of nucleotides
    """
    parser = PDBParser()
    structure = parser.get_structure("structure", in_pdb)
    sequence = ""
    for model in structure:
        for chain in model:
            for residue in chain:
                res = residue.get_resname().replace(" ", "")
                if res in ["A", "C", "G", "U"]:
                    sequence += res
    return sequence</p>
<hr />
<h2 id="file-readmemd">File: /README.md</h2>
<h1 id="rna-torsionbert">RNA-TorsionBERT</h1>
<p><code>RNA-TorsionBERT</code> is a 86.9 MB parameter BERT-based language model that predicts RNA torsional and pseudo-torsional angles from the sequence.</p>
<p><img alt="" src="./img/dnabert_architecture_final.png" /></p>
<p><code>RNA-TorsionBERT</code> is a DNABERT model that was pre-trained on ~4200 RNA structures.</p>
<p>It provides improvement of <a href="https://github.com/tzok/mcq4structures">MCQ</a> over the previous state-of-the-art models like 
<a href="https://github.com/jaswindersingh2/SPOT-RNA-1D">SPOT-RNA-1D</a> or inferred angles from existing methods, on the Test Set (composed of RNA-Puzzles and CASP-RNA).</p>
<h2 id="installation">Installation</h2>
<p>To install RNA-TorsionBERT and it's dependencies following commands can be used in terminal:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>pip<span class="w"> </span>install<span class="w"> </span>-r<span class="w"> </span>requirements.txt<span class="w"> </span>
</code></pre></div>
<h2 id="rna-torsionbert-usage">RNA-TorsionBERT usage</h2>
<p>To run the RNA-TorsionBERT, you can use the following command line:
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>python<span class="w"> </span>-m<span class="w"> </span>src.rna_torsionBERT_cli<span class="w"> </span><span class="o">[</span>--seq_file<span class="o">]</span><span class="w"> </span><span class="o">[</span>--in_fasta<span class="o">]</span><span class="w"> </span><span class="o">[</span>--out_path<span class="o">]</span>
</code></pre></div></p>
<p>The arguments are the following:
- <code>--seq_file</code>: RNA Sequence. 
- <code>--in_fasta</code>: Path to the input sequence fasta file. 
- <code>--out_path</code>: Path to a <code>.csv</code> file where the output will be saved. </p>
<p>You can also import in your python code the class <code>RNATorsionBERTCLI</code> from <code>src.rna_torsionBERT_cli</code>. </p>
<h2 id="tb-mcq">TB-MCQ</h2>
<p>TB-MCQ stands for TorsionBERT-MCQ, which is a scoring function to assess the quality of a predicted structure in torsional angle space.
Given the inferred angles from the structures and the predicted angles from the model, TB-MCQ computes the quality of the predicted angles using 
the <a href="https://github.com/tzok/mcq4structures">MCQ</a> (mean of circular quantities) metric.</p>
<p><img alt="" src="./img/torsion_bert_mcq_T.png" /></p>
<p>To run the TB-MCQ scoring function, you can use the following command line:
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>python<span class="w"> </span>-m<span class="w"> </span>src.rna_torsion_cli<span class="w"> </span><span class="o">[</span>--in_pdb<span class="o">]</span><span class="w"> </span><span class="o">[</span>--out_path<span class="o">]</span>
</code></pre></div>
with:</p>
<ul>
<li><code>--in_pdb</code>: Path to the input PDB file.</li>
<li><code>--out_path</code>: Path to a .csv file where the output will be saved.</li>
</ul>
<h2 id="docker">Docker</h2>
<p>To run the code using <code>Docker</code>, you can use the following command line:
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>docker<span class="w"> </span>build<span class="w"> </span>-t<span class="w"> </span>rna_torsionbert<span class="w"> </span>.
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>docker<span class="w"> </span>run<span class="w"> </span>-it<span class="w"> </span>rna_torsionbert<span class="w"> </span>
</code></pre></div></p>
<p>It will enter into a bash console where you could execute the previous commands with all the installations done. </p>
<p>To have example of commands, you can look at the <code>Makefile</code>.</p>
<h2 id="citation">Citation</h2>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="nc">@article</span><span class="p">{</span><span class="nl">rna_torsion_bert</span><span class="p">,</span>
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="w">    </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Bernard, Clment and Postic, Guillaume and Ghannay, Sahar and Tahi, Fariza}</span><span class="p">,</span>
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="w">    </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{RNA-TorsionBERT: leveraging language models for RNA 3D torsion angles prediction}</span><span class="p">,</span>
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="w">    </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Bioinformatics}</span><span class="p">,</span>
<a id="__codelineno-4-5" name="__codelineno-4-5" href="#__codelineno-4-5"></a><span class="w">    </span><span class="na">volume</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{41}</span><span class="p">,</span>
<a id="__codelineno-4-6" name="__codelineno-4-6" href="#__codelineno-4-6"></a><span class="w">    </span><span class="na">number</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{1}</span><span class="p">,</span>
<a id="__codelineno-4-7" name="__codelineno-4-7" href="#__codelineno-4-7"></a><span class="w">    </span><span class="na">pages</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{btaf004}</span><span class="p">,</span>
<a id="__codelineno-4-8" name="__codelineno-4-8" href="#__codelineno-4-8"></a><span class="w">    </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2025}</span><span class="p">,</span>
<a id="__codelineno-4-9" name="__codelineno-4-9" href="#__codelineno-4-9"></a><span class="w">    </span><span class="na">month</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{01}</span><span class="p">,</span>
<a id="__codelineno-4-10" name="__codelineno-4-10" href="#__codelineno-4-10"></a><span class="w">    </span><span class="na">issn</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{1367-4811}</span><span class="p">,</span>
<a id="__codelineno-4-11" name="__codelineno-4-11" href="#__codelineno-4-11"></a><span class="w">    </span><span class="na">doi</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{10.1093/bioinformatics/btaf004}</span><span class="p">,</span>
<a id="__codelineno-4-12" name="__codelineno-4-12" href="#__codelineno-4-12"></a><span class="w">    </span><span class="na">url</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://doi.org/10.1093/bioinformatics/btaf004}</span><span class="p">,</span>
<a id="__codelineno-4-13" name="__codelineno-4-13" href="#__codelineno-4-13"></a><span class="w">    </span><span class="na">eprint</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://academic.oup.com/bioinformatics/article-pdf/41/1/btaf004/61381586/btaf004.pdf}</span><span class="p">,</span>
<a id="__codelineno-4-14" name="__codelineno-4-14" href="#__codelineno-4-14"></a><span class="p">}</span>
</code></pre></div>
<p>====</p>
<p>Hugging Face's logo Hugging Face</p>
<p>Models
Datasets
Spaces
Docs
Enterprise
Pricing</p>
<p>sayby
/
rna_torsionBERT
Token Classification
Transformers
PyTorch
Safetensors
rna_torsionbert
feature-extraction
biology
RNA
Torsional
Angles
custom_code
Model card
Files
Community</p>
<p>RNA-TorsionBERT
Model Description</p>
<p>RNA-TorsionBERT is a 86.9 MB parameter BERT-based language model that predicts RNA torsional and pseudo-torsional angles from the sequence.</p>
<p>RNA-TorsionBERT is a DNABERT model that was pre-trained on ~4200 RNA structures.</p>
<p>It provides improvement of MCQ over the previous state-of-the-art models like SPOT-RNA-1D or inferred angles from existing methods, on the Test Set (composed of RNA-Puzzles and CASP-RNA).</p>
<p>Key Features</p>
<div class="highlight"><pre><span></span><code>Torsional and Pseudo-torsional angles prediction
Predict sequences up to 512 nucleotides
</code></pre></div>
<p>Usage</p>
<p>Get started generating text with RNA-TorsionBERT by using the following code snippet:</p>
<p>from transformers import AutoModel, AutoTokenizer</p>
<p>tokenizer = AutoTokenizer.from_pretrained("sayby/rna_torsionbert", trust_remote_code=True)
model = AutoModel.from_pretrained("sayby/rna_torsionbert", trust_remote_code=True)</p>
<p>sequence = "ACG CGG GGT GTT"
params_tokenizer = {
    "return_tensors": "pt",
    "padding": "max_length",
    "max_length": 512,
    "truncation": True,
}
inputs = tokenizer(sequence, **params_tokenizer)
output = model(inputs)["logits"]</p>
<div class="highlight"><pre><span></span><code>Please note that it was fine-tuned from a DNABERT-3 model and therefore the tokenizer is the same as the one used for DNABERT. Nucleotide U should therefore be replaced by T in the input sequence.
The output is the sinus and the cosine for each angle. The angles are in the following order: alpha, beta,gamma,delta,epsilon,zeta,chi,eta,theta,eta&#39;,theta&#39;,v0,v1,v2,v3,v4.
</code></pre></div>
<p>To convert the predictions into angles, you can use the following code snippet:</p>
<p>import transformers
from transformers import AutoModel, AutoTokenizer
import numpy as np
import pandas as pd
from typing import Optional, Dict
import os</p>
<p>os.environ["TOKENIZERS_PARALLELISM"] = "false"</p>
<p>transformers.logging.set_verbosity_error()</p>
<p>BACKBONE = [
    "alpha",
    "beta",
    "gamma",
    "delta",
    "epsilon",
    "zeta",
    "chi",
    "eta",
    "theta",
    "eta'",
    "theta'",
    "v0",
    "v1",
    "v2",
    "v3",
    "v4",
]</p>
<p>class RNATorsionBERTHelper:
    def <strong>init</strong>(self):
        self.model_name = "sayby/rna_torsionbert"
        self.tokenizer = AutoTokenizer.from_pretrained(
            self.model_name, trust_remote_code=True
        )
        self.params_tokenizer = {
            "return_tensors": "pt",
            "padding": "max_length",
            "max_length": 512,
            "truncation": True,
        }
        self.model = AutoModel.from_pretrained(self.model_name, trust_remote_code=True)</p>
<div class="highlight"><pre><span></span><code>def predict(self, sequence: str):
    sequence_tok = self.convert_raw_sequence_to_k_mers(sequence)
    inputs = self.tokenizer(sequence_tok, **self.params_tokenizer)
    outputs = self.model(inputs)[&quot;logits&quot;]
    outputs = self.convert_sin_cos_to_angles(
        outputs.cpu().detach().numpy(), inputs[&quot;input_ids&quot;]
    )
    output_angles = self.convert_logits_to_dict(
        outputs[0, :], inputs[&quot;input_ids&quot;][0, :].cpu().detach().numpy()
    )
    output_angles.index = list(sequence)[:-2]  # Because of the 3-mer representation
    return output_angles

def convert_raw_sequence_to_k_mers(self, sequence: str, k_mers: int = 3):
    &quot;&quot;&quot;
    Convert a raw RNA sequence into sequence readable for the tokenizer.
    It converts the sequence into k-mers, and replace U by T
    :return: input readable by the tokenizer
    &quot;&quot;&quot;
    sequence = sequence.upper().replace(&quot;U&quot;, &quot;T&quot;)
    k_mers_sequence = [
        sequence[i : i + k_mers]
        for i in range(len(sequence))
        if len(sequence[i : i + k_mers]) == k_mers
    ]
    return &quot; &quot;.join(k_mers_sequence)

def convert_sin_cos_to_angles(
    self, output: np.ndarray, input_ids: Optional[np.ndarray] = None
):
    &quot;&quot;&quot;
    Convert the raw predictions of the RNA-TorsionBERT into angles.
    It converts the cos and sinus into angles using:
        alpha = arctan(sin(alpha)/cos(alpha))
    :param output: Dictionary with the predictions of the RNA-TorsionBERT per angle
    :param input_ids: the input_ids of the RNA-TorsionBERT. It allows to only select the of the sequence,
        and not the special tokens.
    :return: a np.ndarray with the angles for the sequence
    &quot;&quot;&quot;
    if input_ids is not None:
        output[
            (input_ids == 0)
            | (input_ids == 2)
            | (input_ids == 3)
            | (input_ids == 4)
        ] = np.nan
    pair_indexes, impair_indexes = np.arange(0, output.shape[-1], 2), np.arange(
        1, output.shape[-1], 2
    )
    sin, cos = output[:, :, impair_indexes], output[:, :, pair_indexes]
    tan = np.arctan2(sin, cos)
    angles = np.degrees(tan)
    return angles

def convert_logits_to_dict(self, output: np.ndarray, input_ids: np.ndarray) -&gt; Dict:
    &quot;&quot;&quot;
    Convert the raw predictions into dictionary format.
    It removes the special tokens and only keeps the predictions for the sequence.
    :param output: predictions from the models in angles
    :param input_ids: input ids from the tokenizer
    :return: a dictionary with the predictions for each angle
    &quot;&quot;&quot;
    index_start, index_end = (
        np.where(input_ids == 2)[0][0],
        np.where(input_ids == 3)[0][0],
    )
    output_non_pad = output[index_start + 1 : index_end, :]
    output_angles = {
        angle: output_non_pad[:, angle_index]
        for angle_index, angle in enumerate(BACKBONE)
    }
    out = pd.DataFrame(output_angles)
    return out
</code></pre></div>
<p>if <strong>name</strong> == "<strong>main</strong>":
    sequence = "AGGGCUUUAGUCUUUGGAG"
    rna_torsionbert_helper = RNATorsionBERTHelper()
    output_angles = rna_torsionbert_helper.predict(sequence)
    print(output_angles)</p>
<p>Downloads last month
    988 </p>
<p>Safetensors
Model size
86.9M params
Tensor type
F32
Inference Providers
NEW
Token Classification
This model is not currently available via any of the supported Inference Providers.
The model cannot be deployed to the HF Inference API: The HF Inference API does not support model that require custom code execution.
Model tree for sayby/rna_torsionBERT</p>
<p>Base model
zhihan1996/DNA_bert_3
Finetuned
(1)
this model
TOS
Privacy
About
Jobs
Models
Datasets
Spaces
Pricing
Docs</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>