
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../Stage_C/">
      
      
        <link rel="next" href="../mp_nerf/">
      
      
        
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Geometry & Reconstruction - rna_predict</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#end-of-document" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="rna_predict" class="md-header__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            rna_predict
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Geometry & Reconstruction
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="rna_predict" class="md-nav__button md-logo" aria-label="rna_predict" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    rna_predict
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Guides
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Guides
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_1" >
        
          
          <label class="md-nav__link" for="__nav_2_1" id="__nav_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Getting Started
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Getting Started
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/getting_started/windows_compatibility_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Windows Compatibility
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Best Practices
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/code_quality_best_practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code Quality
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2_2" id="__nav_2_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_coverage_strategy/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Coverage Strategy
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/progressive_coverage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progressive Coverage
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/testing/test_generation_prompt/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Test Generation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_2_3" id="__nav_2_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Debugging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_2_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Debugging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../guides/best_practices/debugging/comprehensive_debugging_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Guide
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Pipeline
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Pipeline
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Overview
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/core_framework/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Core Framework
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Comprehensive Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../overview/Multi_Stage_Implementation_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Multi-Stage Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/Integrated_RNA_3D_Prediction_Pipeline_Final_Comprehensive_Design/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/full_pipeline_specification/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Full Specification
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3" id="__nav_3_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Hydra Integration
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Hydra Integration
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_tutorial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Tutorial
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_gap_analysis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Gap Analysis
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/hydra_integration_master_document/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Master Document
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_2_3_4" id="__nav_3_2_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Components
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="4" aria-labelledby="__nav_3_2_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Components
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageA/StageA_2D_Adjacency/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageB/StageB_Torsion_Pairwise/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageC/StageC_3D_Reconstruction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/stageD/StageD_Diffusion_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Stage D
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../integration/hydra_integration/components/unified_latent/UnifiedLatentMerger/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_3" >
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage A
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage A
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/StageA_RFold/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_code/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Code
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/RFold_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RFold Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/rfold-demo/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Demo
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageA/stage_a_extra/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Extra Notes
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4" >
        
          
          <label class="md-nav__link" for="__nav_3_4" id="__nav_3_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage B
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage B
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageB/Stage_B/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_4_2" >
        
          
          <label class="md-nav__link" for="__nav_3_4_2" id="__nav_3_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    TorsionBERT
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    TorsionBERT
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageB/torsionBert/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageB/torsionbert_code/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Code
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../stageB/torsionBert_full_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_5" checked>
        
          
          <label class="md-nav__link" for="__nav_3_5" id="__nav_3_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Stage C
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Stage C
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Stage_C/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Geometry & Reconstruction
  

    
  </span>
  
  

      </a>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../Unified, Comprehensive Plan for Integrating MP-NeRF into Stage C.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Integration
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mp_nerf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    MP-NeRF Details
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_6" >
        
          
          <label class="md-nav__link" for="__nav_3_6" id="__nav_3_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Unified Latent
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Unified Latent
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../unified_latent/Perceiver_IO/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Perceiver IO
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Energy_Minimization_%26_Molecular_Dynamics_%28MD%29_for_RNA_Structure_Refinement/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Energy & MD
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_8" >
        
          
          <label class="md-nav__link" for="__nav_3_8" id="__nav_3_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Testing
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Testing
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_9" >
        
          
          <label class="md-nav__link" for="__nav_3_9" id="__nav_3_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Kaggle
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Kaggle
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/kaggle_competition/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Competition
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../kaggle_info/M2_Plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    M2 Plan
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Advanced Methods
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    Advanced Methods
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1_1" id="__nav_4_1_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    AlphaFold 3
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_1">
            <span class="md-nav__icon md-icon"></span>
            
  
    AlphaFold 3
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AF3_paper/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Paper
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/AlphaFold3_progress/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Progress
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/af3/Pairwise_Distance_Based_Prediction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Distance Prediction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1_2" >
        
          
          <label class="md-nav__link" for="__nav_4_1_2" id="__nav_4_1_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Diffusion
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Diffusion
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/s4_diffusion/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    S4 Diffusion
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/diffusion/test_time_scaling/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Time Scaling
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/advanced_methods/isosteric_substitutions/RNA_isostericity/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Isosteric Substitutions
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    External Literature
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    External Literature
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/2d_structure_prediction_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    2D Structure
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/external_lit/RNA_papers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    RNA Papers
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Residue Atom Bridging
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            
  
    Residue Atom Bridging
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/audit_report/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Audit Report
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/design_spec/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Design Spec
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/documentation_draft/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Documentation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/implementation_notes/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Implementation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/refactoring_plan/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Refactoring
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/residue_atom_bridging/bridging_caveats_guidelines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Guidelines
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Torsion Calculations
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            
  
    Torsion Calculations
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angles_intro/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Introduction
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Standard_Bond_Lengths_and_Angles_in_RNA_Nucleotides/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Bond Lengths & Angles
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/Torsion_Angles_3D_Coordinates_RNA_Structure/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    3D Coordinates
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/torsion_calculations/torsion_angle_Latent_Manifold_Representation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Latent Manifold
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  



<p>Below is a comprehensive, integrated guide that merges the strengths of all previous versions while addressing their weaknesses, aiming for a thorough, practical reference. It explains how to leverage DSSR/3DNA data and standard RNA covalent geometry to construct RNA 3D structures from torsion angles (e.g., an RNA mp-nerf), plus it provides extended reference tables, discussion of environmental variation, and a roadmap of relevant Python software. The result should be suitable for technical documentationboth verbose and practical.</p>
<p></p>
<ol>
<li>Introduction and Scope</li>
</ol>
<p>Building RNA 3D coordinates from torsion angles requires two main ingredients:
    1.  A library of standard RNA bond lengths, bond angles, and ring-closure constraints (i.e., the geometry of the phosphateribosebase).
    2.  A forward-kinematics or NeRF (Natural Extension Reference Frame) style algorithm that places atoms in 3D given internal coordinates (bond length l, bond angle \theta, and torsion \chi).</p>
<p>DSSR (Dissecting the Spatial Structure of RNA) from the 3DNA suite is analysis-focused: given a 3D RNA, it extracts torsions and structural annotations. While it does not natively do a pure torsionCartesian rebuild, it encodes or references standard nucleic-acid geometry (in its fiber or rebuild modules) that you can adopt. Meanwhile, mp-nerf (originally for proteins) can be extended from amino acids to nucleotides by replacing the protein geometry knowledge base with RNA data.</p>
<p>This document:
    1.  Presents standard RNA bond lengths (Table 1) and bond angles (Table 2).
    2.  Explains environmental factors that cause real-world deviations from these ideals.
    3.  Enumerates Python-based tools that help measure, refine, and validate RNA geometry.
    4.  Outlines how to adapt mp-nerf (or a NeRF pipeline) for an RNA torsion  3D process, referencing DSSR/3DNA data as needed.</p>
<p></p>
<ol>
<li>Standard RNA Bond Lengths</li>
</ol>
<p>Below is a detailed table of average bond lengths (in ) for an RNA nucleotide. These values come from surveys of high-resolution crystal structures (e.g. in the Nucleic Acid Database (NDB), the Nucleic Acid Knowledge Base (NAKB), and small-molecule data from the Cambridge Structural Database (CSD)). Each has a small standard deviation, typically 0.010.02.</p>
<p>Table 1: Approximate Bond Lengths in RNA</p>
<p>Bond    Typical Value ()   Comment / Notes
Ribose Ring     5-membered ring: C1C2C3C4O4
C1C2 ~1.52   Sugar ring CC bond (ribose)
C2C3 ~1.52   Sugar ring CC bond
C3C4 ~1.52   Sugar ring CC bond
C4O4 ~1.45   Sugar ring CO
O4C1 ~1.41   Sugar ring OC
Exocyclic Bonds   <br />
C5C4 ~1.51   Exocyclic from ring to the 5 carbon
C3O3 ~1.42   3-oxygen exocyclic (bridging to phosphate)
C2O2 ~1.41   2-hydroxyl in RNA (absent in DNA)
Glycosidic Bond     Links base ring to the sugar
C1N1 (pyrimidines) /C1N9 (purines)  ~1.47   NC1 is the glycosidic bond; base identity affects small variations (0.01 )
Phosphate Linkages      Tetrahedral geometry at P
PO5   ~1.59   Phosphate bridging to 5-O
PO3   ~1.60   Phosphate bridging to 3-O
PO(non-bridging)   ~1.48   The two non-bridging oxygens (OP1/OP2)
Additional    <br />
O5C5 ~1.44   5-oxygen to 5-carbon (sugarphosphate junction)
O3C3 ~1.43   3-oxygen to 3-carbon (sugarphosphate junction)</p>
<p>Note: DNA is very similar except it lacks the 2-OH (C2O2 bond). The presence of 2-OH slightly shifts ring conformations and can alter bond lengths by ~0.005 relative to deoxyribose.</p>
<p></p>
<ol>
<li>Standard RNA Bond Angles</li>
</ol>
<p>Bond angles around the sugar, exocyclic substituents, and phosphate typically reflect sp-hybridization (109.5) but deviate due to ring strain and resonance. The phosphate group has angles from ~105 to ~111 with bridging vs. non-bridging oxygens, and ~120 between the two non-bridging oxygens.</p>
<p>Table 2: Approximate Bond Angles in RNA</p>
<p>Angle   Typical Value ()   Description / Position
Within Ribose Ring      5-membered ring angles often &lt; 109.5 due to ring strain
C1C2C3 ~101102    Interior angle at C2
C2C3C4 ~102103    Interior angle at C3
C3C4O4 ~105106    Ring angle at C4
C4O4C1 ~109110    Ring angle at O4
O4C1C2 ~106    Angle at C1
Exocyclic (Sugar)     <br />
C5C4C3 ~115    5 exocyclic angle at C4
Phosphate Region        Tetrahedral-like angles around P
O5PO3   ~105110    Bridging O vs bridging O angle (~109 typical)
O5PO(non-bridging)   105111 / 107110   Variation for bridging vs non-bridging O
O1PPO2P (non-bridging)    ~120    Angle between the two non-bridging oxygens
C3O3P   ~118120    Exocyclic angle bridging sugar to phosphate
Glycosidic Region       The angle () is typically measured as a dihedral, but local bond angles are ~110120
O4C1N (glycosidic)  ~108112    e.g., O4C1N1 for pyrimidines, O4C1N9 for purines</p>
<p>Sugar pucker modifies these angles. RNA frequently has C3-endo pucker (A-form), but local C2-endo or mixed puckers cause slight shifts of ~12 or ~0.010.02 in ring bond lengths.</p>
<p></p>
<ol>
<li>Real-World Variation and Influences</li>
</ol>
<p>Though standard values are accurate references, real RNA structures deviate due to:
    1.  Sugar Pucker Changes
       RNA is often in C3-endo (A-form). However, local C2-endo or pseudorotation angles can cause ring angles to shift by a few degrees.
    2.  Base Pairing &amp; Stacking
       Hydrogen bonds (e.g. WatsonCrick, noncanonical pairs) can slightly elongate carbonyl or exocyclic NC bonds by ~0.0030.01.
    3.  Metal Ion Coordination
       Mg (common in functional RNAs) can compress or slightly distort phosphate OPO angles.
    4.  Thermal Fluctuations in MD
       Molecular dynamics simulations at room temperature show bond length fluctuations of ~0.010.02 from their equilibrium.
    5.  Chemical Modification
       Methylations or protonation of bases can shift bond orders, altering standard lengths by a few hundredths of an .</p>
<p>If you see discrepancies &gt; ~0.03 or &gt; ~35 from these standard references, it might indicate either unusual strain, special modifications, or experimental error (like poor crystallographic resolution).</p>
<p></p>
<ol>
<li>Python-Based Tools and Resources</li>
</ol>
<p>Below is a broader set of tools to measure, refine, or even build RNA with standard geometry.
    1.  Barnaba
       Python library for analyzing nucleic-acid structures (PDB or MD trajectories).
       Computes base-pair interactions, torsion angles, sugar puckers, etc.
       Good for validating or comparing your newly built RNA structures angles/lengths to real data.
    2.  MDAnalysis / MDTraj
       Popular Python libraries for reading/writing MD trajectories.
       Provide distance, angle, dihedral computations.
       They do not themselves enforce standard geometry but help confirm it.
    3.  3DNA / DSSR
       Written in C/C++. x3dna-dssr is typically used via CLI, but you can script calls from Python.
       DSSR extracts RNA structural parameters: base pairs, backbone torsions (\alpha,\beta,\gamma,\dots), sugar puckers, etc.
       The older 3DNA rebuild or fiber subprogram can generate ideal duplex forms (A-form) or do partial reconstruction from base-pair parameters.
       Not purely torsionCartesian for single-stranded loops (that is precisely where a NeRF approach can come in).
    4.  PyMOL
       Molecular viewer with a Python API.
       You can script measuring bond lengths/angles or visualize geometry outliers.
    5.  UCSF ChimeraX
       Another advanced viewer with Python-based commands.
       Good for interactive geometry checks and building small linkers.
    6.  Rosetta FARFAR2
       Rosettas RNA de novo folding approach, used for large-scale tertiary structure prediction.
       Enforces standard geometry constraints, but its more heavyweight for folding tasks.
    7.  PHENIX / MolProbity
       Typically used for crystallographic refinement and validation.
       Checks geometry outliers vs. standard bond lengths/angles.
       Scripting possible in PHENIX (Python-based).</p>
<p></p>
<ol>
<li>Converting RNA Torsions to 3D: Adapting mp-nerf or a NeRF Pipeline</li>
</ol>
<p>mp-nerf was designed to convert protein internal coordinates (bond length l, bond angle \theta, dihedral \phi) into 3D in a parallel manner. The concept:
    1.  Protein knowledge base: lists standard bond lengths/angles for NCAC, plus sidechain geometry.
    2.  Algorithm: uses the NeRF method to place each new atom given three references and (l, \theta, \phi).</p>
<p>For RNA:
       Youd replace amino acid residue geometry with nucleotide residue geometry.
       The massively parallel part remains the same: each residue can be built in a parallel block, then unified via rotationtranslation.</p>
<p>6.1. Implementation Outline
    1.  Create an RNA knowledge base (kb_rna.py)
       List bond lengths for PO5, O5C5, C5C4, etc.
       List bond angles at each pivot (O5C5C4, C4C3O3, O3PO5, etc.).
       For the sugar ring, define either:
    1.  A standard set of ring atoms if you want a frozen pucker (e.g., always C3-endo), or
    2.  Torsions for ring closure if you want flexible ring geometry (\nu_0,\nu_1,\ldots).
    2.  Rewrite or Add a rna_fold() function
       Instead of referencing N, CA, C from the protein data, your code references P, O5, C5, C4, C3, O3, C2, C1, base atoms.
       If mp-nerfs approach for sidechains is a separate block, then the base can be treated like a sidechain (i.e., once the sugar is built, place the glycosidic bond and base ring).
    3.  Handling the Sugar Ring
       5-membered ring closure is trickier than a linear chain. Two main strategies:
    1.  Fixed ring approach: Hard-code a standard ribose geometry with a chosen pucker (C3-endo). Then you only define the backbone torsions \alpha,\beta,\gamma,\delta,\epsilon,\zeta plus the glycosidic angle \chi.
    2.  Full ring torsions: If you want the ring to be flexible, define ring torsion angles \nu_0!\dots\nu_4. Then do a small ring-closure routine. This is somewhat like building a proline ring in a protein context.
    4.  Bridging Residues
       Each residue is PO5C5C4C3O3P next residue. You can build them in small parallel blocks and then unify them, the same way mp-nerf does with protein backbone segments.
    5.  Use DSSR or 3DNA for verifying parameters
       If you have an existing RNA structure, run x3dna-dssr -i=struct.pdb --json to see the real bond lengths/angles.
       Confirm they match your knowledge base or adapt to small differences.
    6.  Validation
       Reconstruct a known RNA motif from its known torsions. Then measure RMSD vs. the actual 3D structure to confirm accuracy.</p>
<p>6.2. Practical Example Steps</p>
<p>Suppose you have a short RNA 5-GCAA-3 in a known hairpin. You do:
    1.  Parse the torsions {\alpha,\beta,\gamma,\delta,\epsilon,\zeta,\chi} for each residue from DSSR.
    2.  Initialize the first phosphate in a reference frame (origin).
    3.  For each residue in parallel (like mp-nerf):
       Place P, O5, C5 (using length + angle + torsion).
       Place C4, C3 in the same manner.
       Possibly fix C2, O2, ring closure around O4.
       Attach the base using the glycosidic angle \chi and known base geometry.
    4.  Assemble them into a single chain (rotationtranslation if you do them in separate blocks).
    5.  Compare final 3D coordinates with the original hairpins PDB. Ideally, RMSD is small (0.51.0 or better).</p>
<p></p>
<ol>
<li>Putting It All Together</li>
</ol>
<p>DSSR (and 3DNA) alone does not do a direct torsion3D for single-stranded loops, but it has partial rebuild logic for base-pair steps or standard fiber models. Meanwhile, mp-nerf is excellent for converting internal coordinates to Cartesian quickly but is protein-specific out of the box.</p>
<p>By merging:
       The standard RNA covalent geometry (Tables 1 &amp; 2 above) plus
       A NeRF algorithm (like mp-nerfs parallel approach),</p>
<p>you get a pipeline that can read torsion angles for each residue and produce a valid 3D RNA structure. The approach:
    1.  Provides a backbone that matches known distances and angles.
    2.  Correctly closes or fixes the sugar ring.
    3.  Places each nucleobase with the correct glycosidic bond length and angle (\chi).
    4.  Allows you to integrate with Python-based MD or analysis (e.g., Barnaba, MDAnalysis) for further validation or refinement.</p>
<p></p>
<ol>
<li>References and Further Reading<ol>
<li>Nucleic Acid Knowledge Base or NDB for standard bond-length/angle data:
   http://nucleicacidknowledgebase.org/
   https://ndbserver.rutgers.edu/</li>
<li>DSSR &amp; 3DNA:
   Homepage: http://x3dna.org/
   DSSR paper: Lu et al. (2015) DSSR: an integrated software tool Nucleic Acids Res 43:e142.</li>
<li>mp-nerf:
   Often described in a protein context. E.g., see AlQuraishis pNeRF approach or the repository for massively parallel transformations.</li>
<li>Barnaba:
   Bottaro &amp; Bussi (2019). Nucleic Acids Res. 47(11):e56, also on GitHub.</li>
<li>Rosetta FARFAR2:
   For RNA folding and structure generation with strong geometric constraints.</li>
</ol>
</li>
</ol>
<p>Advanced: If ring closure is needed, consult literature on five-membered ring building approaches (like proline ring closure in proteins or cycpeptidic building in general). Tools like OpenBabels Generate 3D or RDKit can also handle smaller ring conformations, if integrated carefully with your code.</p>
<p></p>
<ol>
<li>Conclusion</li>
</ol>
<p>Yes, you can thoroughly adapt mp-nerf (or any standard NeRF code) for RNA by:
       Substituting a new knowledge base with RNA bond lengths and angles (as tabulated above).
       Implementing minimal or full ring closure for the ribose.
       Possibly labeling the base as a sidechain, using standard geometry for A/U/G/C and the glycosidic angle \chi.
       Optionally referencing DSSR/3DNA to confirm actual parameter values from real PDB structures or to do partial fiber-like rebuilds.</p>
<p>Simultaneously, you can harness Python libraries (Barnaba, MDAnalysis, PyMOL/ChimeraX scripts) to validate that your final geometry remains near standard references. Through this synergy, you get a robust pipeline to convert predicted torsion angles (\alpha, \beta, \gamma, \delta, \epsilon, \zeta, \chi, plus sugar pucker if desired) into 3D atomic coordinates for RNAfully leveraging the high-quality geometry data gleaned from DSSR/3DNA or the Nucleic Acid Knowledge Base.</p>
<p>If you need a deeper dive into code specifics (e.g., rewriting mp-nerfs kb_proteins.py into kb_rna.py), you can replicate the protein code structure, enumerating each RNA atom chain in the correct order, specifying for each new bond:
       bond_length = ,
       bond_angle = ,
       torsion = (the user-supplied or predicted value),
plus ring constraints. Then the parallel approach remains the same.</p>
<h1 id="end-of-document">End of Document.</h1>
<p>Comprehensive Technical Documentation on Standard RNA Geometry Sources and Their Applicability to mp_nerf</p>
<p></p>
<ol>
<li>Introduction</li>
</ol>
<p>When generating three-dimensional coordinates for RNA nucleotides from torsion angles (as in the mp_nerf algorithm), it is essential to use well-established ideal bond lengths and bond angles for the sugarphosphate backbone and the bases. Over the past several decades, multiple groups have collected or refined these geometric parameters through crystallographic surveys or small-molecule databases. Despite incremental updates, the consensus across the community remains that the 1990s work by Parkinson, Gelbin, Clowney, and Berman (often labeled collectively as Parkinson et al. 1996 or the Berman group 1996 data) is the primary standard for RNA geometry.</p>
<p>However, other references sometimes present small numerical refinements or clarify specific subsets (e.g., base ring geometry, definitions of torsions, partial charges). Below is a thorough comparison of these references, emphasizing (1) how they relate to each other, (2) their strengths and weaknesses, and (3) final recommended best practices for choosing a geometry dataset suitable for tools like mp_nerf. This document also addresses critiques or shortcomings of each source, ensuring you have the most balanced and up-to-date guidance possible.</p>
<p></p>
<ol>
<li>Overview of Key References</li>
</ol>
<p>This section briefly introduces the major references frequently cited for RNA geometry. Each subsequent section then discusses them in more detail, noting how each might (or might not) be relevant for building a numeric constants file.
    1.  Parkinson et al. (1996) and Related Berman Group Publications
       Gold standard crystallographic survey from the mid-1990s; these numbers are at the heart of many refinement programs.
    2.  IUPACIUB Joint Commission Recommendations (1982/1983)
       Authoritative for nomenclature (torsion angles , ring numbering), but not for updated numeric geometry.
    3.  Gilski et al. (2019)
       Recent re-analysis of nucleobase geometry (especially WatsonCrick base pairing) using small-molecule crystal structures in the Cambridge Structural Database (CSD).
    4.  Cambridge Structural Database (CSD)
       The worlds largest repository of small-molecule crystallography data. Not a single table, but the raw data underpin many published geometric surveys.
    5.  Nucleic Acid Database (NDB)
       A curated database for nucleic acids. Often mirrors Parkinson et al. (1996), with minor incremental updates.
    6.  Aduri (2007)  AMBER Force Field for Modified Nucleotides
       Provides partial charges and dihedral parameters for modified bases but defers to the 1996 references for standard bond/angle geometry.
    7.  PDB Validation / MolProbity Documentation
       Modern structure validation pipelines (wwPDB, MolProbity) rely on Parkinson (1996) as the reference for nucleic acid geometry.</p>
<p></p>
<ol>
<li>Detailed Examination of Each Reference</li>
</ol>
<p>3.1 Parkinson et al. (1996) / Gelbin, Clowney, Berman (1996)
       Key Publications
       Parkinson G, Vojtechovsky J, Clowney L, Brunger AT, Berman HM (1996) Acta Cryst. D52, 5764.
       Gelbin A, Schneider B, Clowney L, Hsieh S-H, Olson WK, Berman HM (1996) J. Am. Chem. Soc. 118, 519529.
       Clowney L, Jain SC, Srinivasan AR, Westbrook J, Olson WK, Berman HM (1996) J. Am. Chem. Soc. 118, 509518.
       What These Papers Provide
These are systematic surveys of high-resolution nucleic acid structures (both RNA and DNA) that extracted empirical average bond lengths, angles, and torsions for sugar moieties, phosphate groups, and bases. Their data form the ideal geometry often used in standard refinement libraries for crystallography and NMR structure determination.
       Strengths
    1.  Comprehensive and Data-Driven: Among the first large-scale attempts to extract robust average values from crystal structures.
    2.  Widely Adopted: Many structural biology pipelines (e.g., X-PLOR, REFMAC, Phenix, MolProbity) reference these numbers, ensuring consistency with PDB validations.
    3.  Inclusion of SugarPhosphate: Crucially, they detail phosphate bridging angles, sugar ring puckers, glycosidic bond angles, and more, making them a one-stop source for backbone geometry.
       Weaknesses
    1.  Date of Publication: Now ~25+ years old. Some modest refinements have trickled in from expanded data sets.
    2.  Incremental Changes: Later references have observed minor shifts (0.010.02 or a few degrees), especially in base rings.
       Criticisms Addressed
       Age: While often labeled old, no major contradictory study has replaced these data; updates typically confirm only minor numerical adjustments.
       Sparsity of Very High-Resolution RNA Structures in the 90s: Over time, more structures have been added to the NDB, but the fundamental backbone geometry has remained largely unchanged.
       Relevance to mp_nerf
       Primary Source: If mp_nerf needs a canonical set of bond lengths and angles, Parkinson 1996 remains the top choiceparticularly for sugar and phosphate geometry.</p>
<p></p>
<p>3.2 IUPACIUB Joint Commission (Recommendations 1982/1983)
       What These Recommendations Cover
They formalize the nomenclature for polynucleotide chains, defining how to label backbone torsions (, , , , , , ), how to number the sugar ring atoms, the 53 chain direction, etc.
       Strengths
       Authoritative Definitions: Whenever you see , , , , ,  in any RNA or DNA context, these definitions usually trace back to the IUPACIUB documents.
       Nomenclature Consistency: Minimizes confusion about how angles are measured.
       Weaknesses
       Minimal Numeric Data: Not intended to define ideal bond lengths or angles.
       Somewhat Outdated for Fine Geometric Details: These are from the early 1980s, focusing more on naming conventions than empirical geometry.
       Relevance to mp_nerf
       Essential for Torsion Labeling: Great if you want standard naming for angles that mp_nerf might read or produce.
       Not a Source for Bond Lengths/Angles: You still need the 1996 references for numeric geometry.</p>
<p></p>
<p>3.3 Gilski et al. (2019)  Accurate Geometrical Restraints for WC Base Pairs
       Focus of This Study
Published in Acta Cryst. B, Gilski and co-workers re-examined small-molecule crystal structures deposited in the CSD, particularly for nucleobase rings and WatsonCrick hydrogen-bond interactions.
       Strengths
    1.  Recent Analysis: Incorporates more modern, higher-resolution small-molecule data.
    2.  Refined Base Geometry: Bond lengths and angles in the heterocyclic rings might be slightly more precise than the older references.
       Weaknesses
    1.  Limited to Bases: Does not significantly alter sugarphosphate geometry.
    2.  Incremental Adjustments: The reported updates over Parkinson (1996) are typically around 0.010.02 or a few degreesuseful for very high precision, but not a wholesale revision.
       Criticisms Addressed
       Practical Impact: The changes are small enough that many pipeline validations and geometry libraries have not urgently adopted them.
       Relevance to mp_nerf
       Optional Refinement for Bases: If you want state-of-the-art base ring geometry, you can fold Gilskis numbers into your mp_nerf constants.
       Backbone Unchanged: For phosphate and sugar, the older references are still the standard.</p>
<p></p>
<p>3.4 Cambridge Structural Database (CSD)
       What It Is
A massive compilation of small-molecule crystal structures (not just nucleic acids, but all organic and metal-organic compounds).
       Strengths
       Highly Robust: Contains tens of thousands of structures that can be used to statistically derive average geometries for any piece of RNA (ribose, bases).
       Independently Curated: Data from small-molecule crystals often reach higher resolution than large biomacromolecular crystals, giving precise bond lengths.
       Weaknesses
       Requires DIY Data Mining: No built-in universal RNA geometry table is provided.
       Time-Consuming: One must carefully filter for relevant structures (e.g., high resolution, minimal disorder, correct protonation state).
       Criticisms Addressed
       Potential Overlap with Gilski: Gilski (2019) is essentially a curated re-analysis of nucleobases in the CSD. Doing it yourself would presumably yield similar results if you follow rigorous filtering criteria.
       Relevance to mp_nerf
       Indirect: You can rely on published analyses (Parkinson, Gilski) that already tapped into the CSD. No urgent need to re-derive these numbers unless you have a specialized reason.</p>
<p></p>
<p>3.5 Nucleic Acid Database (NDB)
       What It Is
A specialized repository for nucleic acid structures (X-ray, NMR, cryo-EM), curated by the same Rutgers group historically responsible for the 1996 surveys.
       Strengths
       Direct Tabulation of Geometric Values: The NDB often provides updated sugarphosphate bond lengths and angles, close to or slightly refining the 1996 numbers.
       Community-Trusted: Maintained by experts in nucleic acid structure.
       Weaknesses
       Minimal Differences: The data are usually only slightly updated from Parkinson (1996), with typical changes  0.02 or a couple of degrees.
       Not a Drastic Overhaul: The core standard remains effectively the same.
       Criticisms Addressed
       Same as 1996: Some worry it is just re-publishing the old data. However, any expansions typically confirm those older values were robust.
       Relevance to mp_nerf
       Equivalent Choice: Using NDBs geometry tables is effectively the same as using Parkinson (1996). If you want the convenience of a modern website, it is a perfectly valid source.</p>
<p></p>
<p>3.6 Aduri (2007)  AMBER Force Field for Modified Nucleotides
       Main Goal
This study addresses partial charges, torsion parameters, and force-field updates for modified or unusual bases in RNA (e.g., pseudouridine, 2-O-methylated nucleotides).
       Strengths
       Invaluable for Modifications: If your mp_nerf or modeling workflow must handle a large variety of non-standard bases or sugar modifications.
       Force Field Integration: Aligns with well-known AMBER force fields (Cornell et al., parm99, GAFF).
       Weaknesses
       Not a New Source for Standard Geometry: The authors themselves typically reference the same 1996 Berman/Olson data or the Cornell parameters for standard nucleotides.
       Focused on Partial Charges: The impetus is electrostatic and dihedral parameters, not bond length or angle refinement.
       Relevance to mp_nerf
       Only If Handling Modified Residues: Great if you need geometry or charges for unusual groups. For standard A, U, G, C nucleotides, it offers no new geometry.</p>
<p></p>
<p>3.7 PDB Validation / MolProbity
       What It Is
Structural validation tools and guidelines used by the wwPDB, relying on known ideal values to detect outliers.
       Strengths
       Official Confirmation: Reiterates that the PDB strongly references Parkinson (1996) or Engh &amp; Huber (for proteins).
       Community Standard: If your deposited RNA structure has bond lengths/angles far from these references, youll get flagged.
       Weaknesses
       No New Data: MolProbity and wwPDB do not propose alternative geometry sets; they consume the standard references.
       Relevance to mp_nerf
       Consistency: If you adopt the same geometry, your generated coordinates will be consistent with validation norms.</p>
<p></p>
<ol>
<li>Comparing These References and Their Numeric Discrepancies</li>
</ol>
<p>A recurring theme is that nearly all sources agree on the numeric values within very small margins (~0.02, 12). The 1996 references remain the de facto standard, while Gilski (2019) offers incremental updates for base rings. Below is a concise summary of how each compares:
       Parkinson (1996) vs. NDB:
The NDB is essentially an evolution of the same dataset. Differences are typically minuscule.
       Parkinson (1996) vs. Gilski (2019):
Gilski improves base ring geometry slightly. The sugarphosphate geometry remains nearly identical. The changes rarely exceed 0.010.02.
       IUPAC:
Provides definitions, not numeric tables, so no numerical conflict.
       Aduri (2007):
Defers to earlier references for unmodified nucleotides. Again, no conflict.
       CSD (Raw):
The underlying source for many of these surveys. You could re-derive the same means if you replicate Gilskis or Bermans methods.</p>
<p>Hence, there is no major contradiction among references; it is more a matter of how precise you want to be (i.e., whether you incorporate Gilskis minor updates for bases).</p>
<p></p>
<ol>
<li>Strengths and Weaknesses: A Consolidated Table</li>
</ol>
<p>Reference   Strengths   Weaknesses
Parkinson / Gelbin / Clowney (1996) Berman Group    - Empirically derived from high-res structures  - Backbone + base geometry covered  - Still widely used in refinement  - Gold standard for sugarphosphate    - Older dataset (1990s)  - Minor numerical updates exist in newer references
IUPACIUB (1982/83) - Authoritative naming of torsions (, , , , etc.)  - Official polynucleotide nomenclature   - Not an updated numeric source  - Dated in terms of modern structural knowledge
Gilski et al. (2019)    - Latest small-molecule data for base rings  - Focuses on WC pair geometry  - Potentially more precise nucleobase angles/lenths    - Limited to base geometry  - Differences from 1996 sets are small (~0.010.02)
Cambridge Structural Database (CSD) - Vast repository of small-molecule crystals  - Can yield highly precise bond lengths  - The foundation for Gilskis approach   - No single standard table provided  - Must do extensive data mining &amp; filtering
Nucleic Acid Database (NDB) - Modern aggregator of NA structures  - Publishes geometry tables similar to Parkinson (1996)  - Occasionally updated   - Differences from 1996 are minor  - Not a fundamentally new dataset
Aduri (2007)    - AMBER force field parameters for modified nucleotides  - Great for partial charges &amp; unusual bases    - Not an independent geometry source  - Reuses standard references for unmodified nucleotides
PDB Validation / MolProbity - Official pipeline referencing Parkinson (1996)  - Reinforces community standard for geometry  - No new data  - Merely restates established norms</p>
<p></p>
<ol>
<li>Recommendations for mp_nerf Implementation</li>
</ol>
<p>Given the convergence among these sources, the path forward for a reliable constants file in mp_nerf is fairly straightforward:
    1.  Default to Parkinson et al. (1996)
       Use these bond lengths and angles for the sugarphosphate backbone.
       These values are robust, widely accepted, and ensure consistency with PDB validation.
    2.  Optionally Incorporate Gilski (2019) for Bases
       If mp_nerf idealizes base ring structures in a highly precise manner (e.g., if your application is sensitive to subtle ring metrics), you can overwrite the basespecific parameters with Gilskis updates.
       The backbone geometry can remain from the 1996 data.
    3.  Leverage NDB as a Backup
       If it is more convenient to pull from a modern online database, the NDBs valence geometry tables effectively replicate the 1996 numbers (sometimes with trivially small refinements).
       This ensures you are still aligned with standard practice.
    4.  Retain IUPACIUB for Torsion Naming
       For clarity and consistency, use the IUPAC scheme to label , , , , , , and  angles.
       This step avoids confusion in angle definitions.
    5.  Add Specialized Data for Modified Nucleotides (If Needed)
       If mp_nerf must handle modifications such as pseudouridine or 2-O-methyl groups, consult Aduri (2007) or relevant AMBER force-field parameter sets for partial charges and specialized geometry.
       The standard backbone bond lengths typically remain the same, but ring or substituent parameters may differ.</p>
<p>By following this scheme, you ensure your geometry is in line with the broad consensus of structural biology tools and references.</p>
<p></p>
<ol>
<li>Frequently Asked Questions (FAQ)<ol>
<li>Why not use Gilski (2019) for everything?
Gilskis publication focuses heavily on base rings and WatsonCrick hydrogen bonds. It does not supersede the backbone geometry from 1996. The numeric changes for the bases themselves are small enough (&lt;0.02) that for most routine modeling or structure building, the older 1996 numbers are perfectly acceptable.</li>
<li>Are the 1996 data too old?
Despite their age, these data have withstood the test of time. Subsequent expansions of the dataset typically confirm only minimal numerical shifts. No major contradictions or overhauls have emerged in more recent literature.</li>
<li>Do I really need partial charges or dihedral parameters from Aduri (2007)?
Only if you plan to do molecular dynamics or energy calculations involving non-standard nucleotides. If mp_nerf is strictly for coordinate generation of standard A, U, G, C residues, the 1996 bond lengths/angles are sufficient.</li>
<li>Could I do my own CSD-based analysis?
In theory, yesmany researchers do. But it is time-consuming, and the published results from Parkinson, Gilski, and the NDB already represent thorough analyses of that same data. You would likely find near-identical numbers unless you include new structures or different filtering.</li>
<li>What about naming conventions beyond , like sugar ring puckers?
The IUPACIUB recommendations remain the default for naming the sugar ring atoms, while standard pucker notation (C3-endo, C2-endo, etc.) is widely accepted. The numeric definitions of ring conformation angles can be found in the references from the Berman group and in typical computational chemistry software manuals (e.g., AMBER docs).</li>
</ol>
</li>
</ol>
<p></p>
<ol>
<li>Conclusion</li>
</ol>
<p>In summary, Parkinson et al. (1996) (also collectively described as the Berman group 1996 data) remains the central, most authoritative source for standard RNA geometry. Although Gilski et al. (2019) provides slight improvements for nucleobase rings, the sugarphosphate parameters remain effectively unchanged. Alternative sources like the NDB or older IUPAC documentation typically echo these fundamental 1996 values; the CSD underlies them with raw small-molecule data, and the PDB validation pipeline further cements these numbers as the community standard.</p>
<p>For a tool like mp_nerfwhich reconstructs 3D coordinates from torsion anglesthe best practice is to adopt the Parkinson (1996) data for sugarphosphate bond lengths and angles, optionally layering in Gilski (2019) for nucleobase ring geometry if you want the most current refinements. All other references are either duplicative, focus on nomenclature, or address more specialized topics (modified nucleotides, partial charges).</p>
<p>By heeding these guidelines, your geometry constants file will be both authoritative and consistent with widely used structural biology protocolsensuring that the final result is better than the sum of its parts and firmly aligned with the established consensus in RNA structural science.</p>
<p></p>
<p>References and Suggested Reading
    1.  Parkinson G, Vojtechovsky J, Clowney L, Brunger AT, Berman HM (1996). Acta Cryst. D52, 5764.
    2.  Gelbin A, Schneider B, Clowney L, Hsieh S-H, Olson WK, Berman HM (1996). J. Am. Chem. Soc. 118, 519529.
    3.  Clowney L, Jain SC, Srinivasan AR, Westbrook J, Olson WK, Berman HM (1996). J. Am. Chem. Soc. 118, 509518.
    4.  IUPACIUB Joint Commission on Biochemical Nomenclature (1983). Eur. J. Biochem., 131, 915.
    5.  Gilski M, et al. (2019). Acta Cryst. B75, 235254.
    6.  Cambridge Structural Database (CSD): https://www.ccdc.cam.ac.uk/
    7.  Nucleic Acid Database (NDB): http://ndbserver.rutgers.edu
    8.  Aduri R, et al. (2007). J. Chem. Theory Comput. 3(4), 14641475.
    9.  MolProbity / PDB Validation: http://molprobity.biochem.duke.edu</p>
<p></p>
<p>End of Document</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": [], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>